- --------------------- Portal ---------------------
    - Daily Document
- Get the 3D-printer up and running
    - Install defect detection
    - Research resin 3d printers
        - [acrea3d.com/training/microfluidic resins spectra matching/](https://acrea3d.com/training/microfluidic-resins-spectra-matching/)
        - I need an 8k resin 3d printer that I can replace the light source with 385, 365 nm 
    - Try to repair or salvage parts from the Makerbot
        - I've decided to deconstruct it and turn it into a voron 3d printer
    - Print the first truly functional part (Raspberry Pi Case)
    - --------------------- Portal ---------------------
        - Daily Document
    - --------------------- Portal ---------------------
        - Daily Document
    - --------------------- Portal ---------------------
        - Daily Document
    - Work through CAD Modeling course
        - [3D design for 3D printing tutorials - YouTube](https://youtube.com/playlist?list=PLGqRUdq5ULsONnjEEPeBxxStEsobDKAtV)
        - [Onshape 3D Modelling - YouTube](https://youtube.com/playlist?list=PLGqRUdq5ULsMDOxmu10AGPDIOkzNYu7D7) 
- Decide on a list of supplementary lab equipment needed
    - --------------------- Portal ---------------------
        - INDEX
    - Research what kind of equipment is required for microfluidics fabrication and protoplast isolation
        - I'm waiting for 3D-printer parts
            - These parts will help maintain and calibrate the printer to produce more consistent prints
            - Install the BLtouch sensor
        - DIY Spectrophotometers

        - **Fluorescent microscope** 
            - [Fluorescent microscopes are amazing! - YouTube](https://www.youtube.com/watch?v=iYoAiqufB7w) 
            - First mod my light microscope 
            - 
            - --------------------- Portal ---------------------
                - Daily Document
            - Can you compare the OpenFlexure microscope the the FlyPi microscope

            - Flypi
        - incubator
        - Bioreactor
        - Hydroponic system
    - Extra information 
- Create Glowing Plants
    - 
        - 
    - find/pick the plasmid containing GFP gene
    - insert gene into bacteria
        - Agrobacterium
    - purfiy the gene
    - insert into plant
    - perform microprogation 
    - bing
        - Protocols
            - Plasmid isolation
                - Column Based Purfication
                - Chemical Purification
            - **Protocol: Kit-free Alkaline Lysis Plasmid Miniprep** 
                1. Prepare the following solutions:
                    1. **Solution I - Resuspension Buffer**
                        1. 25 mM Tris-HCl (pH 8)
                        2. 50 mM glucose
                            1. 49 mM sucrose
                            2. Glucose should also be added to maintain osmolarity and prevent the buffer from bursting the cells.
                        3. 10 mM EDTA
                            1. EDTA chelates divalent metals (primarily magnesium and calcium). Removal of these cations destabilizes the cell membrane. It also inhibits DNases.
                        - Store Solution I at 4°C
                        - Sure, here are the calculations for each of the parts of solution 1:
                            - To make 31.25 mL of 50 mM glucose from a 1 M stock solution, you need to add **1.5625 mL of 1 M glucose** to a volumetric flask and fill it up to 31.25 mL with distilled water.
                            - To make 31.25 mL of 25 mM Tris-HCl pH 8.0 from a 1 M stock solution, you need to add **0.78125 mL of 1 M Tris-HCl** pH 8.0 to a volumetric flask and fill it up to 31.25 mL with distilled water.
                            - To make 31.25 mL of 10 mM EDTA from a 0.5 M stock solution, you need to add **0.625 mL of 0.5 M EDTA** to a volumetric flask and fill it up to 31.25 mL with distilled water.
                        - After adding each component, you should mix the solution well and label it as solution 1.
                    2. **Solution II - Denaturing Solution**
                        - ![](local:///home/epas/remnote/remnote-GdJuzRmdKsXpCgrQ2/files/ALiEmCyp2J_UAH1T9HjX93wWEh7P6SbK_v70kVgb-XqCqgCvIwg2PXh2OAbwf1_-hVBpPFktzEzbG_okz-Xpov7I9YvqJJb059inOwJ9ApiMSgBYIjDgJkq_JTaITqwD.png) 
                        - 0.2 N NaOH
                        - 1.0% SDS
                        - Store Solution II at room temperature
                        - Sure, here are the calculations for solution 2:
                        - After adding both components, you should mix the solution well and label it as solution 2.
                    3. **Solution III - Renaturing Solution (Potassium Acetate)**
                        1. 120 mL 5M Potassium acetate (**sodium aceate**) 
                            1. 120 ml of 3 M sodium acetate 
                                1. To make 120 mL of 3 M sodium acetate using solid sodium acetate, you can follow these steps:
                                    - Weigh out 29.5 g  of sodium acetate (anhydrous) and add to a suitable container.
                                    - Add 84 mL of distilled water to the container.
                                    - Dissolve the sodium acetate by adding a magnetic stir bar into the container and placing on a magnetic stirrer.
                                    - Adjust the pH to 5.2 with glacial acetic acid.
                                    - Allow the solution to cool overnight.
                                    - Adjust the pH once more to 5.2 with glacial acetic acid.
                                    - Add distilled water until the volume is correct
                                    - [Filter-sterilize the solution](undefined)[1](https://www.bing.com/ck/a?!&&p=e546fba9d7dec9dfJmltdHM9MTY4OTg3ODA1MA&ptn=3&hsh=3&fclid=04c0e38d-272c-11ee-8460-1e17296e2f47&u=a1aHR0cHM6Ly93d3cuYWF0YmlvLmNvbS9yZXNvdXJjZXMvYnVmZmVyLXByZXBhcmF0aW9ucy1hbmQtcmVjaXBlcy9zb2RpdW0tYWNldGF0ZS0zLW0tcGgtNS0y&ntb=1&bc=1).
                            - **alternative for potassium acetate for plasmid extraction** 
                            - To adjust the alkaline lysis plasmid miniprep protocol to use sodium acetate, you may need to change the concentration and pH of the acetate solution, as well as the volume of the ethanol or isopropanol used for precipitation. [This is because sodium acetate has different solubility and buffering properties than potassium acetate, and may affect the efficiency of the neutralization and precipitation steps1](https://www.bing.com/ck/a?!&&p=9dde133f1ee4cb26JmltdHM9MTY4OTg3MjgxNA&ptn=3&hsh=3&fclid=d3a43d9a-271f-11ee-a9d6-af4e1b70b329&u=a1aHR0cHM6Ly8yMDEzLmlnZW0ub3JnL1RlYW06R2xlbmRhbGVfQ0NfQVovUHJvdG9jb2xzL0Fsa2FsaW5lTHlzaXM&ntb=1&bc=1)[2](https://www.bing.com/ck/a?!&&p=515ba45e52bf04ffJmltdHM9MTY4OTg3MjgxNA&ptn=3&hsh=3&fclid=d3a47b04-271f-11ee-aae1-7057d18c9c97&u=a1aHR0cHM6Ly93d3cubGFib3JhdG9yeW5vdGVzLmNvbS9wcm90b2NvbC1wbGFzbWlkLWlzb2xhdGlvbi1ieS1hbGthbGluZS1seXNpcy1tZXRob2QtbWluaXByZXAv&ntb=1&bc=1).
                            - [One possible way to adjust the protocol is to use a 3 M sodium acetate solution at pH 5.2 instead of a 3 M potassium acetate solution at pH 4.8](https://www.bing.com/ck/a?!&&p=b182da821bbff420JmltdHM9MTY4OTg3MjgxOA&ptn=3&hsh=3&fclid=d5e10ff9-271f-11ee-887a-24b4b0f48731&u=a1aHR0cHM6Ly9iaW8ubGlicmV0ZXh0cy5vcmcvQm9va3NoZWx2ZXMvQmlvdGVjaG5vbG9neS9CaW8tT0VSXyUyOENVTlklMjkvMDglM0FfQW5hbHl6aW5nX0ROQS84LjA5JTNBX0ROQV9NaW5pcHJlcF9ieV9BbGthbGluZV9MeXNpc18lMjhBY3Rpdml0eSUyOQ&ntb=1&bc=1)[3](https://www.bing.com/ck/a?!&&p=b182da821bbff420JmltdHM9MTY4OTg3MjgxOA&ptn=3&hsh=3&fclid=d5e10ff9-271f-11ee-887a-24b4b0f48731&u=a1aHR0cHM6Ly9iaW8ubGlicmV0ZXh0cy5vcmcvQm9va3NoZWx2ZXMvQmlvdGVjaG5vbG9neS9CaW8tT0VSXyUyOENVTlklMjkvMDglM0FfQW5hbHl6aW5nX0ROQS84LjA5JTNBX0ROQV9NaW5pcHJlcF9ieV9BbGthbGluZV9MeXNpc18lMjhBY3Rpdml0eSUyOQ&ntb=1&bc=1). [You may also need to increase the volume of ethanol or isopropanol to 2.5 times the volume of the supernatant, instead of 0.7 times, to ensure complete precipitation of the plasmid DNA](https://www.bing.com/ck/a?!&&p=0dafe6e712029db3JmltdHM9MTY4OTg3MjgyMw&ptn=3&hsh=3&fclid=d8d635e7-271f-11ee-a41a-77e29c86c772&u=a1aHR0cHM6Ly9hc2thYmlvbG9naXN0LmFzdS5lZHUvYWxrYWxpbmUtbHlzaXM&ntb=1&bc=1)[4](https://www.bing.com/ck/a?!&&p=0dafe6e712029db3JmltdHM9MTY4OTg3MjgyMw&ptn=3&hsh=3&fclid=d8d635e7-271f-11ee-a41a-77e29c86c772&u=a1aHR0cHM6Ly9hc2thYmlvbG9naXN0LmFzdS5lZHUvYWxrYWxpbmUtbHlzaXM&ntb=1&bc=1). However, these adjustments may not work for all types or sizes of plasmids, so you may need to optimize the protocol for your specific conditions and test the quality and quantity of the DNA before using it for downstream applications.
                        2. 23 mL glacial acetic acid
                        3. 57 mL of dH2O
                        - Store Solution III at 4°C
                        - Sure, here are the calculations for solution 3:
                        - After adding both components, you should mix the solution well and label it as solution 3.
                2. Grow 2 mL [overnight culture of bacteria](https://www.addgene.org/protocols/inoculate-bacterial-culture/) from single colonies containing your plasmid of interest.
                    1. If your tubes are only able to hold 1000 μL, you can modify the protocol by using smaller volumes of liquid. For example, instead of using 2 mL of bacterial culture for each sample, you can use 1 mL. However, you will also need to adjust the volumes of the solutions accordingly. For example, instead of using 250 μL of solution 1, you can use 125 μL. You can use the following table as a guide:
                        1. ![](local:///home/epas/remnote/remnote-GdJuzRmdKsXpCgrQ2/files/SWsPfHftO0A-Am6RqXdjpWnr_Bmt8AfCEUD9xfbBrbO096ZekCWzXRphOFi5tXUKEDZ_43D7lYUzNZh4qf_0BLf9b5XeDFGHu4i91fmPubxhC0EjopbTywdMPdy5EK-m.png) 
                    - ***Pro-Tip*** Refer to appropriate DNA prep protocol for volume of bacteria to grow (low copy plasmids require larger cultures).
                    - ***Pro-Tip*** If your entire overnight culture cannot fit into a single centrifuge tube, aliquot it into several tubes/bottles.
                3. Add 1 mL of the stock culture to a 1.5 mL microfuge tube.
                4. Centrifuge in microfuge tube at 10,000 g for 30 sec.
                5. Pour off the supernatant, being careful not to disturb the bacterial pellet.
                6. Resuspend the pellet in 100 μL of cold Solution I.
                    -  __**Note:**__  __ __   __This step gets all of the bacteria back into suspension, but within a smaller volume of buffer that is compatible with the next solution.__  
                    - 
                7. Vortex the solution for 2 min or until all bacteria are fully resuspended.
                8. Add 200 μL of Solution II and invert the tube carefully 5 times to mix the contents. The contents will become clear and thicker as the proteins and DNA are denatured.
                    - ![](https://media.addgene.org/data/easy-thumbnails/filer_public/cms/filer_public/db/b9/dbb9c846-2c86-4f66-ad9a-fc25fd915f56/exclamation.jpg__50x48_crop_subsampling-2_upscale.png) 
                    -  __**Note:**__  __ __   __This step causes the bacteria to lyse, releasing their contents, including plasmid DNA, into solution.__ 
                    - ***Pro-Tip*** Do not vortex at this stage or the genomic DNA will become sheared and will therefore contaminate your purified plasmid DNA.
                9. Incubate solution on ice for 5 min.
                10. Add 150 μL of cold Solution III to each tube.
                    -  __**Note:**__  __ __   __This step brings the pH back down causing the proteins and genomic DNA to precipitate, while leaving the smaller plasmids free in solution.__ 
                11. Mix by inverting several times. A white precipitate will be formed which contains the bacterial proteins and genomic DNA.
                12. Incubate tube on ice for 5 min.
                13. Centrifuge the tube for 5 min at 12,000 g.
                    -  __**Note:**__  __ __   __Pellet contains proteins, cell fragments, salt and other extra particles from solutions.__ 
                    -  __**Note:**__  __ __   __Supernatant contains the plasmid DNA separated from bacterial chromosomes.__ 
                14. Collect the supernatant into a new tube by pipetting or carefully pouring.
                15.  __(Optional)__  Add 5 μL of 2 mg/mL RNase A to the supernatant in the new tube and incubate at 37 °C for 5 min.
                    -  __**Note:**__  __ __   __Ribonuclease A (RNase A) is a pancreatic ribonuclease that digests single-stranded RNA.__ 
                16.  __(Optional)__  Perform phenol-chloroform extraction - see protocol [below](https://www.addgene.org/protocols/purify-plasmid-dna/#phenol).
                    -  __**Note:**__  __ __   __Phenol-chloroform extraction removes remaining contaminant proteins and RNase A from the DNA sample. When phenol is mixed with the aqueous solution containing DNA, proteins will move into the phenol phase and will be separated from the aqueous DNA.__ 
                    - **Protocol: Phenol-[NO-Chloroform] Extraction of DNA Samples** 
                        - --------------------- Portal ---------------------
                            - Daily Document
                        1. Add an equal volume of TE-saturated phenol to the aqueous DNA sample.
                            - ![](https://media.addgene.org/data/easy-thumbnails/filer_public/cms/filer_public/db/b9/dbb9c846-2c86-4f66-ad9a-fc25fd915f56/exclamation.jpg__50x48_crop_subsampling-2_upscale.png) 
                            - ***Pro-Tip*** Water-saturated phenol can be used if TE-saturated is not available.
                        2. Vortex microfuge tube for 30-60 sec.
                        3. Centrifuge the tube for 5 min at room temperature on the highest setting.
                            -  __**Note:**__  __ __   __You should see clearly separated layers:__ 
                                -  __**Top Phase -**__  __ __   __Aqueous DNA phase__ 
                                -  __**Middle phase -**__  __ __   __A white layer may appear, consisting of precipitated protein particles__ 
                                -  __**Bottom phase -**__  __ __   __Organic phase (phenol or protein?)__  
                        4. Pipet the aqueous DNA layer and place it in a new microfuge tube.
                        5. Add 0.1 volume of 3 M sodium acetate (pH 5.2) and 2.5 volumes of ice-cold ethanol to the recovered aqueous DNA layer.
                        6. Mix well and incubate at -20°C for at least 1 hour or overnight to precipitate the DNA.
                        7. Centrifuge the tube for 15 min at 4°C on the highest setting.
                            - Note: A white pellet of DNA should be visible at the bottom of the tube.
                        8. Carefully remove the supernatant and wash the pellet with 70% ethanol.
                        9. Centrifuge the tube again for 5 min at 4°C on the highest setting and remove the supernatant.
                        10. Air-dry the pellet for 10-15 min or until no ethanol remains.
                        11. Resuspend the pellet in TE buffer or water.
                17. 
                - **Protocol: Ethanol Precipitation** 
            - Isolation of DNA from Agarose and Polyacrylamide Gels
            - Agrobacterium **Bacterial Transformation** 
            - **Micropropagation Protocol for aquatic plants** 
            - **plasmid extraction from ecoli** NOTES
                - You can also  __**automate**__  the process using a pipetting platform such as **CyBio ® FeliX**². The plasmid DNA you obtain should be suitable for modifying plants through agrobacterium⁴. I hope this helps.😊
                - **plasmid DNA purity** 
                    - If I were to use the DNA purified through this protocol will the DNA still be contaminated with Non plasmid DNA
                        - That depends on the quality of your plasmid DNA preparation. You can measure the purity of your plasmid DNA by using the **absorbance ratio** at different wavelengths, such as 230, 260, and 280 nm¹. The ideal ratio of absorbance at 260 nm to the absorbance at 280 nm (A260/A280) for DNA is **1.8**, but it can be in the range of 1.7–1.9³. A lower ratio may indicate the presence of contaminants such as proteins, genomic DNA, or chaotropic agents¹. To ensure high purity and yield of plasmid DNA, you should also consider factors such as culture cell density, duration of growth, medium used, the type of plasmid (high or low copy number), the size of the insert, and the host strain used⁴. Excessively high E. coli cell density can overload the purification system and result in poor quality plasmid DNA⁴.
                            - Source: Conversation with Bing, 7/4/2023
                                - (1) Plasmid Preps: Different Purity, Different Quantities ... - Addgene. [https://blog.addgene.org/plasmid-preps-different-purity-different-quantities-different-uses.](https://blog.addgene.org/plasmid-preps-different-purity-different-quantities-different-uses.)
                                - (2) Plasmid DNA Purification Support—Getting Started. [https://www.thermofisher.com/us/en/home/technical-resources/technical-reference-library/nucleic-acid-purification-analysis-support-center/plasmid-dna-purification-support/plasmid-dna-purification-support-getting-started.html.](https://www.thermofisher.com/us/en/home/technical-resources/technical-reference-library/nucleic-acid-purification-analysis-support-center/plasmid-dna-purification-support/plasmid-dna-purification-support-getting-started.html.)
                                - (3) General Considerations for Plasmid DNA Preparation - MilliporeSigma. [https://www.sigmaaldrich.com/US/en/technical-documents/technical-article/genomics/dna-and-rna-purification/general-considerations-for-plasmid-dna-preparation.](https://www.sigmaaldrich.com/US/en/technical-documents/technical-article/genomics/dna-and-rna-purification/general-considerations-for-plasmid-dna-preparation.)
                                - (4) Recovering Plasmid DNA from Bacterial Culture - Addgene. [https://www.addgene.org/protocols/purify-plasmid-dna/.](https://www.addgene.org/protocols/purify-plasmid-dna/.)
                        - 
                - **freeze dried bacteria viability**
                    - How long can freeze dried bacteria such as e coli stay in the freezer and maintain viability
                        - The viability of freeze-dried bacteria depends on several factors, such as the bacterial strain, the freeze-drying method, the storage temperature, and the storage time. Not all bacteria can be successfully freeze-dried, and some strains may die rapidly once freeze-dried¹. The best way to determine the viability of a strain is to empirically evaluate its stability post–freeze-drying while maintaining a live culture as a backup¹.
                        - 
                        - According to one study, E. coli cells can be stored for up to **3 years** at -20°C or up to **10 years** at -80°C with minimal loss of viability². However, these results may vary depending on the cell density, the freezing and thawing rate, and the presence of cryoprotectants². Another study found that a mixture of bacterial and yeast strains isolated from kefir grains showed better survival rates when freeze-dried in milk medium than in water medium, and when stored at **4°C** than at 25°C or 37°C. The viability of the freeze-dried mixture decreased by about **1 log CFU/mL** after **180 days** of storage at 4°C³. Therefore, it is advisable to use an appropriate medium for freeze-drying bacteria and to store them at low temperatures for optimal viability.
                        - 
                        - Source: Conversation with Bing, 7/4/2023
                        - (1) Storing Bacterial Samples for Optimal Viability | Thermo Fisher .... [https://www.thermofisher.com/us/en/home/industrial/microbiology/microbiology-learning-center/storing-bacterial-samples-optimal-viability.html.](https://www.thermofisher.com/us/en/home/industrial/microbiology/microbiology-learning-center/storing-bacterial-samples-optimal-viability.html.)
                        - (2) Bacterial Freeze Drying Protocol - University of Washington. [https://isolab.ess.washington.edu/resources/manuals/OPS_Diagnostics_BacterialFreezeDryingProtocol.html.](https://isolab.ess.washington.edu/resources/manuals/OPS_Diagnostics_BacterialFreezeDryingProtocol.html.)
                        - (3) Effect of freeze-drying on viability and in vitro probiotic properties .... [https://www.cambridge.org/core/journals/journal-of-dairy-research/article/effect-of-freezedrying-on-viability-and-in-vitro-probiotic-properties-of-a-mixture-of-lactic-acid-bacteria-and-yeasts-isolated-from-kefir/C740F162D779B4438EE04D07B630BC22.](https://www.cambridge.org/core/journals/journal-of-dairy-research/article/effect-of-freezedrying-on-viability-and-in-vitro-probiotic-properties-of-a-mixture-of-lactic-acid-bacteria-and-yeasts-isolated-from-kefir/C740F162D779B4438EE04D07B630BC22.)
                    - How do I revive bacteria that have been freeze-dried?

                        - To revive freeze-dried bacteria, you need to rehydrate them with an appropriate liquid and transfer them to a suitable growth medium. The exact procedure may vary depending on the type of bacteria and the freeze-drying method, but here is a general protocol¹:
                        - 
                        - Disinfect the freeze-dried vial with alcohol and open it carefully using a flame or a glass tubing cutter.
                        - Dilute the bacterial powder in distilled water or physiological saline.
                        - Centrifuge to remove any components of the lyophilized mixture (often contains sugars as lyoprotectants).
                        - Resuspend the bacterial pellet in physiological saline or culture broth.
                        - Add to the different culture solutions in whichever cell concentration you wish.
                        - 
                        - You should use the medium and incubation conditions specified on the product sheet when first reviving strains to ensure optimal conditions for recovery. You should also check the viability of the bacteria after rehydration and store them properly for future use.
                        - 
                        - Source: Conversation with Bing, 7/4/2023
                        - (1) Bacteriology Culture Guide | ATCC. [https://www.atcc.org/resources/culture-guides/bacteriology-culture-guide.](https://www.atcc.org/resources/culture-guides/bacteriology-culture-guide.)
                        - (2) Reviving Freeze-Dried Microorganisms Instructional Guide - ATCC. [https://www.atcc.org/-/media/product-assets/documents/instruction-sheets/reviving-freeze-dried-microorganisms-instructional-guide.pdf?rev=4adbf71c734d4a4ca77aaafe006ec476&hash=9FCB869F6D674982161B5BF3DCACAFEF.](https://www.atcc.org/-/media/product-assets/documents/instruction-sheets/reviving-freeze-dried-microorganisms-instructional-guide.pdf?rev=4adbf71c734d4a4ca77aaafe006ec476&hash=9FCB869F6D674982161B5BF3DCACAFEF.)
                        - (3) Can Salmonella Survive Freeze-drying? - FAQS Clear. [https://www.faqsclear.com/can-salmonella-survive-freeze-drying/.](https://www.faqsclear.com/can-salmonella-survive-freeze-drying/.)
                        - (4) How do you revive freeze dried bacteria? - servicepoint4you.eu. [https://soak.servicepoint4you.eu/qa/04a470049146d/.](https://soak.servicepoint4you.eu/qa/04a470049146d/.)
                        - (5) Does Freezing Really Kill Bacteria? – What the Science Says. [https://dailyhomesafety.com/does-freezing-kill-bacteria/.](https://dailyhomesafety.com/does-freezing-kill-bacteria/.)
                - **ecoli bacteria DH10B endA negative** 
                    - According to the information I found, **E. coli DH10B** is a strain that is **endA1** deficient, which means it lacks the DNA-specific endonuclease I enzyme²¹. This improves the quality of plasmid DNA isolations from this strain². So, to answer your question, **yes**, E. coli DH10B is **endA negative**.
                        - Source: Conversation with Bing, 7/4/2023
                            - (1) Genetic Markers in E. coli and EndA+ versus EndA- Strains. [https://www.promega.com/-/media/files/resources/technical-references/genetic-markers-of-e-coli.pdf.](https://www.promega.com/-/media/files/resources/technical-references/genetic-markers-of-e-coli.pdf.)
                            - (2) NEB® 10-beta Electrocompetent E. coli | NEB. [https://www.neb.com/products/c3020-neb-10-beta-electrocompetent-e-coli.](https://www.neb.com/products/c3020-neb-10-beta-electrocompetent-e-coli.)
                            - (3) NEB® 10-beta Competent E. coli (High Efficiency) | NEB. [https://www.neb.com/products/c3019-neb-10-beta-competent-e-coli-high-efficiency.](https://www.neb.com/products/c3019-neb-10-beta-competent-e-coli-high-efficiency.)
                            - (4) Escherichia coli - Wikipedia. [https://en.wikipedia.org/wiki/Escherichia_coli.](https://en.wikipedia.org/wiki/Escherichia_coli.)
                - **plasmid pJE202** 
                    - Plasmid **pJE202** is also known as **pVIB**. It is an **ampicillin-resistant** plasmid that expresses the **Lux operon genes** from Vibrio fischeri, a bioluminescent bacterium¹. The plasmid can be used to create bioluminescent E. coli by transforming them with the plasmid¹. The plasmid is already present in E. coli DH10B strain, which can be grown at 30°C or below and observed in a dark room for bioluminescence¹. The plasmid is not related to pEG202, which is a yeast two-hybrid vector for expressing LexA fusion proteins².
                        - Source: Conversation with Bing, 7/4/2023
                            - (1) pJE202/pVIB to Create Bioluminescent Bacteria - The ODIN. [https://www.the-odin.com/pje202-pvib-to-create-bioluminescent-bacteria/.](https://www.the-odin.com/pje202-pvib-to-create-bioluminescent-bacteria/.)
                            - (2) pEG202 Sequence and Map - SnapGene. [https://www.snapgene.com/plasmids/yeast_plasmids/pEG202.](https://www.snapgene.com/plasmids/yeast_plasmids/pEG202.)
                    - how many base pairs long is the plasmid

                        - The plasmid **pJE202** is **13,009 base pairs** long¹. You can see the full sequence and map of the plasmid here: [https://novoprolabs.com/vector/Vg4zdeoa.](https://novoprolabs.com/vector/Vg4zdeoa.) The plasmid is different from pEG202, which is a yeast two-hybrid vector that is 10,166 base pairs long².
                        - 
                        - Source: Conversation with Bing, 7/4/2023
                        - (1) pJE202 vector map and sequence - novoprolabs.com. [https://novoprolabs.com/vector/Vg4zdeoa.](https://novoprolabs.com/vector/Vg4zdeoa.)
                        - (2) pEG202 Sequence and Map - SnapGene. [https://www.snapgene.com/plasmids/yeast_plasmids/pEG202.](https://www.snapgene.com/plasmids/yeast_plasmids/pEG202.)
                        - (3) pJE202/pVIB to Create Bioluminescent Bacteria - The ODIN. [https://www.the-odin.com/pje202-pvib-to-create-bioluminescent-bacteria/.](https://www.the-odin.com/pje202-pvib-to-create-bioluminescent-bacteria/.)
                    - Will DNA Ladder 1Kb be sufficient to perform a gel electrophoresis for pJE202 isolation
                        - Yes, DNA Ladder 1 Kb is suitable for pJE202 isolation, since pJE202 is a plasmid of 13,009 bp, which falls within the range of the DNA ladder. You can use the reference band of 1,500 bp to orient yourself on the gel and compare the size of your plasmid.
                        - DNA Ladder 1 Kb is a solution of DNA fragments of different lengths ranging from 100 bp to 15,000 bp, used as a reference to estimate the size of unknown DNA molecules in gel electrophoresis¹. Whether DNA Ladder 1 Kb is sufficient to perform a gel electrophoresis depends on the size and concentration of the DNA samples you want to analyze. Generally, DNA Ladder 1 Kb is suitable for agarose or polyacrylamide gels with 0.7 to 1% concentration²³. However, if your DNA samples are very small (less than 100 bp) or very large (more than 15,000 bp), you may need a different type of DNA ladder, such as 100 bp, 100 bp plus, or 3 kb ladders¹. Also, if your DNA samples are very dilute or very concentrated, you may need to adjust the amount of DNA ladder you load on the gel to match the intensity of the bands².
                        - 
                        - Source: Conversation with Bing, 7/4/2023
                        - (1) DNA Ladders (1 kb, 1 kb plus, 100 bp, 100 bp plus) and Uses - Microbe Notes. [https://microbenotes.com/dna-ladders/.](https://microbenotes.com/dna-ladders/.)
                        - (2) E-Gel™ 1 Kb Plus DNA Ladder - Thermo Fisher Scientific. [https://www.thermofisher.com/order/catalog/product/10488090.](https://www.thermofisher.com/order/catalog/product/10488090.)
                        - (3) 1 kb DNA Ladder Gel Ladder for Electrophoresis Sigma - MilliporeSigma. [https://www.sigmaaldrich.com/US/en/product/sigma/d0428.](https://www.sigmaaldrich.com/US/en/product/sigma/d0428.)
                        - (4) 1 kb DNA Ladder for DNA electrophoresis - MilliporeSigma. [https://www.sigmaaldrich.com/DE/de/product/sigma/d0428.](https://www.sigmaaldrich.com/DE/de/product/sigma/d0428.)
                        - (5) Molecular-weight size marker - Wikipedia. [https://en.wikipedia.org/wiki/Molecular-weight_size_marker.](https://en.wikipedia.org/wiki/Molecular-weight_size_marker.) 
                        - 
                    - if I wanted to peform PCR to amplify the plasmid what primers would i need

                        - To perform PCR to amplify the plasmid, you need primers that can bind to the regions of the plasmid that you want to copy. Primers are short pieces of single-stranded DNA, usually around 20 nucleotides in length, that are complementary to the target sequence³. You can design your own primers using online tools or databases, or use existing primers that have been reported for your plasmid. For example, for pJE202 plasmid, one possible pair of primers is:
                        - 
                        - Forward primer: 5'-GAATTCATGTGGCATATCTCGAAGTAC-3'
                        - Reverse primer: 5'-GCGGCCGCTTACAGTTTCTGTTTGTTC-3'
                        - 
                        - These primers contain EcoRI and NotI restriction sites at their 5' ends, respectively, and hybridize to the start and stop codons of the lux operon genes in the plasmid⁴. You can use these primers to amplify the entire plasmid or a specific fragment of it. Alternatively, you can use other primers that target different regions of the plasmid, depending on your purpose.
                        - 
                        - Source: Conversation with Bing, 7/4/2023
                        - (1) Polymerase chain reaction (PCR) (article) | Khan Academy. [https://www.khanacademy.org/science/ap-biology/gene-expression-and-regulation/biotechnology/a/polymerase-chain-reaction-pcr.](https://www.khanacademy.org/science/ap-biology/gene-expression-and-regulation/biotechnology/a/polymerase-chain-reaction-pcr.)
                        - (2) pJE202 vector map and sequence - novoprolabs.com. [https://novoprolabs.com/vector/Vg4zdeoa.](https://novoprolabs.com/vector/Vg4zdeoa.)
                        - (3) Addgene: Plasmid Cloning by PCR (with Protocols). [https://www.addgene.org/protocols/pcr-cloning/.](https://www.addgene.org/protocols/pcr-cloning/.)
                        - (4) Addgene: Sequencing Primers. [https://www.addgene.org/mol-bio-reference/sequencing-primers/.](https://www.addgene.org/mol-bio-reference/sequencing-primers/.)
            - **culturing protocol for agrobacterium**
            - 
        - Plants being used for genetic modification 
        - What type of water is best for these types of biological experiments
        - **lb agar vs regular agar for growing bacteria**

        - 
        - **Plasmid purification and DNA gel extraction with glass syringe filters** 
        - 
        - **columns for plasmid extraction** 
    - 
- Grow plant seedlings
    - Get soil from a garden center or a nursery
    - Choose a suitable location with adequate sunlight and ventilation
    - Have on hand bigger containers to move the plants into once they mature
        - 3d print small vases
        - buy grow bags
    - Create a watering routine based on the plant species and the weather conditions
    - This should be relatively straightforward. You'll need a location with appropriate light (or grow lamps), potting soil, water, and your choice of seedlings. The time for this will depend on the growth rate of the plants, but you should plan for several weeks to months.
- Use seedlings to create protoplasts
- Use the protoplasts as a cheap source of cells for prototyping microfluidics designs
    - Design microfluidic channels and chambers using software tools or sketches
    - Test the microfluidic designs using a 3D printer for fabrication
    - Load the protoplasts into the microfluidic devices and observe their behavior and viability
    - This could be a good way to test your microfluidic devices before using more expensive or hard-to-obtain cells. You'll need to make sure your devices are compatible with the size and properties of the plant protoplasts.
- Source the equipment needed to do soft lithography, a more advanced form of microfluidics fabrication
- Finalize the designs for mouse cell culture, starting with mouse neurons
- Find/create a renewable supply of neuronal cells
- Start experimenting with
