{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# PaperQA - A Question Answering Dataset for Academic Papers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "!pip install paper-qa\n",
        "#!pip install git+https://github.com/blackadad/paper-scraper.git\n",
        "!pip install sentence-transformers\n",
        "#!pip install -U angle-emb\n",
        "api_key = \"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\"\n",
        "import os\n",
        "from re import T\n",
        "os.environ['OPENAI_API_KEY'] = api_key\n",
        "import nest_asyncio\n",
        "nest_asyncio.apply()\n",
        "!pip install langchain\n",
        "import langchain\n",
        "from langchain.cache import InMemoryCache\n",
        "langchain.llm_cache = InMemoryCache()\n",
        "model_name = \"ggrn/e5-small-v2\" # fast\n",
        "#model_name = \"WhereIsAI/UAE-Large-V1\" # slow\n",
        "model_kwargs = {'device': 'cpu'}\n",
        "from langchain.embeddings import HuggingFaceEmbeddings\n",
        "TOKENIZERS_PARALLELISM=True\n",
        "embeddings = HuggingFaceEmbeddings(model_name=model_name, model_kwargs=model_kwargs)\n",
        "!export DOI2PDF='https://sci-hub.ru/'\n",
        "os.environ['DOI2PDF'] = 'https://sci-hub.ru/'\n",
        "#os.environ[\"SEMANTIC_SCHOLAR_API_KEY\"]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [
        {
          "ename": "KeyboardInterrupt",
          "evalue": "",
          "output_type": "error",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
            "Cell \u001b[0;32mIn[2], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mre\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m M\n\u001b[0;32m----> 2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mpaperqa\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m Docs\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mos\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# Set the API key\u001b[39;00m\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/site-packages/paperqa/__init__.py:1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mdocs\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m Answer, Docs, PromptCollection, Doc, Text, Context\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mversion\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m __version__\n\u001b[1;32m      4\u001b[0m __all__ \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m      5\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mDocs\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m      6\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAnswer\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     11\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mContext\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     12\u001b[0m ]\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/site-packages/paperqa/docs.py:13\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mchat_models\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m ChatOpenAI\n\u001b[1;32m     12\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01membeddings\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mopenai\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m OpenAIEmbeddings\n\u001b[0;32m---> 13\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmemory\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m ConversationTokenBufferMemory\n\u001b[1;32m     14\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmemory\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mchat_memory\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m BaseChatMemory\n\u001b[1;32m     15\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mschema\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01membeddings\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m Embeddings\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/site-packages/langchain/memory/__init__.py:54\u001b[0m\n\u001b[1;32m     34\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmemory\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mchat_message_histories\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m (\n\u001b[1;32m     35\u001b[0m     AstraDBChatMessageHistory,\n\u001b[1;32m     36\u001b[0m     CassandraChatMessageHistory,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     51\u001b[0m     ZepChatMessageHistory,\n\u001b[1;32m     52\u001b[0m )\n\u001b[1;32m     53\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmemory\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mcombined\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m CombinedMemory\n\u001b[0;32m---> 54\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmemory\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mentity\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m (\n\u001b[1;32m     55\u001b[0m     ConversationEntityMemory,\n\u001b[1;32m     56\u001b[0m     InMemoryEntityStore,\n\u001b[1;32m     57\u001b[0m     RedisEntityStore,\n\u001b[1;32m     58\u001b[0m     SQLiteEntityStore,\n\u001b[1;32m     59\u001b[0m     UpstashRedisEntityStore,\n\u001b[1;32m     60\u001b[0m )\n\u001b[1;32m     61\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmemory\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mkg\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m ConversationKGMemory\n\u001b[1;32m     62\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmemory\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmotorhead_memory\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m MotorheadMemory\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/site-packages/langchain/memory/entity.py:11\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain_core\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mprompts\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m BasePromptTemplate\n\u001b[1;32m      9\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain_core\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpydantic_v1\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m BaseModel, Field\n\u001b[0;32m---> 11\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mchains\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mllm\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m LLMChain\n\u001b[1;32m     12\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmemory\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mchat_memory\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m BaseChatMemory\n\u001b[1;32m     13\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmemory\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mprompt\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m (\n\u001b[1;32m     14\u001b[0m     ENTITY_EXTRACTION_PROMPT,\n\u001b[1;32m     15\u001b[0m     ENTITY_SUMMARIZATION_PROMPT,\n\u001b[1;32m     16\u001b[0m )\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/site-packages/langchain/chains/__init__.py:21\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;124;03m\"\"\"**Chains** are easily reusable components linked together.\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \n\u001b[1;32m      3\u001b[0m \u001b[38;5;124;03mChains encode a sequence of calls to components like models, document retrievers,\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[38;5;124;03m    Chain --> <name>Chain  # Examples: LLMChain, MapReduceChain, RouterChain\u001b[39;00m\n\u001b[1;32m     18\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m     20\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mchains\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mapi\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m APIChain\n\u001b[0;32m---> 21\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mchains\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mapi\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mopenapi\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mchain\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m OpenAPIEndpointChain\n\u001b[1;32m     22\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mchains\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mcombine_documents\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m AnalyzeDocumentChain\n\u001b[1;32m     23\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mchains\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mcombine_documents\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmap_reduce\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m MapReduceDocumentsChain\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/site-packages/langchain/chains/api/openapi/chain.py:12\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mrequests\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m Response\n\u001b[1;32m     11\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mcallbacks\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmanager\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m CallbackManagerForChainRun, Callbacks\n\u001b[0;32m---> 12\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mchains\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mapi\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mopenapi\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mrequests_chain\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m APIRequesterChain\n\u001b[1;32m     13\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mchains\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mapi\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mopenapi\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mresponse_chain\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m APIResponderChain\n\u001b[1;32m     14\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mchains\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m Chain\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/site-packages/langchain/chains/api/openapi/requests_chain.py:40\u001b[0m\n\u001b[1;32m     35\u001b[0m     \u001b[38;5;129m@property\u001b[39m\n\u001b[1;32m     36\u001b[0m     \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_type\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28mstr\u001b[39m:\n\u001b[1;32m     37\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mapi_requester\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m---> 40\u001b[0m \u001b[38;5;28;43;01mclass\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;21;43;01mAPIRequesterChain\u001b[39;49;00m\u001b[43m(\u001b[49m\u001b[43mLLMChain\u001b[49m\u001b[43m)\u001b[49m\u001b[43m:\u001b[49m\n\u001b[1;32m     41\u001b[0m \u001b[38;5;250;43m    \u001b[39;49m\u001b[38;5;124;43;03m\"\"\"Get the request parser.\"\"\"\u001b[39;49;00m\n\u001b[1;32m     43\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;129;43m@classmethod\u001b[39;49m\n\u001b[1;32m     44\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43;01mdef\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;21;43mis_lc_serializable\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mcls\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m-\u001b[39;49m\u001b[38;5;241;43m>\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43mbool\u001b[39;49m\u001b[43m:\u001b[49m\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/site-packages/pydantic/main.py:138\u001b[0m, in \u001b[0;36mpydantic.main.ModelMetaclass.__new__\u001b[0;34m()\u001b[0m\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/site-packages/pydantic/utils.py:693\u001b[0m, in \u001b[0;36mpydantic.utils.smart_deepcopy\u001b[0;34m()\u001b[0m\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/copy.py:146\u001b[0m, in \u001b[0;36mdeepcopy\u001b[0;34m(x, memo, _nil)\u001b[0m\n\u001b[1;32m    144\u001b[0m copier \u001b[38;5;241m=\u001b[39m _deepcopy_dispatch\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;28mcls\u001b[39m)\n\u001b[1;32m    145\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m copier \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 146\u001b[0m     y \u001b[38;5;241m=\u001b[39m \u001b[43mcopier\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmemo\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    147\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    148\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28missubclass\u001b[39m(\u001b[38;5;28mcls\u001b[39m, \u001b[38;5;28mtype\u001b[39m):\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/copy.py:231\u001b[0m, in \u001b[0;36m_deepcopy_dict\u001b[0;34m(x, memo, deepcopy)\u001b[0m\n\u001b[1;32m    229\u001b[0m memo[\u001b[38;5;28mid\u001b[39m(x)] \u001b[38;5;241m=\u001b[39m y\n\u001b[1;32m    230\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m key, value \u001b[38;5;129;01min\u001b[39;00m x\u001b[38;5;241m.\u001b[39mitems():\n\u001b[0;32m--> 231\u001b[0m     y[deepcopy(key, memo)] \u001b[38;5;241m=\u001b[39m \u001b[43mdeepcopy\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmemo\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    232\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m y\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/copy.py:172\u001b[0m, in \u001b[0;36mdeepcopy\u001b[0;34m(x, memo, _nil)\u001b[0m\n\u001b[1;32m    170\u001b[0m                 y \u001b[38;5;241m=\u001b[39m x\n\u001b[1;32m    171\u001b[0m             \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 172\u001b[0m                 y \u001b[38;5;241m=\u001b[39m \u001b[43m_reconstruct\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmemo\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mrv\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    174\u001b[0m \u001b[38;5;66;03m# If is its own copy, don't memoize.\u001b[39;00m\n\u001b[1;32m    175\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m y \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m x:\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/copy.py:271\u001b[0m, in \u001b[0;36m_reconstruct\u001b[0;34m(x, memo, func, args, state, listiter, dictiter, deepcopy)\u001b[0m\n\u001b[1;32m    269\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m state \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    270\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m deep:\n\u001b[0;32m--> 271\u001b[0m         state \u001b[38;5;241m=\u001b[39m \u001b[43mdeepcopy\u001b[49m\u001b[43m(\u001b[49m\u001b[43mstate\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmemo\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    272\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(y, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m__setstate__\u001b[39m\u001b[38;5;124m'\u001b[39m):\n\u001b[1;32m    273\u001b[0m         y\u001b[38;5;241m.\u001b[39m__setstate__(state)\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/copy.py:146\u001b[0m, in \u001b[0;36mdeepcopy\u001b[0;34m(x, memo, _nil)\u001b[0m\n\u001b[1;32m    144\u001b[0m copier \u001b[38;5;241m=\u001b[39m _deepcopy_dispatch\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;28mcls\u001b[39m)\n\u001b[1;32m    145\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m copier \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 146\u001b[0m     y \u001b[38;5;241m=\u001b[39m \u001b[43mcopier\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmemo\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    147\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    148\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28missubclass\u001b[39m(\u001b[38;5;28mcls\u001b[39m, \u001b[38;5;28mtype\u001b[39m):\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/copy.py:211\u001b[0m, in \u001b[0;36m_deepcopy_tuple\u001b[0;34m(x, memo, deepcopy)\u001b[0m\n\u001b[1;32m    210\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_deepcopy_tuple\u001b[39m(x, memo, deepcopy\u001b[38;5;241m=\u001b[39mdeepcopy):\n\u001b[0;32m--> 211\u001b[0m     y \u001b[38;5;241m=\u001b[39m \u001b[43m[\u001b[49m\u001b[43mdeepcopy\u001b[49m\u001b[43m(\u001b[49m\u001b[43ma\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmemo\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43ma\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mx\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m    212\u001b[0m     \u001b[38;5;66;03m# We're not going to put the tuple in the memo, but it's still important we\u001b[39;00m\n\u001b[1;32m    213\u001b[0m     \u001b[38;5;66;03m# check for it, in case the tuple contains recursive mutable structures.\u001b[39;00m\n\u001b[1;32m    214\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/copy.py:211\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    210\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_deepcopy_tuple\u001b[39m(x, memo, deepcopy\u001b[38;5;241m=\u001b[39mdeepcopy):\n\u001b[0;32m--> 211\u001b[0m     y \u001b[38;5;241m=\u001b[39m [\u001b[43mdeepcopy\u001b[49m\u001b[43m(\u001b[49m\u001b[43ma\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmemo\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mfor\u001b[39;00m a \u001b[38;5;129;01min\u001b[39;00m x]\n\u001b[1;32m    212\u001b[0m     \u001b[38;5;66;03m# We're not going to put the tuple in the memo, but it's still important we\u001b[39;00m\n\u001b[1;32m    213\u001b[0m     \u001b[38;5;66;03m# check for it, in case the tuple contains recursive mutable structures.\u001b[39;00m\n\u001b[1;32m    214\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/copy.py:146\u001b[0m, in \u001b[0;36mdeepcopy\u001b[0;34m(x, memo, _nil)\u001b[0m\n\u001b[1;32m    144\u001b[0m copier \u001b[38;5;241m=\u001b[39m _deepcopy_dispatch\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;28mcls\u001b[39m)\n\u001b[1;32m    145\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m copier \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 146\u001b[0m     y \u001b[38;5;241m=\u001b[39m \u001b[43mcopier\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmemo\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    147\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    148\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28missubclass\u001b[39m(\u001b[38;5;28mcls\u001b[39m, \u001b[38;5;28mtype\u001b[39m):\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/copy.py:231\u001b[0m, in \u001b[0;36m_deepcopy_dict\u001b[0;34m(x, memo, deepcopy)\u001b[0m\n\u001b[1;32m    229\u001b[0m memo[\u001b[38;5;28mid\u001b[39m(x)] \u001b[38;5;241m=\u001b[39m y\n\u001b[1;32m    230\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m key, value \u001b[38;5;129;01min\u001b[39;00m x\u001b[38;5;241m.\u001b[39mitems():\n\u001b[0;32m--> 231\u001b[0m     y[deepcopy(key, memo)] \u001b[38;5;241m=\u001b[39m \u001b[43mdeepcopy\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmemo\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    232\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m y\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/copy.py:161\u001b[0m, in \u001b[0;36mdeepcopy\u001b[0;34m(x, memo, _nil)\u001b[0m\n\u001b[1;32m    159\u001b[0m reductor \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(x, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m__reduce_ex__\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[1;32m    160\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m reductor \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 161\u001b[0m     rv \u001b[38;5;241m=\u001b[39m reductor(\u001b[38;5;241m4\u001b[39m)\n\u001b[1;32m    162\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    163\u001b[0m     reductor \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(x, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m__reduce__\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m)\n",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
          ]
        }
      ],
      "source": [
        "from re import M\n",
        "from paperqa import Docs\n",
        "import os\n",
        "\n",
        "# Set the API key\n",
        "api_key = \"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\"\n",
        "\n",
        "# Optionally set the environment variable (if needed elsewhere)\n",
        "os.environ['OPENAI_API_KEY'] = api_key\n",
        "\n",
        "# Initialize Docs with the API key\n",
        "docs = Docs(llm='gpt-3.5-turbo', openai_api_key=api_key, memory=True, embeddings=embeddings)\n",
        "\n",
        "# load the papers from Mitochondria Papers folder\n",
        "\n",
        "mito_papers = os.listdir('Mitochondria Papers/')\n",
        "\n",
        "for paper in mito_papers:\n",
        "    docs.add(\"Mitochondria Papers/\"+paper, chunk_chars=2500)\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "\n",
        "# Query and print the answer\n",
        "answer = docs.query(\"What is the current understanding of the role of mitochondria in animal regeneration and aging, and what future research directions are being considered to harness these mechanisms for whole-body regeneration?\")\n",
        "print(answer.formatted_answer)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "import pickle\n",
        "\n",
        "# save\n",
        "with open(\"MitochondrialPapers.pkl\", \"wb\") as f:\n",
        "    pickle.dump(docs, f)\n",
        "\n",
        "# load\n",
        "with open(\"MitochondrialPapers.pkl\", \"rb\") as f:\n",
        "    docs = pickle.load(f)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "import os\n",
        "\n",
        "\n",
        "from paperqa import Docs\n",
        "\n",
        "try:\n",
        "    docs = Docs(llm='gpt-3.5-turbo', openai_api_key=api_key)\n",
        "    print(\"Initialization successful.\")\n",
        "except Exception as e:\n",
        "    print(f\"Initialization failed: {e}\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "import paperscraper\n",
        "# Set the API key\n",
        "api_key = \"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\"\n",
        "\n",
        "# Optionally set the environment variable (if needed elsewhere)\n",
        "os.environ['OPENAI_API_KEY'] = api_key\n",
        "\n",
        "# Initialize Docs with the API key\n",
        "#docs = Docs(llm='gpt-3.5-turbo', openai_api_key=api_key)\n",
        "import paperqa\n",
        "\n",
        "keyword_search = 'bispecific antibody manufacture'\n",
        "papers = paperscraper.search_papers(keyword_search)\n",
        "docs = paperqa.Docs(openai_api_key=api_key)\n",
        "for path,data in papers.items():\n",
        "    try:\n",
        "        #docs.add(path)\n",
        "        print(path, data['title'])\n",
        "    except ValueError as e:\n",
        "        # sometimes this happens if PDFs aren't downloaded or readable\n",
        "        print('Could not read', path, e)\n",
        "answer = docs.query(\"What manufacturing challenges are unique to bispecific antibodies?\")\n",
        "print(answer)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "ename": "KeyboardInterrupt",
          "evalue": "",
          "output_type": "error",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
            "Cell \u001b[0;32mIn[2], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mnest_asyncio\u001b[39;00m\n\u001b[1;32m      3\u001b[0m nest_asyncio\u001b[38;5;241m.\u001b[39mapply()\n\u001b[0;32m----> 4\u001b[0m papers \u001b[38;5;241m=\u001b[39m \u001b[43mpaperscraper\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msearch_papers\u001b[49m\u001b[43m(\u001b[49m\u001b[43mquery\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mbayesian model selection\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      5\u001b[0m \u001b[43m                                    \u001b[49m\u001b[43mlimit\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      6\u001b[0m \u001b[43m                                    \u001b[49m\u001b[43mpdir\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mdownloaded-papers\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/site-packages/paperscraper/lib.py:578\u001b[0m, in \u001b[0;36msearch_papers\u001b[0;34m(query, limit, pdir, semantic_scholar_api_key, _paths, _limit, _offset, logger, year, verbose, scraper, batch_size, search_type)\u001b[0m\n\u001b[1;32m    576\u001b[0m     loop \u001b[38;5;241m=\u001b[39m asyncio\u001b[38;5;241m.\u001b[39mnew_event_loop()\n\u001b[1;32m    577\u001b[0m     asyncio\u001b[38;5;241m.\u001b[39mset_event_loop(loop)\n\u001b[0;32m--> 578\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mloop\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_until_complete\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    579\u001b[0m \u001b[43m    \u001b[49m\u001b[43ma_search_papers\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    580\u001b[0m \u001b[43m        \u001b[49m\u001b[43mquery\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    581\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlimit\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlimit\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    582\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpdir\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpdir\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    583\u001b[0m \u001b[43m        \u001b[49m\u001b[43msemantic_scholar_api_key\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msemantic_scholar_api_key\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    584\u001b[0m \u001b[43m        \u001b[49m\u001b[43m_paths\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m_paths\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    585\u001b[0m \u001b[43m        \u001b[49m\u001b[43m_limit\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m_limit\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    586\u001b[0m \u001b[43m        \u001b[49m\u001b[43m_offset\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m_offset\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    587\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlogger\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlogger\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    588\u001b[0m \u001b[43m        \u001b[49m\u001b[43myear\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43myear\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    589\u001b[0m \u001b[43m        \u001b[49m\u001b[43mverbose\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mverbose\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    590\u001b[0m \u001b[43m        \u001b[49m\u001b[43mscraper\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mscraper\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    591\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbatch_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbatch_size\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    592\u001b[0m \u001b[43m        \u001b[49m\u001b[43msearch_type\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msearch_type\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    593\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    594\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/site-packages/nest_asyncio.py:93\u001b[0m, in \u001b[0;36m_patch_loop.<locals>.run_until_complete\u001b[0;34m(self, future)\u001b[0m\n\u001b[1;32m     91\u001b[0m     f\u001b[38;5;241m.\u001b[39m_log_destroy_pending \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m     92\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m f\u001b[38;5;241m.\u001b[39mdone():\n\u001b[0;32m---> 93\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_run_once\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     94\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_stopping:\n\u001b[1;32m     95\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/site-packages/nest_asyncio.py:129\u001b[0m, in \u001b[0;36m_patch_loop.<locals>._run_once\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    127\u001b[0m     handle \u001b[38;5;241m=\u001b[39m ready\u001b[38;5;241m.\u001b[39mpopleft()\n\u001b[1;32m    128\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m handle\u001b[38;5;241m.\u001b[39m_cancelled:\n\u001b[0;32m--> 129\u001b[0m         \u001b[43mhandle\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_run\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    130\u001b[0m handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/asyncio/events.py:80\u001b[0m, in \u001b[0;36mHandle._run\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     78\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_run\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m     79\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m---> 80\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_context\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_callback\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_args\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     81\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m (\u001b[38;5;167;01mSystemExit\u001b[39;00m, \u001b[38;5;167;01mKeyboardInterrupt\u001b[39;00m):\n\u001b[1;32m     82\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/asyncio/tasks.py:350\u001b[0m, in \u001b[0;36mTask.__wakeup\u001b[0;34m(self, future)\u001b[0m\n\u001b[1;32m    342\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m__step(exc)\n\u001b[1;32m    343\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    344\u001b[0m     \u001b[38;5;66;03m# Don't pass the value of `future.result()` explicitly,\u001b[39;00m\n\u001b[1;32m    345\u001b[0m     \u001b[38;5;66;03m# as `Future.__iter__` and `Future.__await__` don't need it.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    348\u001b[0m     \u001b[38;5;66;03m# instead of `__next__()`, which is slower for futures\u001b[39;00m\n\u001b[1;32m    349\u001b[0m     \u001b[38;5;66;03m# that return non-generator iterators from their `__iter__`.\u001b[39;00m\n\u001b[0;32m--> 350\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m__step\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    351\u001b[0m \u001b[38;5;28mself\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/site-packages/nest_asyncio.py:205\u001b[0m, in \u001b[0;36m_patch_task.<locals>.step\u001b[0;34m(task, exc)\u001b[0m\n\u001b[1;32m    203\u001b[0m curr_task \u001b[38;5;241m=\u001b[39m curr_tasks\u001b[38;5;241m.\u001b[39mget(task\u001b[38;5;241m.\u001b[39m_loop)\n\u001b[1;32m    204\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 205\u001b[0m     \u001b[43mstep_orig\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mexc\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    206\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    207\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m curr_task \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/asyncio/tasks.py:267\u001b[0m, in \u001b[0;36mTask.__step\u001b[0;34m(***failed resolving arguments***)\u001b[0m\n\u001b[1;32m    263\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    264\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m exc \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    265\u001b[0m         \u001b[38;5;66;03m# We use the `send` method directly, because coroutines\u001b[39;00m\n\u001b[1;32m    266\u001b[0m         \u001b[38;5;66;03m# don't have `__iter__` and `__next__` methods.\u001b[39;00m\n\u001b[0;32m--> 267\u001b[0m         result \u001b[38;5;241m=\u001b[39m coro\u001b[38;5;241m.\u001b[39msend(\u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[1;32m    268\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    269\u001b[0m         result \u001b[38;5;241m=\u001b[39m coro\u001b[38;5;241m.\u001b[39mthrow(exc)\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/site-packages/paperscraper/lib.py:495\u001b[0m, in \u001b[0;36ma_search_papers.<locals>.process_paper\u001b[0;34m(paper, i)\u001b[0m\n\u001b[1;32m    493\u001b[0m \u001b[38;5;28;01masync\u001b[39;00m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mprocess_paper\u001b[39m(paper, i):\n\u001b[1;32m    494\u001b[0m     path \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(pdir, \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mpaper[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpaperId\u001b[39m\u001b[38;5;124m\"\u001b[39m]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.pdf\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m--> 495\u001b[0m     success \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mawait\u001b[39;00m scraper\u001b[38;5;241m.\u001b[39mscrape(paper, path, i\u001b[38;5;241m=\u001b[39mi, logger\u001b[38;5;241m=\u001b[39mlogger)\n\u001b[1;32m    496\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m success:\n\u001b[1;32m    497\u001b[0m         bibtex \u001b[38;5;241m=\u001b[39m paper[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcitationStyles\u001b[39m\u001b[38;5;124m\"\u001b[39m][\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbibtex\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/site-packages/paperscraper/scraper.py:63\u001b[0m, in \u001b[0;36mScraper.scrape\u001b[0;34m(self, paper, path, i, logger)\u001b[0m\n\u001b[1;32m     61\u001b[0m scraper \u001b[38;5;241m=\u001b[39m scrapers[j]\n\u001b[1;32m     62\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m---> 63\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mawait\u001b[39;00m scraper\u001b[38;5;241m.\u001b[39mfunction(paper, path, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mscraper\u001b[38;5;241m.\u001b[39mkwargs)\n\u001b[1;32m     64\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m result \u001b[38;5;129;01mand\u001b[39;00m (\u001b[38;5;129;01mnot\u001b[39;00m scraper\u001b[38;5;241m.\u001b[39mcheck_pdf \u001b[38;5;129;01mor\u001b[39;00m check_pdf(path)):\n\u001b[1;32m     65\u001b[0m         scrape_result[scraper\u001b[38;5;241m.\u001b[39mname] \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msuccess\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/site-packages/paperscraper/lib.py:239\u001b[0m, in \u001b[0;36mopenaccess_scraper\u001b[0;34m(paper, path, session)\u001b[0m\n\u001b[1;32m    237\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m    238\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 239\u001b[0m     \u001b[38;5;28;01mawait\u001b[39;00m link_to_pdf(url, path, session)\n\u001b[1;32m    240\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
            "File \u001b[0;32m~/Documents/GitHub/ResearchAgentSwarm/.conda/lib/python3.11/site-packages/paperscraper/lib.py:103\u001b[0m, in \u001b[0;36mlink_to_pdf\u001b[0;34m(url, path, session)\u001b[0m\n\u001b[1;32m    101\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUnable to download \u001b[39m\u001b[38;5;132;01m{\u001b[39;00murl\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m, status code \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mr\u001b[38;5;241m.\u001b[39mstatus\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    102\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpdf\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m r\u001b[38;5;241m.\u001b[39mheaders[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mContent-Type\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n\u001b[0;32m--> 103\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43;01mwith\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mpath\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mwb\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mas\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mf\u001b[49m\u001b[43m:\u001b[49m\n\u001b[1;32m    104\u001b[0m \u001b[43m        \u001b[49m\u001b[43mf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwrite\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43;01mawait\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    105\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m\n",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
          ]
        },
        {
          "ename": "",
          "evalue": "",
          "output_type": "error",
          "traceback": [
            "\u001b[1;31mThe Kernel crashed while executing code in the the current cell or a previous cell. Please review the code in the cell(s) to identify a possible cause of the failure. Click <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. View Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
          ]
        }
      ],
      "source": [
        "import paperscraper\n",
        "import nest_asyncio\n",
        "nest_asyncio.apply()\n",
        "papers = paperscraper.search_papers(query='bayesian model selection',\n",
        "                                    limit=1,\n",
        "                                    pdir='downloaded-papers')\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "!pip install nougat-ocr\n",
        "#$ nougat path/to/file.pdf -o output_directory\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "!pip install transformers\n",
        "from transformers import AutoTokenizer, AutoModelForTokenClassification, pipeline\n",
        "#tokenizer = AutoTokenizer.from_pretrained(\"studio-ousia/luke-large\")\n",
        "#model = AutoModelForTokenClassification.from_pretrained(\"studio-ousia/luke-large\")\n",
        "tokenizer = AutoTokenizer.from_pretrained(\"dbmdz/electra-large-discriminator-finetuned-conll03-english\")\n",
        "model = AutoModelForTokenClassification.from_pretrained(\"dbmdz/electra-large-discriminator-finetuned-conll03-english\")\n",
        "\n",
        "nlp = pipeline(\"ner\", model=model, tokenizer=tokenizer, aggregation_strategy='simple')\n",
        "text = \"Recent studies have shown that multilingual pretrained language models can be effectively improved with cross-lingual alignment information from entities.\"\n",
        "ner_results = nlp(text)\n",
        "print(ner_results)\n",
        "# save to file txt\n",
        "with open('ner_results.txt', 'w') as f:\n",
        "    print(ner_results, file=f)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "!nougat '/Users/tomriddle1/Documents/GitHub/ResearchAgentSwarm/Mitochondria Papers/izawa2017.pdf' -o \"/Users/tomriddle1/Documents/GitHub/ResearchAgentSwarm/swarm_files\""
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Create Summary"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [],
      "source": [
        "import json\n",
        "import tempfile\n",
        "\n",
        "# Function to clean entities based on new lines and remove leading hyphens\n",
        "def clean_and_separate_entities(entities_list):\n",
        "    entities_str = '\\n'.join(entities_list)\n",
        "    cleaned_entities = []\n",
        "    dirty_entities = []\n",
        "\n",
        "    for line in entities_str.split('\\n'):\n",
        "        stripped_line = line.strip()\n",
        "        if stripped_line.startswith('-'):\n",
        "            # Remove the leading hyphen and any extra space after it\n",
        "            cleaned_entities.append(stripped_line.lstrip('-').strip())\n",
        "        else:\n",
        "            dirty_entities.append(stripped_line)\n",
        "\n",
        "    return cleaned_entities, dirty_entities\n",
        "def test_clean_and_separate_entities():\n",
        "    \n",
        "    # Define the summary JSON file path\n",
        "    SUMMARY_JSON = \"summaries.json\"\n",
        "\n",
        "    # Read the summaries.json file\n",
        "    with open(SUMMARY_JSON, \"r\") as file:\n",
        "        summaries_json = json.load(file)\n",
        "\n",
        "    # Extract the first entities entry\n",
        "    first_entities_list = summaries_json[0][\"entities\"][0]\n",
        "\n",
        "    # Clean the entities and separate the uncleaned ones\n",
        "    cleaned_entities, dirty_entities = clean_and_separate_entities(first_entities_list)\n",
        "\n",
        "    # Save the results to a temporary file\n",
        "    with tempfile.NamedTemporaryFile(mode='w', delete=False, suffix='.json') as temp_file:\n",
        "        json.dump({\n",
        "            \"cleaned_entities\": cleaned_entities,\n",
        "            \"dirty_entities\": dirty_entities\n",
        "        }, temp_file, indent=4)\n",
        "\n",
        "    print(\"Results saved in:\", temp_file.name)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [],
      "source": [
        "import re\n",
        "def extract_topics_with_justification(topic_text):\n",
        "    # Regular expression pattern for identifying topics with their justifications\n",
        "    topic_pattern = re.compile(r'(\\d+)\\.\\s+([^\\n]+)(\\n\\s+-[^\\n]+)*')\n",
        "    topics = topic_pattern.findall(topic_text)\n",
        "    \n",
        "    extracted_topics = []\n",
        "    for match in topics:\n",
        "        topic = match[1].strip()\n",
        "        justification = ' '.join(match[2].split('\\n')).strip()\n",
        "        # Remove \"Justification:\" if it starts with it\n",
        "        if justification.lower().startswith('- justification:'):\n",
        "            justification = justification[len('- justification:'):].strip()\n",
        "        # Remove the - if it starts with it\n",
        "        if justification.startswith('-'):\n",
        "            justification = justification[1:].strip()\n",
        "        extracted_topics.append({\"topic\": topic, \"justification\": justification})\n",
        "\n",
        "    return extracted_topics\n",
        "\n",
        "\n",
        "\n",
        "def test_extract_topics_with_justification():\n",
        "    # Adjusted topic text\n",
        "    topic_text_list = []\n",
        "    topic_text_list.append(\"**Topics Identified:**\\n\\n1. Importance of Mitochondria in Energy Production, Signaling, and Apoptosis\\n   - Mitochondria as the powerhouse of the cell\\n   - Role of mitochondria in energy production, signaling, and apoptosis\\n   - Significance of studying mitochondrial function and involvement in diseases\\n\\n2. Challenges with Traditional Methods of Mitochondrial Isolation\\n   - Limitations of traditional methods like differential centrifugation\\n   - Potential damage to mitochondrial double membrane and variable viability\\n\\n3. Innovative Techniques for Mitochondrial Isolation\\n   - Nitrogen cavitation for gentle disruption and release of intact mitochondria\\n   - Affinity purification using anti-TOM22 magnetic beads for efficient isolation\\n   - Filtration-based methods to reduce isolation time and improve viability\\n   - Differential isopycnic density gradient centrifugation for separation based on buoyant density\\n\\n4. Quality Control Measures for Validating Mitochondrial Isolation\\n   - Assessment of mitochondrial respiration, metabolic activity, protein import, and membrane fusion\\n   - High-resolution respirometry and bioluminescent measurements of ATP synthesis\\n\\n5. Importance of Continued Refinement and Standardization of Techniques\\n   - Advancing understanding of mitochondrial biology and implications in health and disease\\n   - Need for standardized protocols to facilitate comparisons and translation of research findings into clinical applications\\n\\n**Notes**: The summary provides a comprehensive overview of the importance of mitochondria, challenges with traditional methods of isolation, innovative techniques for isolation, quality control measures, and the need for continued refinement and standardization. The topics cover the main ideas and themes discussed in the summary, providing a clear and comprehensive analysis of the content.\") \n",
        "    topic_text_list.append(\"**Topic List:**\\n\\n1. Challenges in isolating intact mitochondria from plant cells\\n   - Cell walls, mitochondrial membranes, and large amounts of starting material\\n2. Comprehensive protocol for isolating intact mitochondria from plant cells\\n   - Grinding, filtering, centrifuging, and resuspending\\n3. Characterization and analysis of isolated mitochondria\\n   - Purity, integrity, and functionality assessment\\n   - Techniques: protein profiling, enzymatic activity assays, respiratory chain measurements, and oxygen consumption analysis\\n4. Storage of purified mitochondria\\n   - Long-term storage at -80C\\n5. Adaptation of isolation process for different tissue types and plant species\\n   - Consideration of phenolic compounds and metabolite profiles\\n6. Validation and controls for quality and functionality assurance\\n7. Downstream applications of isolated mitochondria\\n   - Protein and tRNA uptake experiments, enzyme activity assays, Western blot analyses, and mass spectrometry analyses\\n\\n**Notes:**\\n- The revised summary provides a comprehensive overview of the topic, covering various aspects of isolating intact mitochondria from plant cells.\\n- The topics are specific and non-repetitive, ensuring a clear and distinct representation of the core themes.\\n- The summary is focused on the technical process and considerations involved in isolating mitochondria, as well as the analysis and applications of the isolated mitochondria.\")\n",
        "    topic_text_list.append(\"**Topics Identified:**\\n\\n1. Importance of mitochondrial research in understanding cellular biology and addressing diseases related to mitochondrial dysfunction\\n    - Justification: The summary highlights the crucial role of mitochondrial research in understanding cellular biology and addressing diseases related to mitochondrial dysfunction.\\n\\n2. Significance of gentle and effective mitochondrial isolation techniques\\n    - Justification: The summary emphasizes the importance of gentle and effective isolation techniques for studying mitochondrial biology and developing mitochondrial-based therapies.\\n\\n3. Overview of macroscale mitochondrial isolation techniques\\n    - Justification: The summary discusses macroscale mitochondrial isolation techniques, such as manual homogenization and differential filtration-based isolation.\\n\\n4. Advancements in microscale and nanoscale mitochondrial isolation techniques\\n    - Justification: The summary mentions microscale and nanoscale techniques, including microfluidic techniques and nanoprobe-based technologies, for mitochondrial isolation.\\n\\n5. Breakthroughs in sub-cellular isolation techniques for mitochondria\\n    - Justification: The summary highlights breakthroughs in sub-cellular isolation techniques that enable the isolation of mitochondria from subcellular compartments with minimal disruption.\\n\\n6. Challenges in mitochondrial isolation techniques\\n    - Justification: The summary mentions challenges such as the presence of whole cell contaminants in mitochondrial isolates and the time sensitivity of isolated mitochondria.\\n\\n7. Emerging therapeutic approach: Autologous mitochondrial transplants\\n    - Justification: The summary discusses the development of autologous mitochondrial transplants as an emerging therapeutic approach.\\n\\n8. Contributions of the London Centre for Nanotechnology and the McCully laboratory\\n    - Justification: The summary mentions the significant contributions of the London Centre for Nanotechnology and the McCully laboratory in optimizing differential filtration-based mitochondrial isolation for use in cellular models.\\n\\n9. Role of Stem Cell Research & Therapy in advancing mitochondrial medicine\\n    - Justification: The summary highlights the role of Stem Cell Research & Therapy in providing in-depth overviews of advancements in mitochondrial research and facilitating the development of novel therapies for mitochondrial diseases.\")\n",
        "    topic_text_list.append(\"Topics:\\n1. Genetic modifications to enhance mitochondrial autonomy\\n   - Justification: The main focus of the report is exploring genetic modifications to enhance the autonomy of mitochondria from nuclear-encoded proteins and functions.\\n2. Role of mitochondria in cellular function\\n   - Justification: The report highlights the crucial role played by mitochondria in cellular function.\\n3. Coordination between mtDNA and nuclear DNA\\n   - Justification: The report discusses the coordination required between mtDNA and nuclear DNA, as most proteins are encoded by nuclear DNA.\\n4. Therapeutic strategies for mitochondrial diseases\\n   - Justification: The report mentions that enhancing mitochondrial autonomy could lead to new therapeutic strategies for mitochondrial diseases.\\n5. Research on genome engineering, programmable nucleases, and base editors\\n   - Justification: The report mentions that recent research in genome engineering, programmable nucleases, and base editors shows promise for treating hereditary mitochondrial diseases.\\n6. Challenges in genetic manipulation of mtDNA\\n   - Justification: The report discusses challenges such as mtDNA mutations, resistance to genetic manipulation, and limitations in mtDNA recombination.\\n7. Advancements in protein-only gene editing platforms\\n   - Justification: The report mentions advancements in protein-only gene editing platforms as potential solutions to the challenges in genetic manipulation of mtDNA.\\n8. Somatic mitochondrial DNA-replaced cells\\n   - Justification: The report mentions the generation of somatic mitochondrial DNA-replaced cells as a potential solution to the challenges in genetic manipulation of mtDNA.\\n9. Mitochondrial nucleoids and their role in maintaining genetic autonomy\\n   - Justification: The report highlights the concept of mitochondrial nucleoids and their role in maintaining genetic autonomy as a key area of study.\\n10. Mitochondrial epigenomics and gene expression regulation\\n    - Justification: The report emphasizes the importance of understanding mitochondrial epigenomics and gene expression regulation in different cellular contexts, including stress conditions, for identifying genetic modifications that could enhance mitochondrial autonomy.\")\n",
        "    for topic_text in topic_text_list:\n",
        "        extracted_topics = extract_topics_with_justification(topic_text)\n",
        "        print(f'Extracted topics: {extracted_topics}')\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "141.27s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "'\\nfor url in urls:\\n    try:\\n        pdf =  pdfx.PDFx(url)\\n        metadata = pdf.get_metadata()\\n        print(f\\'Metadata: {metadata}\\')\\n        references_list = pdf.get_references()\\n        print(f\\'References: {references_list}\\')\\n        references_dict = pdf.get_references_as_dict()\\n        print(f\\'References dict: {references_dict}\\')\\n        papers = paperscraper.link_to_pdf(url, pdir=\\'downloaded-papers\\')\\n        print(f\\'Papers: {papers}\\')\\n    except:\\n        print(\"Error in extracting references\")\\n        continue\\n#pdf.download_pdfs(\"target-directory\")\\n\\n'"
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "import re\n",
        "#!pip install pdfx\n",
        "import pdfx\n",
        "#!pip install paperscraper\n",
        "#import paperscraper\n",
        "import nest_asyncio\n",
        "nest_asyncio.apply()\n",
        "import os\n",
        "!export DOI2PDF='https://sci-hub.ru/'\n",
        "os.environ['DOI2PDF'] = 'https://sci-hub.ru/'\n",
        "def extract_urls(reference_text):\n",
        "    # Regular expression pattern for identifying URLs\n",
        "    url_pattern = re.compile(r'https?://[^\\s,]+')\n",
        "    urls = url_pattern.findall(reference_text)\n",
        "    return urls\n",
        "\n",
        "\n",
        "def test_extract_urls():\n",
        "    # Define the reference text\n",
        "    reference_text = \"\"\"\\n\\nAmerican Institute of Physics. (2023). The powerhouse of the future: Artificial cells. Phys.org. Retrieved from https://phys.org/news/2023-03-powerhouse-future-artificial-cells.html\\n\\nNational Institutes of Health. (2023). Artificial mitochondria transfer (AMT) and transplant. PMC. Retrieved from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5511681/\\n\\nNature. (2023). Spatiotemporal simulations of mitochondrial dynamics. Nature.com. Retrieved from https://www.nature.com/articles/s41598-019-54159-1\\n\\nSogang University & Harbin Institute of Technology. (2023). Artificial organelles for sustainable chemical energy conversion and production: Artificial mitochondria and chloroplasts. Biophysics Reviews. Retrieved from https://publishing.aip.org/publications/latest-content/the-powerhouse-of-the-future-artificial-cells/\"\"\"\n",
        "\n",
        "    urls = extract_urls(reference_text)\n",
        "    print(f'Extracted URLs: {urls}')\n",
        "\n",
        "#pdf = pdfx.PDFx(\"filename-or-url.pdf\")\n",
        "#urls = ['/Users/tomriddle1/Documents/GitHub/ResearchAgentSwarm/2308.00352.pdf']\n",
        "#!pdfx -v /Users/tomriddle1/Documents/GitHub/ResearchAgentSwarm/Literature_Review/gpt_researcher_outputs/0e1ea044ac594b668db126787ad7f47d.pdf -o /Users/tomriddle1/Documents/GitHub/ResearchAgentSwarm/swarm_files/test.txt\n",
        "\n",
        "\"\"\"\n",
        "for url in urls:\n",
        "    try:\n",
        "        pdf =  pdfx.PDFx(url)\n",
        "        metadata = pdf.get_metadata()\n",
        "        print(f'Metadata: {metadata}')\n",
        "        references_list = pdf.get_references()\n",
        "        print(f'References: {references_list}')\n",
        "        references_dict = pdf.get_references_as_dict()\n",
        "        print(f'References dict: {references_dict}')\n",
        "        papers = paperscraper.link_to_pdf(url, pdir='downloaded-papers')\n",
        "        print(f'Papers: {papers}')\n",
        "    except:\n",
        "        print(\"Error in extracting references\")\n",
        "        continue\n",
        "#pdf.download_pdfs(\"target-directory\")\n",
        "\n",
        "\"\"\"\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Questions: [('Content-Based Question', 'How do genetic modifications contribute to increasing mitochondrial autonomy from nuclear-encoded proteins and functions?'), ('Analytical Question', 'What are the key tools and methods used in modifying the mitochondrial genome to study the interplay between nuclear and mitochondrial genomes?'), ('Creative/Scenario-Based Question', 'Imagine a future where mitochondrial autonomy from nuclear-encoded proteins and functions is fully achieved. How might this impact our understanding of cellular functions and the development of new treatments for mitochondrial diseases?'), ('Contextual/Relational Question', 'How does the research on modifying the mitochondrial genome relate to other areas of genetic engineering and its potential for future advancements in the field?'), ('User-Interactive Question', 'What are your thoughts on the ethical considerations surrounding genetic modifications in mitochondrial genome engineering? How do you think society should approach this research?')]\n",
            "Extracted hypothetical questions: [{'question_type': 'Content-Based Question', 'question': 'How do genetic modifications contribute to increasing mitochondrial autonomy from nuclear-encoded proteins and functions?'}, {'question_type': 'Analytical Question', 'question': 'What are the key tools and methods used in modifying the mitochondrial genome to study the interplay between nuclear and mitochondrial genomes?'}, {'question_type': 'Creative/Scenario-Based Question', 'question': 'Imagine a future where mitochondrial autonomy from nuclear-encoded proteins and functions is fully achieved. How might this impact our understanding of cellular functions and the development of new treatments for mitochondrial diseases?'}, {'question_type': 'Contextual/Relational Question', 'question': 'How does the research on modifying the mitochondrial genome relate to other areas of genetic engineering and its potential for future advancements in the field?'}, {'question_type': 'User-Interactive Question', 'question': 'What are your thoughts on the ethical considerations surrounding genetic modifications in mitochondrial genome engineering? How do you think society should approach this research?'}]\n",
            "Questions: [('Analytical Question', 'How do theoretical models help in understanding mitochondrial ATP production?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where mitochondrial ATP production could be replicated outside the cellular environment. How could this impact medical research and treatments?'), ('Contextual/Relational Question', 'How does the research on mitochondrial ATP production relate to the broader field of cellular bioenergetics?'), ('User-Interactive Question', 'How would you approach studying the replication of mitochondrial ATP production outside the cellular environment?')]\n",
            "Extracted hypothetical questions: [{'question_type': 'Analytical Question', 'question': 'How do theoretical models help in understanding mitochondrial ATP production?'}, {'question_type': 'Creative/Scenario-Based Question', 'question': 'Imagine a scenario where mitochondrial ATP production could be replicated outside the cellular environment. How could this impact medical research and treatments?'}, {'question_type': 'Contextual/Relational Question', 'question': 'How does the research on mitochondrial ATP production relate to the broader field of cellular bioenergetics?'}, {'question_type': 'User-Interactive Question', 'question': 'How would you approach studying the replication of mitochondrial ATP production outside the cellular environment?'}]\n"
          ]
        }
      ],
      "source": [
        "import re\n",
        "\n",
        "def extract_hypothetical_questions(hypothetical_questions_text):\n",
        "    # Regular expression pattern for identifying hypothetical questions\n",
        "    question_pattern = re.compile(r'\\d+\\.\\s+([A-Za-z\\/-]+ Question):\\n\\s+-\\s+([^\\n]+)')\n",
        "    questions = question_pattern.findall(hypothetical_questions_text)\n",
        "    print(f'Questions: {questions}')\n",
        "    if len(questions) == 0:\n",
        "        return hypothetical_questions_text\n",
        "    return [{\"question_type\": question_type, \"question\": question} for question_type, question in questions]\n",
        "def test_extract_hypothetical_questions():\n",
        "    # Example hypothetical questions text\n",
        "    hypothetical_questions_text_1 = \"1. Content-Based Question:\\n   - How do genetic modifications contribute to increasing mitochondrial autonomy from nuclear-encoded proteins and functions?\\n\\n2. Analytical Question:\\n   - What are the key tools and methods used in modifying the mitochondrial genome to study the interplay between nuclear and mitochondrial genomes?\\n\\n3. Creative/Scenario-Based Question:\\n   - Imagine a future where mitochondrial autonomy from nuclear-encoded proteins and functions is fully achieved. How might this impact our understanding of cellular functions and the development of new treatments for mitochondrial diseases?\\n\\n4. Contextual/Relational Question:\\n   - How does the research on modifying the mitochondrial genome relate to other areas of genetic engineering and its potential for future advancements in the field?\\n\\n5. User-Interactive Question:\\n   - What are your thoughts on the ethical considerations surrounding genetic modifications in mitochondrial genome engineering? How do you think society should approach this research?\"\n",
        "    hypothetical_questions_text_2 = \"1. Content-Based Question: \\n   - What does this report investigate regarding mitochondrial ATP production?\\n   - How does this report contribute to our understanding of mitochondrial function?\\n   - What are the key findings regarding the replication of mitochondrial ATP production outside the cellular environment?\\n\\n2. Analytical Question:\\n   - How do theoretical models help in understanding mitochondrial ATP production?\\n   - What experimental evidence supports the concept of artificial organelles for ATP synthesis?\\n   - What are the implications of studying mitochondrial dynamics and stress responses for ex vivo methods of ATP synthesis?\\n\\n3. Creative/Scenario-Based Question:\\n   - Imagine a scenario where mitochondrial ATP production could be replicated outside the cellular environment. How could this impact medical research and treatments?\\n   - If artificial organelles capable of ATP synthesis were successfully developed, what potential applications could they have in various industries?\\n   - How might the understanding of mitochondrial dynamics and stress responses lead to the development of innovative approaches for ATP synthesis?\\n\\n4. Contextual/Relational Question:\\n   - How does the research on mitochondrial ATP production relate to the broader field of cellular bioenergetics?\\n   - In what ways does the replication of mitochondrial ATP production outside cells build upon previous studies in the field?\\n   - How do the findings in this report align with or challenge existing theories and models of mitochondrial function?\\n\\n5. User-Interactive Question:\\n   - How would you approach studying the replication of mitochondrial ATP production outside the cellular environment?\\n   - Can you think of any potential limitations or ethical considerations in developing artificial organelles for ATP synthesis?\\n   - What questions or areas of research would you like to see explored further in the study of mitochondrial dynamics and stress responses?\"\n",
        "    hypothetical_questions = []\n",
        "    hypothetical_questions.append(hypothetical_questions_text_1)\n",
        "    hypothetical_questions.append(hypothetical_questions_text_2)\n",
        "    for hypothetical_questions_text in hypothetical_questions:\n",
        "        extracted_hypothetical_questions = extract_hypothetical_questions(hypothetical_questions_text)\n",
        "        print(f'Extracted hypothetical questions: {extracted_hypothetical_questions}')\n",
        "\n",
        "test_extract_hypothetical_questions()\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[{'subject': 'mitochondria',\n",
              "  'relationship': 'responsible for',\n",
              "  'target': 'energy production'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated from',\n",
              "  'target': 'plant cells'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated for',\n",
              "  'target': 'studies involving mitochondrial DNA, protein profiling, and enzymatic activity assays'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated using',\n",
              "  'target': 'continuous colloidal density gradients'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'improved methods'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'slight modifications'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'traditional plant protoplast isolation'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'mammalian mitochondria extraction protocols'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'adjustments in isolation medium compositions'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'reduced need for heavy labor, expensive equipment, and large amounts of starting material'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'used for',\n",
              "  'target': 'respiratory chain measurements, western blot analyses, and mass spectrometry'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'used for',\n",
              "  'target': 'protein and tRNA uptake experiments, enzyme activity assays, and western blot analyses'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'used for',\n",
              "  'target': 'targeted multiple reaction monitoring or quantification by dimethyl or other isotope labels'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed for',\n",
              "  'target': 'purity and integrity'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed using',\n",
              "  'target': 'proteinase digestion assays, electron microscopy, mitochondrial membrane potential measurement, and electron transport chain activity measurement'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed to',\n",
              "  'target': 'confirm the intactness and functional capacity'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed to',\n",
              "  'target': 'evaluate the mitochondrial purity'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed to',\n",
              "  'target': 'evaluate the mitochondrial integrity'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed to',\n",
              "  'target': 'measure the mitochondrial membrane potential'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed to',\n",
              "  'target': 'measure the electron transport chain activity'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed at',\n",
              "  'target': 'the DNA and protein levels'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed using',\n",
              "  'target': 'electron microscopy'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed using',\n",
              "  'target': 'proteinase digestion assays'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed using',\n",
              "  'target': 'mitochondrial membrane potential measurement'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed using',\n",
              "  'target': 'electron transport chain activity measurement'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated from',\n",
              "  'target': 'Arabidopsis thaliana'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated using',\n",
              "  'target': 'continuous colloidal density gradients'},\n",
              " {'subject': 'mitochondria', 'relationship': 'isolated at', 'target': '4 C'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated for',\n",
              "  'target': 'studies involving mitochondrial DNA, protein profiling, and enzymatic activity assays'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'tailored isolation protocol'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'minimized damage to ensure the integrity'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'reduced contamination from other organelles'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'improved methods'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'mammalian mitochondria extraction protocols'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'adjustments in isolation medium compositions'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'reduced need for heavy labor, expensive equipment, and large amounts of starting material'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'minimal contamination from other organelles'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'improved methods'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'slight modifications'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'traditional plant protoplast isolation'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'mammalian mitochondria extraction protocols'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'adjustments in isolation medium compositions'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'reduced need for heavy labor, expensive equipment, and large amounts of starting material'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'minimal contamination from other organelles'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated from',\n",
              "  'target': 'Arabidopsis thaliana'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated using',\n",
              "  'target': 'continuous colloidal density gradients'},\n",
              " {'subject': 'mitochondria', 'relationship': 'isolated at', 'target': '4 C'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'used for',\n",
              "  'target': 'protein and tRNA uptake experiments'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'used for',\n",
              "  'target': 'enzyme activity assays'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'used for',\n",
              "  'target': 'western blot analyses'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'used for',\n",
              "  'target': 'mass spectrometry analyses'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'used for',\n",
              "  'target': 'targeted multiple reaction monitoring or quantification by dimethyl or other isotope labels'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed for',\n",
              "  'target': 'purity and integrity'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed using',\n",
              "  'target': 'proteinase digestion assays'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed using',\n",
              "  'target': 'electron microscopy'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'responsible for',\n",
              "  'target': 'energy production'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated from',\n",
              "  'target': 'plant cells'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated for',\n",
              "  'target': 'studies involving mitochondrial DNA, protein profiling, and enzymatic activity assays'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated using',\n",
              "  'target': 'continuous colloidal density gradients'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'improved methods'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'slight modifications'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'traditional plant protoplast isolation'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'mammalian mitochondria extraction protocols'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'adjustments in isolation medium compositions'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated with',\n",
              "  'target': 'reduced need for heavy labor, expensive equipment, and large amounts of starting material'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'used for',\n",
              "  'target': 'respiratory chain measurements, western blot analyses, and mass spectrometry'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'used for',\n",
              "  'target': 'protein and tRNA uptake experiments, enzyme activity assays, and western blot analyses'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'used for',\n",
              "  'target': 'targeted multiple reaction monitoring or quantification by dimethyl or other isotope labels'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed for',\n",
              "  'target': 'purity and integrity'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed using',\n",
              "  'target': 'proteinase digestion assays, electron microscopy, mitochondrial membrane potential measurement, and electron transport chain activity measurement'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed to',\n",
              "  'target': 'confirm the intactness and functional capacity'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed to',\n",
              "  'target': 'evaluate the mitochondrial purity'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed to',\n",
              "  'target': 'evaluate the mitochondrial integrity'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed to',\n",
              "  'target': 'measure the mitochondrial membrane potential'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed to',\n",
              "  'target': 'measure the electron transport chain activity'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed at',\n",
              "  'target': 'the DNA and protein levels'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed using',\n",
              "  'target': 'electron microscopy'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed using',\n",
              "  'target': 'proteinase digestion assays'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed using',\n",
              "  'target': 'mitochondrial membrane potential measurement'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed using',\n",
              "  'target': 'electron transport chain activity measurement'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated from',\n",
              "  'target': 'Arabidopsis thaliana'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated using',\n",
              "  'target': 'continuous colloidal density gradients'},\n",
              " {'subject': 'mitochondria', 'relationship': 'isolated at', 'target': '4 C'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'used for',\n",
              "  'target': 'protein and tRNA uptake experiments'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'used for',\n",
              "  'target': 'enzyme activity assays'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'used for',\n",
              "  'target': 'western blot analyses'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'used for',\n",
              "  'target': 'mass spectrometry analyses'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'used for',\n",
              "  'target': 'targeted multiple reaction monitoring or quantification by dimethyl or other isotope labels'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed for',\n",
              "  'target': 'purity and integrity'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed using',\n",
              "  'target': 'proteinase digestion assays'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed using',\n",
              "  'target': 'electron microscopy'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed using',\n",
              "  'target': 'mitochondrial membrane potential measurement'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'assessed using',\n",
              "  'target': 'electron transport chain activity measurement'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated from',\n",
              "  'target': 'Arabidopsis thaliana'},\n",
              " {'subject': 'mitochondria',\n",
              "  'relationship': 'isolated using',\n",
              "  'target': 'continuous colloidal density gradients'},\n",
              " {'subject': 'mitochondria', 'relationship': 'isolated at', 'target': '4 C'}]"
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "def clean_entity_relationships(entity_relationships_text):\n",
        "    # Regular expression pattern for identifying entity relationships\n",
        "    entity_pattern = re.compile(r'\\d+\\.\\s+\\((.+?),\\s+(.+?),\\s+(.+?)\\)')\n",
        "    entity_relationships = entity_pattern.findall(entity_relationships_text)\n",
        "    return [{\"subject\": relationship[0], \"relationship\": relationship[1], \"target\": relationship[2]} for relationship in entity_relationships]\n",
        "\n",
        "# Example entity relationships text\n",
        "entity_relationships_text =  \"Entity Relationships:\\n\\n1. (mitochondria, responsible for, energy production)\\n2. (mitochondria, isolated from, plant cells)\\n3. (mitochondria, isolated for, studies involving mitochondrial DNA, protein profiling, and enzymatic activity assays)\\n4. (mitochondria, isolated using, continuous colloidal density gradients)\\n5. (mitochondria, isolated with, improved methods)\\n6. (mitochondria, isolated with, slight modifications)\\n7. (mitochondria, isolated with, traditional plant protoplast isolation)\\n8. (mitochondria, isolated with, mammalian mitochondria extraction protocols)\\n9. (mitochondria, isolated with, adjustments in isolation medium compositions)\\n10. (mitochondria, isolated with, reduced need for heavy labor, expensive equipment, and large amounts of starting material)\\n11. (mitochondria, used for, respiratory chain measurements, western blot analyses, and mass spectrometry)\\n12. (mitochondria, used for, protein and tRNA uptake experiments, enzyme activity assays, and western blot analyses)\\n13. (mitochondria, used for, targeted multiple reaction monitoring or quantification by dimethyl or other isotope labels)\\n14. (mitochondria, assessed for, purity and integrity)\\n15. (mitochondria, assessed using, proteinase digestion assays, electron microscopy, mitochondrial membrane potential measurement, and electron transport chain activity measurement)\\n16. (mitochondria, assessed to, confirm the intactness and functional capacity)\\n17. (mitochondria, assessed to, evaluate the mitochondrial purity)\\n18. (mitochondria, assessed to, evaluate the mitochondrial integrity)\\n19. (mitochondria, assessed to, measure the mitochondrial membrane potential)\\n20. (mitochondria, assessed to, measure the electron transport chain activity)\\n21. (mitochondria, assessed at, the DNA and protein levels)\\n22. (mitochondria, assessed using, electron microscopy)\\n23. (mitochondria, assessed using, proteinase digestion assays)\\n24. (mitochondria, assessed using, mitochondrial membrane potential measurement)\\n25. (mitochondria, assessed using, electron transport chain activity measurement)\\n26. (mitochondria, isolated from, Arabidopsis thaliana)\\n27. (mitochondria, isolated using, continuous colloidal density gradients)\\n28. (mitochondria, isolated at, 4 C)\\n29. (mitochondria, isolated for, studies involving mitochondrial DNA, protein profiling, and enzymatic activity assays)\\n30. (mitochondria, isolated with, tailored isolation protocol)\\n31. (mitochondria, isolated with, minimized damage to ensure the integrity)\\n32. (mitochondria, isolated with, reduced contamination from other organelles)\\n33. (mitochondria, isolated with, improved methods)\\n34. (mitochondria, isolated with, mammalian mitochondria extraction protocols)\\n35. (mitochondria, isolated with, adjustments in isolation medium compositions)\\n36. (mitochondria, isolated with, reduced need for heavy labor, expensive equipment, and large amounts of starting material)\\n37. (mitochondria, isolated with, minimal contamination from other organelles)\\n38. (mitochondria, isolated with, improved methods)\\n39. (mitochondria, isolated with, slight modifications)\\n40. (mitochondria, isolated with, traditional plant protoplast isolation)\\n41. (mitochondria, isolated with, mammalian mitochondria extraction protocols)\\n42. (mitochondria, isolated with, adjustments in isolation medium compositions)\\n43. (mitochondria, isolated with, reduced need for heavy labor, expensive equipment, and large amounts of starting material)\\n44. (mitochondria, isolated with, minimal contamination from other organelles)\\n45. (mitochondria, isolated from, Arabidopsis thaliana)\\n46. (mitochondria, isolated using, continuous colloidal density gradients)\\n47. (mitochondria, isolated at, 4 C)\\n48. (mitochondria, used for, protein and tRNA uptake experiments)\\n49. (mitochondria, used for, enzyme activity assays)\\n50. (mitochondria, used for, western blot analyses)\\n51. (mitochondria, used for, mass spectrometry analyses)\\n52. (mitochondria, used for, targeted multiple reaction monitoring or quantification by dimethyl or other isotope labels)\\n53. (mitochondria, assessed for, purity and integrity)\\n54. (mitochondria, assessed using, proteinase digestion assays)\\n55. (mitochondria, assessed using, electron microscopy)\\n56. (mitochondria, assessedThe article discusses the protocol for isolating mitochondria from plant cells. Mitochondria are double-membraned organelles responsible for energy production in eukaryotic cells. The isolation of mitochondria is crucial for various studies involving mitochondrial DNA, protein profiling, and enzymatic activity assays.\\n\\nThe isolation process is challenging due to the presence of cell walls, vacuoles, and secondary metabolites in plant cells. The protocol must be tailored to minimize damage to the mitochondria and ensure their integrity. Specificity in isolation protocols is required as different plant species and tissue types have varying phenolic compounds and metabolite profiles. Earlier methods led to contamination with nuclei and chloroplasts, but recent advancements have improved isolation methods, reducing the need for heavy labor, expensive equipment, and large amounts of starting material.\\n\\nThe protocol for isolating intact mitochondria involves several steps. First, the preparation of grinding medium, wash buffer, and gradient solutions is necessary. The plant material is then homogenized in the grinding medium to release the mitochondria, which are then filtered and centrifuged to pellet the mitochondria. The mitochondrial pellet is resuspended in the wash buffer. Oxygen consumption measurements are crucial for determining the intactness and functional capacity of the isolated mitochondria. Evaluation of mitochondrial purity and integrity can be done through proteinase digestion assays, electron microscopy, and checks of mitochondrial membrane potential and electron transport chain activity.\\n\\nOnce purified, the isolated mitochondria can be used for various studies, including protein and tRNA uptake experiments, enzyme activity assays, and western blot analyses. For mass spectrometry analyses, targeted multiple reaction monitoring (MRM) or quantification by dimethyl or other isotope labels can be employed.\\n\\nIn conclusion, the isolation of mitochondria from plant cells is a delicate process that requires careful consideration of the specific requirements of the plant species and tissue type. Recent advancements have made the process more effective and accessible for a range of tissue types and species, allowing for a broader application of mitochondrial studies across different plant species.\\n\\nReferences:\\n- Plant Methods. (2015). https://plantmethods.biomedcentral.com/articles/10.1186/s13007-015-0099-x\\n- NCBI. (2018). https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5908444/\\n- NCBI. (2018). https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7640673/\\n- NCBI. (2018). https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4687074/Extraction and Categorization:\\n\\n1. (mitochondria, responsible for, energy production)\\n2. (mitochondria, isolated from, plant cells)\\n3. (mitochondria, isolated for, studies involving mitochondrial DNA, protein profiling, and enzymatic activity assays)\\n4. (mitochondria, isolated using, continuous colloidal density gradients)\\n5. (mitochondria, isolated with, improved methods)\\n6. (mitochondria, isolated with, slight modifications)\\n7. (mitochondria, isolated with, traditional plant protoplast isolation)\\n8. (mitochondria, isolated with, mammalian mitochondria extraction protocols)\\n9. (mitochondria, isolated with, adjustments in isolation medium compositions)\\n10. (mitochondria, isolated with, reduced need for heavy labor, expensive equipment, and large amounts of starting material)\\n11. (mitochondria, used for, respiratory chain measurements, western blot analyses, and mass spectrometry)\\n12. (mitochondria, used for, protein and tRNA uptake experiments, enzyme activity assays, and western blot analyses)\\n13. (mitochondria, used for, targeted multiple reaction monitoring or quantification by dimethyl or other isotope labels)\\n14. (mitochondria, assessed for, purity and integrity)\\n15. (mitochondria, assessed using, proteinase digestion assays, electron microscopy, mitochondrial membrane potential measurement, and electron transport chain activity measurement)\\n16. (mitochondria, assessed to, confirm the intactness and functional capacity)\\n17. (mitochondria, assessed to, evaluate the mitochondrial purity)\\n18. (mitochondria, assessed to, evaluate the mitochondrial integrity)\\n19. (mitochondria, assessed to, measure the mitochondrial membrane potential)\\n20. (mitochondria, assessed to, measure the electron transport chain activity)\\n21. (mitochondria, assessed at, the DNA and protein levels)\\n22. (mitochondria, assessed using, electron microscopy)\\n23. (mitochondria, assessed using, proteinase digestion assays)\\n24. (mitochondria, assessed using, mitochondrial membrane potential measurement)\\n25. (mitochondria, assessed using, electron transport chain activity measurement)\\n26. (mitochondria, isolated from, Arabidopsis thaliana)\\n27. (mitochondria, isolated using, continuous colloidal density gradients)\\n28. (mitochondria, isolated at, 4 C)\\n29. (mitochondria, used for, protein and tRNA uptake experiments)\\n30. (mitochondria, used for, enzyme activity assays)\\n31. (mitochondria, used for, western blot analyses)\\n32. (mitochondria, used for, mass spectrometry analyses)\\n33. (mitochondria, used for, targeted multiple reaction monitoring or quantification by dimethyl or other isotope labels)\\n34. (mitochondria, assessed for, purity and integrity)\\n35. (mitochondria, assessed using, proteinase digestion assays)\\n36. (mitochondria, assessed using, electron microscopy)\\n37. (mitochondria, assessed using, mitochondrial membrane potential measurement)\\n38. (mitochondria, assessed using, electron transport chain activity measurement)\\n39. (mitochondria, isolated from, Arabidopsis thaliana)\\n40. (mitochondria, isolated using, continuous colloidal density gradients)\\n41. (mitochondria, isolated at, 4 C)\"\n",
        "\n",
        "# Clean the entity relationships\n",
        "cleaned_entity_relationships = clean_entity_relationships(entity_relationships_text)\n",
        "\n",
        "# Output the cleaned entity relationships\n",
        "cleaned_entity_relationships\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "148.82s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: pydantic in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (2.5.2)\n",
            "Requirement already satisfied: annotated-types>=0.4.0 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from pydantic) (0.6.0)\n",
            "Requirement already satisfied: pydantic-core==2.14.5 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from pydantic) (2.14.5)\n",
            "Requirement already satisfied: typing-extensions>=4.6.1 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from pydantic) (4.8.0)\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "154.84s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: instructor in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (0.4.5)\n",
            "Requirement already satisfied: aiohttp<4.0.0,>=3.9.1 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from instructor) (3.9.1)\n",
            "Requirement already satisfied: docstring-parser<0.16,>=0.15 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from instructor) (0.15)\n",
            "Requirement already satisfied: openai<2.0.0,>=1.1.0 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from instructor) (1.7.1)\n",
            "Requirement already satisfied: pydantic<3.0.0,>=2.0.2 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from instructor) (2.5.2)\n",
            "Requirement already satisfied: rich<14.0.0,>=13.7.0 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from instructor) (13.7.0)\n",
            "Requirement already satisfied: typer<0.10.0,>=0.9.0 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from instructor) (0.9.0)\n",
            "Requirement already satisfied: attrs>=17.3.0 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.9.1->instructor) (23.1.0)\n",
            "Requirement already satisfied: multidict<7.0,>=4.5 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.9.1->instructor) (6.0.4)\n",
            "Requirement already satisfied: yarl<2.0,>=1.0 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.9.1->instructor) (1.9.2)\n",
            "Requirement already satisfied: frozenlist>=1.1.1 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.9.1->instructor) (1.4.0)\n",
            "Requirement already satisfied: aiosignal>=1.1.2 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.9.1->instructor) (1.3.1)\n",
            "Requirement already satisfied: async-timeout<5.0,>=4.0 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.9.1->instructor) (4.0.3)\n",
            "Requirement already satisfied: anyio<5,>=3.5.0 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from openai<2.0.0,>=1.1.0->instructor) (3.7.1)\n",
            "Requirement already satisfied: distro<2,>=1.7.0 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from openai<2.0.0,>=1.1.0->instructor) (1.8.0)\n",
            "Requirement already satisfied: httpx<1,>=0.23.0 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from openai<2.0.0,>=1.1.0->instructor) (0.25.2)\n",
            "Requirement already satisfied: sniffio in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from openai<2.0.0,>=1.1.0->instructor) (1.3.0)\n",
            "Requirement already satisfied: tqdm>4 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from openai<2.0.0,>=1.1.0->instructor) (4.66.1)\n",
            "Requirement already satisfied: typing-extensions<5,>=4.7 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from openai<2.0.0,>=1.1.0->instructor) (4.8.0)\n",
            "Requirement already satisfied: annotated-types>=0.4.0 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from pydantic<3.0.0,>=2.0.2->instructor) (0.6.0)\n",
            "Requirement already satisfied: pydantic-core==2.14.5 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from pydantic<3.0.0,>=2.0.2->instructor) (2.14.5)\n",
            "Requirement already satisfied: markdown-it-py>=2.2.0 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from rich<14.0.0,>=13.7.0->instructor) (3.0.0)\n",
            "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from rich<14.0.0,>=13.7.0->instructor) (2.17.2)\n",
            "Requirement already satisfied: click<9.0.0,>=7.1.1 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from typer<0.10.0,>=0.9.0->instructor) (8.1.7)\n",
            "Requirement already satisfied: idna>=2.8 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from anyio<5,>=3.5.0->openai<2.0.0,>=1.1.0->instructor) (3.6)\n",
            "Requirement already satisfied: exceptiongroup in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from anyio<5,>=3.5.0->openai<2.0.0,>=1.1.0->instructor) (1.2.0)\n",
            "Requirement already satisfied: certifi in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from httpx<1,>=0.23.0->openai<2.0.0,>=1.1.0->instructor) (2023.11.17)\n",
            "Requirement already satisfied: httpcore==1.* in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from httpx<1,>=0.23.0->openai<2.0.0,>=1.1.0->instructor) (1.0.2)\n",
            "Requirement already satisfied: h11<0.15,>=0.13 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from httpcore==1.*->httpx<1,>=0.23.0->openai<2.0.0,>=1.1.0->instructor) (0.14.0)\n",
            "Requirement already satisfied: mdurl~=0.1 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from markdown-it-py>=2.2.0->rich<14.0.0,>=13.7.0->instructor) (0.1.2)\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "160.68s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: openai in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (1.7.1)\n",
            "Requirement already satisfied: anyio<5,>=3.5.0 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from openai) (3.7.1)\n",
            "Requirement already satisfied: distro<2,>=1.7.0 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from openai) (1.8.0)\n",
            "Requirement already satisfied: httpx<1,>=0.23.0 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from openai) (0.25.2)\n",
            "Requirement already satisfied: pydantic<3,>=1.9.0 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from openai) (2.5.2)\n",
            "Requirement already satisfied: sniffio in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from openai) (1.3.0)\n",
            "Requirement already satisfied: tqdm>4 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from openai) (4.66.1)\n",
            "Requirement already satisfied: typing-extensions<5,>=4.7 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from openai) (4.8.0)\n",
            "Requirement already satisfied: idna>=2.8 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from anyio<5,>=3.5.0->openai) (3.6)\n",
            "Requirement already satisfied: exceptiongroup in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from anyio<5,>=3.5.0->openai) (1.2.0)\n",
            "Requirement already satisfied: certifi in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from httpx<1,>=0.23.0->openai) (2023.11.17)\n",
            "Requirement already satisfied: httpcore==1.* in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from httpx<1,>=0.23.0->openai) (1.0.2)\n",
            "Requirement already satisfied: h11<0.15,>=0.13 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from httpcore==1.*->httpx<1,>=0.23.0->openai) (0.14.0)\n",
            "Requirement already satisfied: annotated-types>=0.4.0 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from pydantic<3,>=1.9.0->openai) (0.6.0)\n",
            "Requirement already satisfied: pydantic-core==2.14.5 in /Users/tomriddle1/anaconda3/envs/codeint/lib/python3.10/site-packages (from pydantic<3,>=1.9.0->openai) (2.14.5)\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/cb4a90a0dfcf4e5c9e7398d85fa23b78.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/cb4a90a0dfcf4e5c9e7398d85fa23b78.json\n",
            "Questions: [('Content-Based Question', 'What is the role of mitochondria in plant cells, and why is isolating them important for studying their function?'), ('Analytical Question', 'How do traditional methods of isolating mitochondria from plant cells compare to the recent advancements that combine plant protoplast isolation with animal mitochondria extraction?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where a researcher wants to study the mitochondria of a specific plant species. How could the simplified method for rice mitochondria be adapted and improved to suit their research needs?'), ('Contextual/Relational Question', 'How do the limitations of the current methods for isolating plant mitochondria, such as the requirement for large starting material and their effectiveness with green leaf tissues, impact the broader field of plant physiology research?'), ('User-Interactive Question', 'If you were a plant biologist, how would you approach the challenge of isolating mitochondria from a specific plant species that is known to have unique characteristics and requirements?')]\n",
            "Storing data for file_id: cb4a90a0dfcf4e5c9e7398d85fa23b78\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/cb4a90a0dfcf4e5c9e7398d85fa23b78.json\n",
            "Questions: [('Content-Based Question', 'How does the research on isolating mitochondria from model and crop plants contribute to our understanding of plant respiration and the role of mitochondria?'), ('Analytical Question', 'What are the key procedures and techniques involved in isolating mitochondria from model and crop plants, as described in scientific journals and books?'), ('Creative/Scenario-Based Question', 'Imagine you are a plant biologist studying the role of mitochondria in rice plants. How would you use the simplified method for isolating mitochondria to further your research?'), ('Contextual/Relational Question', 'How do the protocols for isolating mitochondria from plant cell culture differ from those used for model and crop plants, and what implications does this have for studying plant respiration?'), ('User-Interactive Question', 'Have you ever conducted any research involving mitochondria in plants? If so, what methods did you use for isolating mitochondria and what were your findings?')]\n",
            "Storing data for file_id: cb4a90a0dfcf4e5c9e7398d85fa23b78\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/cb4a90a0dfcf4e5c9e7398d85fa23b78.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/cb4a90a0dfcf4e5c9e7398d85fa23b78.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/ff5a9d31dd07438e9b76b1be3d917cca.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/ff5a9d31dd07438e9b76b1be3d917cca.json\n",
            "Questions: [('Content-Based Question', 'How do cryopreservation techniques using cryoprotectants like trehalose, DMSO, glycerol, and ethylene glycol contribute to the viability and functionality preservation of isolated mitochondria?'), ('Analytical Question', 'What are the key quality control measures that need to be implemented to ensure the structural and functional integrity of mitochondria during the isolation process?'), ('Creative/Scenario-Based Question', 'Imagine a future where mitochondrial transplantation becomes a widely accepted treatment for neurodegenerative diseases. How might this impact the healthcare system and the lives of patients?'), ('Contextual/Relational Question', 'How does investigating mitochondrial dysfunction in cellular models of neurodegenerative diseases contribute to the development of therapeutic approaches for mitochondrial diseases?'), ('User-Interactive Question', 'If you were a researcher studying mitochondrial transplantation, what additional methods or techniques would you explore to improve the safety and efficacy of this therapeutic approach?')]\n",
            "Storing data for file_id: ff5a9d31dd07438e9b76b1be3d917cca\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/ff5a9d31dd07438e9b76b1be3d917cca.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How do the sources contribute to our understanding of mitochondrial preservation and viability?'), ('Analytical Question', 'What are the key findings and insights from the sources regarding standardized protocols for assessing mitochondrial function?'), ('Creative/Scenario-Based Question', 'Imagine you are a researcher in the field of mitochondrial medicine. How could the information from these sources help you develop innovative preservation methods?'), ('Contextual/Relational Question', 'How do the findings from these sources relate to the broader field of mitochondrial medicine and its potential impact on healthcare?'), ('User-Interactive Question', 'What are your thoughts on the importance of mitochondrial preservation and viability in advancing medical research and potential therapeutic interventions?')]\n",
            "Storing data for file_id: ff5a9d31dd07438e9b76b1be3d917cca\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/ff5a9d31dd07438e9b76b1be3d917cca.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/ff5a9d31dd07438e9b76b1be3d917cca.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/a2c5e4f152574f7e84097824587b9394.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/a2c5e4f152574f7e84097824587b9394.json\n",
            "Questions: [('Content-Based Question', 'How have recent advancements in gentle mitochondrial isolation techniques improved the preservation of mitochondrial integrity and function?'), ('Analytical Question', 'What are the key factors that have contributed to the optimization of differential filtration-based techniques for isolating high-quality and quantity mitochondria from cellular models?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where mitochondrial transplantation becomes a widely accepted therapy. How might this impact the field of medicine and our understanding of mitochondrial function?'), ('Contextual/Relational Question', 'How do the advancements in microscale and nanoscale isolation techniques, such as nanoprobe-based technologies, contribute to our understanding of mitochondrial heterogeneity and dysfunction?'), ('User-Interactive Question', 'How do you think breakthroughs in mitochondrial research and medicine can potentially revolutionize the development of novel therapies and our overall understanding of mitochondrial function?')]\n",
            "Storing data for file_id: a2c5e4f152574f7e84097824587b9394\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/a2c5e4f152574f7e84097824587b9394.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How do mitochondria contribute to signal integration within cells?'), ('Analytical Question', 'What would be the consequences if mitochondria were unable to perform their role as signal integrators within cells?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where mitochondria lose their ability to integrate signals within cells. How would this impact the overall functioning of the organism?'), ('Contextual/Relational Question', 'How do the physiological processes that rely on mitochondria as signal integrators relate to other cellular processes, such as gene expression or cellular metabolism?'), ('User-Interactive Question', 'Can you think of any examples where the role of mitochondria as signal integrators within cells could be disrupted, and what effects it might have on cellular function?')]\n",
            "Storing data for file_id: a2c5e4f152574f7e84097824587b9394\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/a2c5e4f152574f7e84097824587b9394.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/a2c5e4f152574f7e84097824587b9394.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6829ccbe621c4f47a47e02bb959c7136.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6829ccbe621c4f47a47e02bb959c7136.json\n",
            "Questions: [('Content-Based Question', 'How can advanced search techniques and APIs like Historical PowerTrack and Full-Archive Search be used to explore Wikidata through Twitter?'), ('Analytical Question', 'What are the challenges associated with using Twitter to explore Wikidata, such as the volume of tweets and limited API access?'), ('Creative/Scenario-Based Question', 'Imagine you are a researcher interested in studying the historical evolution of discussions about Wikidata on Twitter. How would you use hashtags like #WikidataQuery or #SPARQL, along with a date range, to narrow down your search results?'), ('Contextual/Relational Question', \"How do Twitter's APIs, including the Historical PowerTrack API and Full-Archive Search API, provide powerful tools for programmatically searching and retrieving tweets related to Wikidata?\"), ('User-Interactive Question', 'Have you ever used publicly shared dashboards, like the #taycan hashtag dashboard, to gain insights into the historical evolution of discussions, interactions, top influencers, and the tone of tweets? If so, what valuable information did you discover?')]\n",
            "Storing data for file_id: 6829ccbe621c4f47a47e02bb959c7136\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6829ccbe621c4f47a47e02bb959c7136.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6829ccbe621c4f47a47e02bb959c7136.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/c693d2576ff948ca9c68e186e535d3f7.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/c693d2576ff948ca9c68e186e535d3f7.json\n",
            "Questions: [('Content-Based Question', 'How does synthetic biology replicate mitochondrial ATP production in artificial environments, and what are the challenges associated with it?'), ('Analytical Question', 'What are the key factors that impact ATP production in synthetic environments, and how can they be optimized for energy efficiency?'), ('Creative/Scenario-Based Question', 'Imagine a future where synthetic environments can provide a limitless and sustainable supply of ATP. How would this impact various industries such as medicine, energy, and biotechnology?'), ('Contextual/Relational Question', 'How does the concept of synthetic environments for mitochondrial ATP production relate to the field of synthetic biology as a whole, and what other applications can be explored using this technology?'), ('User-Interactive Question', 'If you were a researcher working on synthetic environments for ATP production, what specific challenges would you focus on addressing, and how would you approach solving them?')]\n",
            "Storing data for file_id: c693d2576ff948ca9c68e186e535d3f7\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/c693d2576ff948ca9c68e186e535d3f7.json\n",
            "Questions: [('Content-Based Question', 'How do Sogang University and Harbin Institute of Technology propose to explore sustainable chemical energy conversion and production through artificial organelles?'), ('Analytical Question', 'What are the potential advantages of using artificial mitochondria and chloroplasts as a solution for sustainable energy production?'), ('Creative/Scenario-Based Question', 'Imagine a world where artificial organelles are successfully implemented for energy production. How might this impact global energy consumption and the environment?'), ('Contextual/Relational Question', 'How does the research on artificial organelles in sustainable energy production relate to the articles published in Nature and AIP Publishing?'), ('User-Interactive Question', 'What are your thoughts on the potential of artificial organelles as a solution for sustainable energy? How do you think this technology could be further developed and implemented in real-world applications?')]\n",
            "Storing data for file_id: c693d2576ff948ca9c68e186e535d3f7\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/c693d2576ff948ca9c68e186e535d3f7.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/c693d2576ff948ca9c68e186e535d3f7.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/113e61b80ff84f9ebc77f984f302540e.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/113e61b80ff84f9ebc77f984f302540e.json\n",
            "Questions: [('Content-Based Question', 'How do recent advancements in isolating mitochondria from plant cell cultures simplify the process and eliminate the need for ultracentrifugation?'), ('Analytical Question', 'What are the specific methods used to assess mitochondrial purity and evaluate outer membrane integrity and respiratory function in plant mitochondria?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where the traditional method of isolating mitochondria from plant cell cultures is still being used. How would the introduction of these simplified methods impact the efficiency and cost-effectiveness of mitochondrial research?'), ('Contextual/Relational Question', 'How do the advancements in isolating mitochondria in plant species using density gradient centrifugation and marker enzyme assays relate to the broader field of plant mitochondrial research and its applications in studying crop species?'), ('User-Interactive Question', 'If you were a researcher studying mitochondrial structure and function in plant species, how would these improved methods for isolating mitochondria impact your research process and outcomes?')]\n",
            "Storing data for file_id: 113e61b80ff84f9ebc77f984f302540e\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/113e61b80ff84f9ebc77f984f302540e.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/113e61b80ff84f9ebc77f984f302540e.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/25502641c6f6461cbaa3eaf2aba53f97.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/25502641c6f6461cbaa3eaf2aba53f97.json\n",
            "Questions: [('Content-Based Question', 'How do mitochondria regulate their dynamics and why is it important for cellular homeostasis, energy production, and apoptosis?'), ('Analytical Question', 'What are the different techniques used to simulate mitochondrial fusion and fission processes outside the cell, and how do they contribute to our understanding of these processes?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where researchers successfully manipulate mitochondrial dynamics using pharmacological and genetic approaches. How could this advancement impact drug discovery and our understanding of related diseases?'), ('Contextual/Relational Question', 'How can the study of natural and artificial mitochondrial transfer techniques help us better understand cell-cell fusion and manipulate mitochondrial function?'), ('User-Interactive Question', 'Can you think of any other potential applications or research areas where studying mitochondrial dynamics and manipulating fusion and fission processes could be beneficial?')]\n",
            "Storing data for file_id: 25502641c6f6461cbaa3eaf2aba53f97\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/25502641c6f6461cbaa3eaf2aba53f97.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How do mitochondrial dynamics, including fission and fusion processes, contribute to our understanding of mitochondrial-related diseases?'), ('Analytical Question', 'What are the key methods and techniques used to simulate mitochondrial dynamics, and how do they help researchers explore potential therapeutic interventions?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where a new therapeutic intervention for mitochondrial-related diseases is discovered through the study of mitochondrial dynamics. How would this breakthrough impact the field of medicine?'), ('Contextual/Relational Question', 'How do the processes of mitochondrial fission and fusion relate to each other, and how do they collectively contribute to the overall health and function of mitochondria?'), ('User-Interactive Question', 'What are some potential applications of the knowledge gained from studying mitochondrial dynamics in developing treatments for other diseases or conditions?')]\n",
            "Storing data for file_id: 25502641c6f6461cbaa3eaf2aba53f97\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/25502641c6f6461cbaa3eaf2aba53f97.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/25502641c6f6461cbaa3eaf2aba53f97.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/19bc86d438fd4d06b14a99c32595b0ad.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/19bc86d438fd4d06b14a99c32595b0ad.json\n",
            "Questions: [('Content-Based Question', 'How can we create or mimic the essential cellular conditions that support mitochondrial functions, such as ATP production, outside of the cellular environment?'), ('Analytical Question', 'What are the challenges and future directions in creating artificial cells with biologically realistic energy-generation methods?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where a biomimetic artificial organelle capable of ATP production is successfully constructed. How could this technology be applied in the field of medicine?'), ('Contextual/Relational Question', 'How do mitochondrial dynamics and cellular stress responses contribute to disease modeling, and how can they be targeted for therapeutic purposes?'), ('User-Interactive Question', 'Considering the sources mentioned in the summary (such as Nature Communications and Boundless Biology), how would you approach designing an experiment to replicate mitochondrial ATP production outside of the cellular environment?')]\n",
            "Storing data for file_id: 19bc86d438fd4d06b14a99c32595b0ad\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/19bc86d438fd4d06b14a99c32595b0ad.json\n",
            "Questions: [('Content-Based Question', 'How does understanding mitochondrial dynamics and stress responses contribute to our knowledge of ATP production and replication?'), ('Analytical Question', 'What are the different sources of evidence, including theoretical models and experimental research, that support the understanding of mitochondrial ATP production and replication?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where artificial organelles capable of ATP synthesis are successfully developed. How could this impact the field of cellular research and the development of new therapies?'), ('Contextual/Relational Question', 'How does the understanding of mitochondrial function and the development of methods to mimic cellular conditions relate to the goal of replicating ATP production outside of the cellular environment?'), ('User-Interactive Question', 'How would you approach designing an experiment to study mitochondrial ATP production and replication using artificial organelles?')]\n",
            "Storing data for file_id: 19bc86d438fd4d06b14a99c32595b0ad\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/19bc86d438fd4d06b14a99c32595b0ad.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/19bc86d438fd4d06b14a99c32595b0ad.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/2d8a595d69384ac88ffe6cfed2beb386.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/2d8a595d69384ac88ffe6cfed2beb386.json\n",
            "Questions: [('Content-Based Question', 'How does treating data as a product prioritize data quality, consistency, and accessibility in knowledge graph analysis?'), ('Analytical Question', 'What are the benefits of using standardized languages like RDF and query languages like SPARQL in constructing knowledge graphs for data integration and interoperability?'), ('Creative/Scenario-Based Question', 'Imagine you are part of a team responsible for building a knowledgeable team for knowledge graph analysis. How would you ensure that the team has expertise in semantic web standards, knowledge representation, and graph databases?'), ('Contextual/Relational Question', 'How does the scalability and agility of knowledge graphs accommodate data growth and adapt to new sources in data preparation for AI applications?'), ('User-Interactive Question', 'How can continuous research and development in knowledge graph analysis, embeddings, acquisition, and reasoning contribute to solving data challenges and revolutionizing data analytics?')]\n",
            "Storing data for file_id: 2d8a595d69384ac88ffe6cfed2beb386\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/2d8a595d69384ac88ffe6cfed2beb386.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/2d8a595d69384ac88ffe6cfed2beb386.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/8fee094a8fd542b2863f828292058780.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/8fee094a8fd542b2863f828292058780.json\n",
            "Questions: [('Content-Based Question', 'How does mitochondrial research contribute to our understanding of cellular metabolism, signaling, and disease pathogenesis?'), ('Analytical Question', 'What are the advantages and disadvantages of macroscale and microscale/nanoscale techniques for isolating mitochondria in terms of purity, integrity, and resolution?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where a researcher needs to study mitochondrial heterogeneity and dysfunction. How would they choose between using magnetic beads and antibodies versus gradient centrifugation and nitrogen cavitation for isolating mitochondria?'), ('Contextual/Relational Question', 'How does the choice of isolation technique for mitochondria impact the quality of the obtained mitochondria and the accuracy of scientific research and disease pathogenesis studies?'), ('User-Interactive Question', 'In your opinion, what factors should researchers prioritize when selecting an isolation technique for mitochondria in their scientific research? How would you balance yield, purity, integrity, and specific needs?')]\n",
            "Storing data for file_id: 8fee094a8fd542b2863f828292058780\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/8fee094a8fd542b2863f828292058780.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/8fee094a8fd542b2863f828292058780.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/fa591421670b49f4bdffc9b3e53be8c2.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/fa591421670b49f4bdffc9b3e53be8c2.json\n",
            "Error in extracting knowledge: Error code: 400 - {'error': {'message': \"This model's maximum context length is 4097 tokens. However, your messages resulted in 4463 tokens. Please reduce the length of the messages.\", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}\n",
            "Error in extracting dirty knowledge: Error in extracting knowledge\n",
            " Trying again\n",
            "Error in extracting knowledge: Error code: 400 - {'error': {'message': \"This model's maximum context length is 4097 tokens. However, your messages resulted in 4463 tokens. Please reduce the length of the messages.\", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}\n",
            "Error summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/fa591421670b49f4bdffc9b3e53be8c2.md: Error in extracting knowledge\n",
            "Error summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/fa591421670b49f4bdffc9b3e53be8c2.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/83ff310aa5414ce3b7da3b2273101518.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/83ff310aa5414ce3b7da3b2273101518.json\n",
            "Questions: [('Content-Based Question', 'How does a 3D printed microfluidic platform contribute to the isolation of mitochondria?'), ('Analytical Question', 'What are the advantages and disadvantages of using micro- and nanoscale techniques for mitochondrial isolation?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where a chemical-free microfluidic extraction method for functional mitochondria is successfully implemented. How could this impact the field of mitochondrial research?'), ('Contextual/Relational Question', 'How do recent advances in microfluidic technologies for cell separation contribute to the optimization of mitochondrial isolation protocols?'), ('User-Interactive Question', 'Have you ever encountered any challenges or limitations when working with microfluidic platforms for mitochondrial extraction? How did you overcome them?')]\n",
            "Storing data for file_id: 83ff310aa5414ce3b7da3b2273101518\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/83ff310aa5414ce3b7da3b2273101518.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How does the novel method combining coculture and droplet microfluidics enable high-throughput and quantitative isolation of mitochondria?'), ('Analytical Question', 'What are the advantages of using coculture and droplet microfluidics in isolating mitochondria compared to traditional methods?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where the novel method for isolating mitochondria was not successful. How would this impact research in the field of mitochondrial biology?'), ('Contextual/Relational Question', 'How does the ability to isolate mitochondria in a high-throughput and quantitative manner contribute to our understanding of mitochondrial function and its role in various biological processes?'), ('User-Interactive Question', 'How might the combination of coculture and droplet microfluidics be further optimized to improve the efficiency and accuracy of isolating mitochondria?')]\n",
            "Storing data for file_id: 83ff310aa5414ce3b7da3b2273101518\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/83ff310aa5414ce3b7da3b2273101518.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/83ff310aa5414ce3b7da3b2273101518.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/5687242aecfd458b8f0bdc1a8e9942e1.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/5687242aecfd458b8f0bdc1a8e9942e1.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How have researchers replicated components of mitochondria in artificial environments, and what advancements have been made in terms of energy efficiency?'), ('Analytical Question', 'What are the challenges involved in enabling self-adaptation of artificial mitochondria in changing environments, and how can these challenges be addressed?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where synthetic cells with autonomous energy generation and molecule synthesis are widely available. How could this technology revolutionize the field of regenerative medicine?'), ('Contextual/Relational Question', 'How do the methodologies of biochemical engineering, synthetic biology, and computational modeling contribute to the construction of artificial mitochondria, and how do they complement each other in this process?'), ('User-Interactive Question', 'Considering the potential applications of artificial mitochondria, such as drug delivery systems and manufacturing compounds, how do you think the development of regulatory frameworks can ensure responsible research and ethical considerations in this field?')]\n",
            "Storing data for file_id: 5687242aecfd458b8f0bdc1a8e9942e1\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/5687242aecfd458b8f0bdc1a8e9942e1.json\n",
            "Questions: [('Content-Based Question', 'How are advancements in computational modeling and protective membrane structures contributing to the development of artificial organelles and cells?'), ('Analytical Question', 'What challenges need to be overcome in constructing artificial mitochondria and chloroplasts to make them viable powerhouses of the future?'), ('Creative/Scenario-Based Question', 'Imagine a world where artificial organelles and cells are successfully developed. How might this technology revolutionize the field of sustainable chemical energy conversion and production?'), ('Contextual/Relational Question', 'How do the methodologies of computational modeling and the use of protective membrane structures contribute to the progress being made in the development of complex synthetic cells?'), ('User-Interactive Question', 'What ethical considerations do you think should be taken into account when exploring the potential applications of artificial organelles and cells?')]\n",
            "Storing data for file_id: 5687242aecfd458b8f0bdc1a8e9942e1\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/5687242aecfd458b8f0bdc1a8e9942e1.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/5687242aecfd458b8f0bdc1a8e9942e1.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/d186b27fa8f74a20998bb45769dd60d5.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/d186b27fa8f74a20998bb45769dd60d5.json\n",
            "Questions: [('Content-Based Question', 'How does the combination of traditional plant protoplast isolation with mammalian mitochondria extraction protocols improve the isolation of pure mitochondria from different plant species and tissues?'), ('Analytical Question', 'What are the specific steps involved in the protocol for isolating intact mitochondria from plant cells, and how do these steps ensure mitochondrial integrity and functionality?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where a researcher wants to isolate mitochondria from a rare plant species with a complex cell wall structure. How would they modify the protocol to ensure successful isolation?'), ('Contextual/Relational Question', 'How do the advancements in isolating pure mitochondria from plant cells contribute to the broader field of mitochondrial research and its applications in studying different plant species?'), ('User-Interactive Question', 'As a molecular biologist, how would you adapt the protocol for isolating intact mitochondria from plant cells to study the specific mitochondrial functions in a particular plant species of interest?')]\n",
            "Storing data for file_id: d186b27fa8f74a20998bb45769dd60d5\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/d186b27fa8f74a20998bb45769dd60d5.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/d186b27fa8f74a20998bb45769dd60d5.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/76725ca31bcc4918b4b806f0493de4e9.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/76725ca31bcc4918b4b806f0493de4e9.json\n",
            "Questions: [('Content-Based Question', 'How does ATP synthesis occur within mitochondria, and what are the key factors involved in this process?'), ('Analytical Question', 'What are the thermodynamic and kinetic principles that govern ATP hydrolysis, and how do they differ from those governing ATP synthesis?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where the spatial and temporal dynamics of ATP synthase are disrupted. How might this impact cellular energetics and overall cellular function?'), ('Contextual/Relational Question', 'How does understanding the mechanisms of ATP synthesis in mitochondria contribute to the development of therapeutic strategies for neurodegenerative and age-associated pathologies?'), ('User-Interactive Question', 'How might computational frameworks and synthetic environments for mitochondrial ATP production be used to engage students in bioenergetic research and enhance their understanding of cellular energetics?')]\n",
            "Storing data for file_id: 76725ca31bcc4918b4b806f0493de4e9\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/76725ca31bcc4918b4b806f0493de4e9.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How does IF1 regulate mitochondrial ATP synthase?'), ('Analytical Question', 'What would happen if there was a deficiency of IF1 in the regulation of mitochondrial ATP synthase?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where IF1 is completely absent in the regulation of mitochondrial ATP synthase. How would this affect the energy production in the cell?'), ('Contextual/Relational Question', 'How does the regulation of mitochondrial ATP synthase by IF1 relate to other processes involved in cellular energy metabolism?'), ('User-Interactive Question', 'Can you think of any other molecules or proteins that might interact with IF1 in the regulation of mitochondrial ATP synthase?')]\n",
            "Storing data for file_id: 76725ca31bcc4918b4b806f0493de4e9\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/76725ca31bcc4918b4b806f0493de4e9.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/76725ca31bcc4918b4b806f0493de4e9.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/b03d541d30c047d886d44feab151b41e.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/b03d541d30c047d886d44feab151b41e.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How does oxidative phosphorylation contribute to energy production in mitochondria?'), ('Analytical Question', 'What are the potential benefits and challenges of engineering new pathways for sustained energy production independent of traditional substrates?'), ('Creative/Scenario-Based Question', 'Imagine a world where humans can generate energy through metabolic switching. How might this impact our daily lives and the environment?'), ('Contextual/Relational Question', 'How does the manipulation of mitochondrial dynamics and the development of artificial organelles relate to the goal of bioengineering efficient energy production systems?'), ('User-Interactive Question', 'In your opinion, what are some ethical considerations that should be taken into account when exploring strategies for enhancing mitochondrial function through synthetic biology?')]\n",
            "Storing data for file_id: b03d541d30c047d886d44feab151b41e\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/b03d541d30c047d886d44feab151b41e.json\n",
            "Questions: [('Content-Based Question', 'How does optimizing exercise training impact mitochondrial function and overall health?'), ('Analytical Question', 'What are the specific pathways that can be targeted to alleviate mitochondrial dysfunction in neurodegenerative diseases?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where researchers discover a new method to enhance mitochondrial function. How could this discovery be applied to improve human health?'), ('Contextual/Relational Question', 'How does understanding the molecular and supramolecular structure of the mitochondrial oxidative phosphorylation system provide insights into pathological conditions?'), ('User-Interactive Question', 'How can individuals incorporate mitohormesis into their exercise routines to potentially enhance mitochondrial function?')]\n",
            "Storing data for file_id: b03d541d30c047d886d44feab151b41e\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/b03d541d30c047d886d44feab151b41e.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/b03d541d30c047d886d44feab151b41e.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/641d140e02854f2f9cd77da5d347168d.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/641d140e02854f2f9cd77da5d347168d.json\n",
            "Error in extracting knowledge: Error code: 400 - {'error': {'message': \"This model's maximum context length is 4097 tokens. However, your messages resulted in 4230 tokens. Please reduce the length of the messages.\", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}\n",
            "Error in extracting dirty knowledge: Error in extracting knowledge\n",
            " Trying again\n",
            "Error in extracting knowledge: Error code: 400 - {'error': {'message': \"This model's maximum context length is 4097 tokens. However, your messages resulted in 4230 tokens. Please reduce the length of the messages.\", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}\n",
            "Error summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/641d140e02854f2f9cd77da5d347168d.md: Error in extracting knowledge\n",
            "Error summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/641d140e02854f2f9cd77da5d347168d.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/eaa052c350e6467282c2f1d4a7767d17.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/eaa052c350e6467282c2f1d4a7767d17.json\n",
            "Questions: [('Content-Based Question', 'How are the purity and integrity of isolated mitochondria assessed in research?'), ('Analytical Question', 'What are the challenges faced in accurately detecting abnormalities in isolated mitochondria?'), ('Creative/Scenario-Based Question', \"Imagine a scenario where the purity of isolated mitochondria is compromised during research. How might this affect the reliability of the study's findings?\"), ('Contextual/Relational Question', 'How do rapid and effective isolation methods, such as differential filtration, contribute to improving the quality of isolated mitochondria?'), ('User-Interactive Question', 'Can you think of any other methods or techniques that could be used to assess the integrity and functionality of isolated mitochondria?')]\n",
            "Storing data for file_id: eaa052c350e6467282c2f1d4a7767d17\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/eaa052c350e6467282c2f1d4a7767d17.json\n",
            "Questions: [('Content-Based Question', 'How do rapid techniques like tissue dissociation and differential filtration contribute to efficient isolation and purification of mitochondria for biogenetical studies?'), ('Analytical Question', 'What are the methods used to estimate mitochondrial integrity, such as enzymatic activities or membrane permeability, and why is it crucial to maintain it?'), ('Creative/Scenario-Based Question', 'Imagine you are a researcher studying mitochondria. How would you use high-resolution lipid profiling to gather detailed information about mitochondria from cell culture samples?'), ('Contextual/Relational Question', 'How can the detection of mitochondrial abnormalities aid in the diagnosis and treatment of various conditions, and what are some examples of such conditions?'), ('User-Interactive Question', 'Have you ever used the Seahorse XFe24 analyzer or any other equipment to measure respiratory function in isolated mitochondria? If so, what were your findings, and how did they contribute to your research or understanding of mitochondrial function?')]\n",
            "Storing data for file_id: eaa052c350e6467282c2f1d4a7767d17\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/eaa052c350e6467282c2f1d4a7767d17.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/eaa052c350e6467282c2f1d4a7767d17.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/514db44d0e4244269605aa45b1a8dbc2.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/514db44d0e4244269605aa45b1a8dbc2.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How does Neo4j enhance search, AI, and data analytics through the use of knowledge graphs?'), ('Analytical Question', 'What are the key steps involved in preparing data for import into Neo4j, and why is this process important?'), ('Creative/Scenario-Based Question', 'Imagine you are working on a project that requires importing CSV data into Neo4j. What challenges might you encounter during the import process, and how would you overcome them?'), ('Contextual/Relational Question', 'How does the Neo4j Data Importer tool simplify the process of importing data into Neo4j, and what benefits does it provide for users?'), ('User-Interactive Question', 'Have you ever worked with a graph database platform like Neo4j? If so, what were some of the challenges you faced during the data import and verification process? If not, what questions or concerns do you have about using a graph database for knowledge graph creation and management?')]\n",
            "Storing data for file_id: 514db44d0e4244269605aa45b1a8dbc2\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/514db44d0e4244269605aa45b1a8dbc2.json\n",
            "Questions: [('Content-Based Question', \"How can beginners utilize Neo4j's comprehensive documentation to import data from API sources and relational databases?\"), ('Analytical Question', 'What are the advantages and disadvantages of importing data from API sources compared to importing data from relational databases using Neo4j?'), ('Creative/Scenario-Based Question', \"Imagine you are a data scientist working on a project that requires importing data from both API sources and a relational database. How would you leverage Neo4j's documentation and resources to efficiently import and integrate this data into a knowledge graph?\"), ('Contextual/Relational Question', 'How does the information provided in O\\'Reilly Media\\'s book \"Building Knowledge Graphs\" complement the practical insights and step-by-step instructions offered in Neo4j\\'s documentation for data import?'), ('User-Interactive Question', \"Have you ever encountered challenges or limitations when importing data from API sources or relational databases? How do you think Neo4j's documentation and O'Reilly Media's book could help address those challenges?\")]\n",
            "Storing data for file_id: 514db44d0e4244269605aa45b1a8dbc2\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/514db44d0e4244269605aa45b1a8dbc2.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/514db44d0e4244269605aa45b1a8dbc2.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/34c57b9fb1bb4e4e822e0e1baad1772c.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/34c57b9fb1bb4e4e822e0e1baad1772c.json\n",
            "Questions: [('Content-Based Question', 'How do bioengineering techniques, such as nanotechnology and mitochondrial transplantation, contribute to enhancing mitochondrial function and self-sufficiency?'), ('Analytical Question', 'What are the potential applications of regenerative medicine and tissue engineering in treating conditions like ischemia-reperfusion injury, nerve damage, and neurological disorders through the delivery and retention of healthy mitochondria?'), ('Creative/Scenario-Based Question', 'Imagine a future where nanotechnology-mediated approaches for enhancing mitochondrial self-sufficiency have been successfully translated into clinical applications. How might this impact the treatment and management of mitochondrial diseases?'), ('Contextual/Relational Question', 'How can interdisciplinary collaboration between bioengineers, molecular biologists, and clinicians help overcome challenges in peptide self-assembly and understanding molecular mechanisms for the detection and treatment of mitochondrial diseases?'), ('User-Interactive Question', 'What are some potential ethical considerations that need to be addressed when utilizing bioengineering techniques, such as nanotechnology and mitochondrial transplantation, for enhancing mitochondrial function and self-sufficiency?')]\n",
            "Storing data for file_id: 34c57b9fb1bb4e4e822e0e1baad1772c\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/34c57b9fb1bb4e4e822e0e1baad1772c.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/34c57b9fb1bb4e4e822e0e1baad1772c.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/2b224a28aab246e898126d440e8abd74.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/2b224a28aab246e898126d440e8abd74.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How does understanding SPARQL contribute to effective information access and manipulation in Wikidata?'), ('Analytical Question', \"What are the potential research and innovation opportunities that can be explored using Wikidata's vast knowledge base?\"), ('Creative/Scenario-Based Question', 'Imagine you are a researcher looking to analyze data from Wikidata. How would leveraging the Query Service tutorial enhance your data analysis capabilities?'), ('Contextual/Relational Question', 'How does the accessibility and reusability of data promoted by the Query Service tutorial contribute to building a community of informed users in the evolving digital landscape?'), ('User-Interactive Question', 'Have you ever used SPARQL or the Query Service tutorial to access and manipulate information in Wikidata? If so, how has it impacted your research or data analysis process?')]\n",
            "Storing data for file_id: 2b224a28aab246e898126d440e8abd74\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/2b224a28aab246e898126d440e8abd74.json\n",
            "Questions: [('Content-Based Question', 'How did Dr. Keren Shatzman contribute to the educational tools and gateways for accessing information through the Query Service Tutorial for Wikidata?'), ('Analytical Question', 'What are the key features and capabilities of the Wikidata Query Service, as explained in the user manual developed by Dr. Keren Shatzman?'), ('Creative/Scenario-Based Question', 'Imagine you are a student interested in exploring a specific topic using the Wikidata Query Service. How would you use the educational tools and gateways developed by Dr. Keren Shatzman to access relevant information?'), ('Contextual/Relational Question', 'How does the support from the broader Wikidata community contribute to the effectiveness and accessibility of the educational tools and gateways developed by Dr. Keren Shatzman?'), ('User-Interactive Question', 'Have you ever used the Wikidata Query Service or any similar tools to access information? If so, how did it enhance your understanding and exploration of a specific topic?')]\n",
            "Storing data for file_id: 2b224a28aab246e898126d440e8abd74\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/2b224a28aab246e898126d440e8abd74.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/2b224a28aab246e898126d440e8abd74.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/cc60050b611b4ac2bd8727ea791457b0.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/cc60050b611b4ac2bd8727ea791457b0.json\n",
            "Questions: [('Content-Based Question', 'What are the starting materials used for isolating functionally competent mitochondria according to the protocol described in the report?'), ('Analytical Question', 'How does the protocol ensure the purity of the isolated mitochondria?'), ('Creative/Scenario-Based Question', 'Imagine you are a researcher studying mitochondrial function in neuronal tissues. How would you modify the protocol to accommodate the specific requirements of neuronal tissues?'), ('Contextual/Relational Question', 'How does the addition of fatty-acid-free BSA in the buffer system contribute to the isolation of high-quality mitochondria?'), ('User-Interactive Question', 'Have you ever conducted any experiments involving mitochondrial isolation? If so, what challenges did you encounter and how did you overcome them?')]\n",
            "Storing data for file_id: cc60050b611b4ac2bd8727ea791457b0\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/cc60050b611b4ac2bd8727ea791457b0.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/cc60050b611b4ac2bd8727ea791457b0.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6f37373983f54caf8eae14058cb786c8.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6f37373983f54caf8eae14058cb786c8.json\n",
            "Questions: [('Content-Based Question', 'How can mitochondrial manipulation, including gene therapy, be used to treat mitochondrial diseases?'), ('Analytical Question', 'What are the key ethical considerations and safety measures that need to be taken into account when developing and implementing mitochondrial manipulation therapies?'), ('Creative/Scenario-Based Question', 'Imagine a future where mitochondrial manipulation therapies are widely accessible. How might this impact the treatment and management of mitochondrial diseases?'), ('Contextual/Relational Question', 'How do preclinical studies and clinical trial design contribute to the development and evaluation of safe and effective mitochondrial manipulation therapies?'), ('User-Interactive Question', 'Considering the intergenerational impact and potential risks associated with germline modification, how can equitable access to mitochondrial manipulation therapies be ensured while maintaining ethical standards?')]\n",
            "Storing data for file_id: 6f37373983f54caf8eae14058cb786c8\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6f37373983f54caf8eae14058cb786c8.json\n",
            "Questions: [('Content-Based Question', 'What are the key ethical, social, and policy considerations discussed in the article \"Mitochondrial Replacement Techniques\"?'), ('Analytical Question', 'How do the National Academies of Sciences, Engineering, and Medicine contribute to the responsible implementation of mitochondrial replacement techniques?'), ('Creative/Scenario-Based Question', 'Imagine a society where mitochondrial replacement techniques are widely accepted. How might this impact the perception of genetic diseases and disabilities?'), ('Contextual/Relational Question', 'How do the ethical considerations for mitochondrial replacement techniques compare to those of other reproductive technologies, such as in vitro fertilization?'), ('User-Interactive Question', 'What are your thoughts on the ethical and social implications of mitochondrial replacement techniques? How do you think these techniques should be regulated to ensure responsible implementation?')]\n",
            "Storing data for file_id: 6f37373983f54caf8eae14058cb786c8\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6f37373983f54caf8eae14058cb786c8.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6f37373983f54caf8eae14058cb786c8.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/fa3be607e77a4b93a7bbc61135bea4bd.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/fa3be607e77a4b93a7bbc61135bea4bd.json\n",
            "Questions: [('Content-Based Question', 'How does Wikidata differ from traditional knowledge bases in terms of data structure and accessibility?'), ('Analytical Question', 'What are the potential advantages and limitations of utilizing SPARQL queries to retrieve information from Wikidata?'), ('Creative/Scenario-Based Question', 'Imagine you are a researcher studying the impact of climate change on biodiversity. How could you utilize Wikidata and its structured data to gather relevant information for your study?'), ('Contextual/Relational Question', 'How does the collaborative nature of Wikidata, with regular updates made by the community of editors, contribute to its reliability and accuracy as a knowledge base?'), ('User-Interactive Question', 'Have you ever used Wikidata or a similar structured data resource for question answering? If so, what was your experience like? If not, what potential applications can you envision for such a resource in your field or area of interest?')]\n",
            "Storing data for file_id: fa3be607e77a4b93a7bbc61135bea4bd\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/fa3be607e77a4b93a7bbc61135bea4bd.json\n",
            "Questions: [('Content-Based Question', \"How does Dr. Keren Shatzman's discussion on the Query Service Tutorial for Wikidata highlight the potential for creative and innovative use of data?\"), ('Analytical Question', 'What are the key features and benefits of the SPARQL tutorial offered by Wikidata for users interested in learning querying?'), ('Creative/Scenario-Based Question', 'Imagine you are a researcher looking to explore the creative possibilities of the growing knowledge base. How could you utilize the Query Service Tutorial for Wikidata to enhance your innovative use of data?'), ('Contextual/Relational Question', \"How does the assistance provided by Wikipedia's Help:Wikidata page relate to the Query Service Tutorial for Wikidata discussed by Dr. Keren Shatzman?\"), ('User-Interactive Question', 'Have you ever used the Query Service Tutorial for Wikidata or the SPARQL tutorial offered by Wikidata? If so, what was your experience like and how did it enhance your understanding of querying?')]\n",
            "Storing data for file_id: fa3be607e77a4b93a7bbc61135bea4bd\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/fa3be607e77a4b93a7bbc61135bea4bd.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/fa3be607e77a4b93a7bbc61135bea4bd.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/be556138d3cc4cea9f89110c8cae3f9f.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/be556138d3cc4cea9f89110c8cae3f9f.json\n",
            "Questions: [('Content-Based Question', 'How does understanding the molecular and supramolecular structure of the mitochondrial oxidative phosphorylation system contribute to the engineering of new pathways for sustained energy production?'), ('Analytical Question', 'What are the potential targetable pathways for alleviating mitochondrial dysfunction that could be engineered for sustained energy production?'), ('Creative/Scenario-Based Question', 'Imagine a future where alternative energy production pathways are successfully engineered in mitochondria. How might this impact the treatment of diseases related to mitochondrial dysfunction?'), ('Contextual/Relational Question', 'How do the adaptive responses to mitochondrial stress and quality control mechanisms provide insights into alternative energy production pathways?'), ('User-Interactive Question', 'How might computational frameworks for analyzing and simulating mitochondrial ATP synthesis be used to model and discover new energy production pathways?')]\n",
            "Storing data for file_id: be556138d3cc4cea9f89110c8cae3f9f\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/be556138d3cc4cea9f89110c8cae3f9f.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How does the study in \"10 MFSD7C Switches Mitochondrial ATP Synthesis to Thermogenesis in Response to Heme\" contribute to our understanding of mitochondrial energy production?'), ('Analytical Question', 'What are the key findings and implications of the study on the ADP/ATP carrier in terms of substrate binding and the transport cycle?'), ('Creative/Scenario-Based Question', 'Imagine if we could manipulate the switch between ATP synthesis and thermogenesis in mitochondria. How could this knowledge be applied to develop new energy production pathways?'), ('Contextual/Relational Question', 'How does the study on the ADP/ATP carrier relate to other research on mitochondrial function and energy production pathways?'), ('User-Interactive Question', 'How might the discovery of alternative pathways for energy production impact various fields, such as medicine, bioengineering, or environmental sustainability?')]\n",
            "Storing data for file_id: be556138d3cc4cea9f89110c8cae3f9f\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/be556138d3cc4cea9f89110c8cae3f9f.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/be556138d3cc4cea9f89110c8cae3f9f.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/e572c6d677854443adc2481c01b7cf1d.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/e572c6d677854443adc2481c01b7cf1d.json\n",
            "Questions: [('Content-Based Question', 'What are the top five rubber exporting countries and their respective export values in 2021?'), ('Analytical Question', 'How does the market concentration of rubber exports, as indicated by the Shannon Entropy, impact the global rubber industry?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where Thailand and Indonesia significantly decrease their rubber production. How would this impact the global rubber market and the economies of other exporting countries?'), ('Contextual/Relational Question', 'How does the projected compound annual growth rate (CAGR) of the global rubber market between 2024 and 2032 relate to the challenges faced by rubber exporting countries in maintaining their dominance?'), ('User-Interactive Question', 'As a consumer, how do you think the growth in construction activities and the development of novel rubber products will affect the availability and pricing of rubber-based products in the future?')]\n",
            "Storing data for file_id: e572c6d677854443adc2481c01b7cf1d\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/e572c6d677854443adc2481c01b7cf1d.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/e572c6d677854443adc2481c01b7cf1d.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/b9b557c218a74e36a5894c7be8be79a3.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/b9b557c218a74e36a5894c7be8be79a3.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How does the preservation of isolated mitochondria impact research and therapy?'), ('Analytical Question', 'What are the challenges faced by current techniques in maintaining the viability and functionality of isolated mitochondria?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where you are tasked with developing a new preservation solution for isolated mitochondria. What factors would you consider in order to optimize its effectiveness?'), ('Contextual/Relational Question', 'How do nanotechnology and gene editing contribute to the potential of transplantation and preservation of isolated mitochondria?'), ('User-Interactive Question', 'Do you think standardization and quality control measures are important in the field of isolated mitochondria preservation? Why or why not?')]\n",
            "Storing data for file_id: b9b557c218a74e36a5894c7be8be79a3\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/b9b557c218a74e36a5894c7be8be79a3.json\n",
            "Questions: [('Content-Based Question', 'How does the technique of differential isopycnic density gradient centrifugation contribute to the study of isolated mitochondria?'), ('Analytical Question', 'What are the potential implications of the ability to keep isolated mitochondria active and coupled for approximately 1 hour on ice using the differential isopycnic density gradient centrifugation technique?'), ('Creative/Scenario-Based Question', 'Imagine a future where the storable preparation of mitochondria becomes widely available. How might this impact the field of clinical medicine and the treatment of mitochondrial-related diseases?'), ('Contextual/Relational Question', 'How does the ability to study isolated mitochondria using the differential isopycnic density gradient centrifugation technique enhance our understanding of mitochondrial function and dysfunction?'), ('User-Interactive Question', 'How do you think the differential isopycnic density gradient centrifugation technique could be further improved to extend the duration of active and coupled isolated mitochondria on ice?')]\n",
            "Storing data for file_id: b9b557c218a74e36a5894c7be8be79a3\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/b9b557c218a74e36a5894c7be8be79a3.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/b9b557c218a74e36a5894c7be8be79a3.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/545d51e83b5c4eeb8176c2abac159bf6.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/545d51e83b5c4eeb8176c2abac159bf6.json\n",
            "Questions: [('Content-Based Question', 'How does ATP contribute to cellular processes such as cellular differentiation, the cell cycle, and cell growth?'), ('Analytical Question', 'What are the challenges faced by researchers when studying mitochondrial function in cell-free environments, and how do they overcome them?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where there is a significant decrease in ATP production in mitochondria due to environmental changes. How might this impact calcium signaling and the regulation of apoptosis?'), ('Contextual/Relational Question', 'How do mitochondrial fusion, fission, and mtDNA integrity relate to bioenergetic functions, and how can mutations affect these processes?'), ('User-Interactive Question', 'In your opinion, what ethical and legal considerations should be taken into account when using human cells and tissues in mitochondrial research?')]\n",
            "Storing data for file_id: 545d51e83b5c4eeb8176c2abac159bf6\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/545d51e83b5c4eeb8176c2abac159bf6.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How does the process of tissue dissociation and differential filtration contribute to the isolation and purification of mitochondria for transplantation?'), ('Analytical Question', 'What are the potential challenges or limitations in isolating and purifying mitochondria for transplantation using tissue dissociation and differential filtration?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where the isolation and purification of mitochondria for transplantation using tissue dissociation and differential filtration is successful. How could this advance medical treatments or therapies?'), ('Contextual/Relational Question', 'How does the isolation and purification of mitochondria for transplantation using tissue dissociation and differential filtration relate to other methods of obtaining mitochondria for research or therapeutic purposes?'), ('User-Interactive Question', 'What are some ethical considerations that should be taken into account when using isolated and purified mitochondria for transplantation?')]\n",
            "Storing data for file_id: 545d51e83b5c4eeb8176c2abac159bf6\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/545d51e83b5c4eeb8176c2abac159bf6.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/545d51e83b5c4eeb8176c2abac159bf6.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/3be46da3b4e948eaa5e926bf04fd65c4.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/3be46da3b4e948eaa5e926bf04fd65c4.json\n",
            "Questions: [('Content-Based Question', 'How does microfluidic technology revolutionize the isolation of mitochondria?'), ('Analytical Question', 'What are the advantages of using microfluidic technology for mitochondrial isolation compared to traditional techniques?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where a researcher needs to isolate mitochondria from a large number of samples. How could microfluidic technology help in this situation?'), ('Contextual/Relational Question', 'How does the use of microfluidic devices in mitochondrial isolation preserve the physiological environment of the mitochondria?'), ('User-Interactive Question', 'Have you ever encountered a situation where high-throughput mitochondrial isolation would be beneficial? How do you think microfluidic technology could be applied in that scenario?')]\n",
            "Storing data for file_id: 3be46da3b4e948eaa5e926bf04fd65c4\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/3be46da3b4e948eaa5e926bf04fd65c4.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/3be46da3b4e948eaa5e926bf04fd65c4.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/c9e653a050724fe8bc0abc30d7e3a46c.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/c9e653a050724fe8bc0abc30d7e3a46c.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How does autologous mitochondrial transplantation differ from other transplantation methods in terms of immune response? '), ('Analytical Question', 'What evidence supports the claim that mitochondrial transplantation reduces inflammation and has cardio-protective effects?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where mitochondrial transplantation becomes a widely accepted therapy. How might this impact the field of organ transplantation and the treatment of immune-related disorders?'), ('Contextual/Relational Question', 'How does the immune response to transplanted mitochondria relate to the rejection of cardiac allografts?'), ('User-Interactive Question', 'What are some potential strategies to overcome the challenges faced by mitochondrial transfer/transplantation, particularly in addressing the immune response?')]\n",
            "Storing data for file_id: c9e653a050724fe8bc0abc30d7e3a46c\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/c9e653a050724fe8bc0abc30d7e3a46c.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How do the findings in this report contribute to our understanding of immune mechanisms and the safety of autologous transplantation?'), ('Analytical Question', 'What are the potential strategies that researchers can develop based on the findings in this report to minimize immune response to mitochondrial transplantation?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where immune response to mitochondrial transplantation is completely eliminated. How would this impact the therapeutic potential of the transplantation?'), ('Contextual/Relational Question', 'How do the challenges in allogeneic transplantation discussed in this report relate to the safety and efficacy of autologous transplantation?'), ('User-Interactive Question', 'How would you approach the development of strategies to minimize immune response to mitochondrial transplantation based on the findings in this report?')]\n",
            "Storing data for file_id: c9e653a050724fe8bc0abc30d7e3a46c\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/c9e653a050724fe8bc0abc30d7e3a46c.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/c9e653a050724fe8bc0abc30d7e3a46c.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/0e1ea044ac594b668db126787ad7f47d.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/0e1ea044ac594b668db126787ad7f47d.json\n",
            "Questions: [('Content-Based Question', 'How do mitochondria contribute to cellular processes, and why is it important to study their structure and function?'), ('Analytical Question', 'What are the different techniques used to assess the purity, integrity, and function of isolated mitochondria?'), ('Creative/Scenario-Based Question', 'Imagine you are a scientist studying mitochondria. What innovative methods could you develop to optimize the isolation process and minimize contamination?'), ('Contextual/Relational Question', 'How do NAD+, NADH, organic acids, matrix dehydrogenases, and oxidative phosphorylation capacity relate to mitochondrial function and ATP production?'), ('User-Interactive Question', 'Have you ever encountered challenges in isolating mitochondria or studying their function? How did you overcome them, and what lessons did you learn from the process?')]\n",
            "Storing data for file_id: 0e1ea044ac594b668db126787ad7f47d\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/0e1ea044ac594b668db126787ad7f47d.json\n",
            "Questions: [('Content-Based Question', 'What are the different techniques used to evaluate mitochondrial preparations and understand their function in health and disease?'), ('Analytical Question', 'How can researchers ensure the reliability of their findings when evaluating mitochondrial preparations?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where a researcher is studying mitochondrial function in a specific disease model. What considerations should they keep in mind when selecting and optimizing isolation methods?'), ('Contextual/Relational Question', 'How do the articles mentioned in the summary contribute to the overall understanding of mitochondrial quality assessment and integrity?'), ('User-Interactive Question', 'Have you ever encountered challenges in evaluating mitochondrial preparations? How did you address them, and what lessons did you learn from the process?')]\n",
            "Storing data for file_id: 0e1ea044ac594b668db126787ad7f47d\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/0e1ea044ac594b668db126787ad7f47d.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/0e1ea044ac594b668db126787ad7f47d.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/7a3222732a35428ab90990e3fae6f0d4.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/7a3222732a35428ab90990e3fae6f0d4.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'What are the main techniques involved in the protocol for extracting functionally competent mitochondria?'), ('Analytical Question', 'How does the protocol ensure the functional competence of the extracted mitochondria?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where the protocol for extracting functionally competent mitochondria is not followed accurately. What potential consequences could arise?'), ('Contextual/Relational Question', 'How does the MitoTag approach contribute to the isolation of specific cell types within tissues during the extraction of mitochondria?'), ('User-Interactive Question', 'Have you ever encountered a situation where the isolation of high-quality mitochondria was crucial for your research or work? If so, how did you ensure the quality and functionality of the extracted mitochondria?')]\n",
            "Storing data for file_id: 7a3222732a35428ab90990e3fae6f0d4\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/7a3222732a35428ab90990e3fae6f0d4.json\n",
            "Questions: [('Content-Based Question', 'How does tissue dissociation and differential filtration enable rapid isolation and purification of mitochondria for transplantation?'), ('Analytical Question', 'What are the implications of high throughput microplate respiratory measurements using minimal quantities of isolated mitochondria?'), ('Creative/Scenario-Based Question', 'Imagine you are a researcher working with cerebral organoids. How could the optimization of differential filtration-based mitochondrial isolation benefit your work?'), ('Contextual/Relational Question', 'How does the protocol for isolating mitochondria from cultured cells and extracting differently located mitochondrial proteins contribute to the field of stem cell research?'), ('User-Interactive Question', 'How do you think the novel approach to quantify mitochondrial content and intrinsic bioenergetic efficiency across organs could impact transplantation procedures?')]\n",
            "Storing data for file_id: 7a3222732a35428ab90990e3fae6f0d4\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/7a3222732a35428ab90990e3fae6f0d4.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/7a3222732a35428ab90990e3fae6f0d4.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/f8e2a08ffd5149d6a128d23d772abc38.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/f8e2a08ffd5149d6a128d23d772abc38.json\n",
            "Questions: [('Content-Based Question', 'What are the challenges associated with traditional methods of isolating mitochondria while preserving their structure and function?'), ('Analytical Question', 'How do emerging methods such as nitrogen cavitation, affinity purification, filtration-based methods, and density gradient centrifugation address the limitations of traditional methods in isolating mitochondria?'), ('Creative/Scenario-Based Question', \"Imagine you are a researcher working on developing a new method for isolating mitochondria. How would you incorporate the use of anti-TOM22 magnetic beads for affinity purification, like Franko's method?\"), ('Contextual/Relational Question', 'How do the research articles in \"Frontiers in Cardiovascular Medicine\" and \"Frontiers in Physiology\" provide valuable insights into the challenges and advancements in mitochondrial isolation techniques for clinical applications?'), ('User-Interactive Question', 'Have you ever encountered challenges in isolating mitochondria while preserving their structure and function? How did you overcome those challenges, or what improvements would you suggest for existing methods?')]\n",
            "Storing data for file_id: f8e2a08ffd5149d6a128d23d772abc38\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/f8e2a08ffd5149d6a128d23d772abc38.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'What is the main focus of the study by Cowan, D.B., and McCully, J.D. on rapid isolation and purification of mitochondria for transplantation?'), ('Analytical Question', 'How does the method proposed by Cowan, D.B., and McCully, J.D. for isolating and purifying mitochondria differ from existing techniques?'), ('Creative/Scenario-Based Question', 'Imagine a future where mitochondria transplantation becomes a common medical procedure. What potential benefits and challenges might arise from this advancement?'), ('Contextual/Relational Question', 'How does the research by Cowan, D.B., and McCully, J.D. on mitochondria transplantation contribute to the broader field of regenerative medicine?'), ('User-Interactive Question', 'If you were a researcher in the field, how would you design an experiment to test the effectiveness of mitochondria transplantation in improving cellular function?')]\n",
            "Storing data for file_id: f8e2a08ffd5149d6a128d23d772abc38\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/f8e2a08ffd5149d6a128d23d772abc38.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/f8e2a08ffd5149d6a128d23d772abc38.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6a666e95c0674c2d84596d6065a4dad6.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6a666e95c0674c2d84596d6065a4dad6.json\n",
            "Questions: [('Content-Based Question', 'How do the latest advancements in gentle mitochondrial isolation contribute to our understanding of mitochondrial biology and the development of novel therapies?'), ('Analytical Question', 'What are the main challenges that need to be addressed in the field of gentle mitochondrial isolation, and how can they be overcome to ensure reproducibility, mitochondrial integrity, and yield?'), ('Creative/Scenario-Based Question', 'Imagine a world where gentle mitochondrial isolation techniques have been perfected. How might this revolutionize the treatment of chronic and rare diseases associated with mitochondrial dysfunction?'), ('Contextual/Relational Question', 'How does stem cell research and therapy intersect with the functional assessment and optimization of gentle mitochondrial isolation techniques, and what role does it play in advancing mitochondrial medicine?'), ('User-Interactive Question', 'What are some potential applications of differential filtration-based mitochondrial isolation and mitochondrial transplants to cerebral organoids in the field of regenerative medicine, and how might they impact patient outcomes?')]\n",
            "Storing data for file_id: 6a666e95c0674c2d84596d6065a4dad6\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6a666e95c0674c2d84596d6065a4dad6.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6a666e95c0674c2d84596d6065a4dad6.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/77c098f492e841019170e5eb977a63b9.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/77c098f492e841019170e5eb977a63b9.json\n",
            "Questions: [('Content-Based Question', 'How does the concentration of natural rubber production and export in a few countries impact the global economy?'), ('Analytical Question', 'What are the main challenges faced by the rubber industry, and how do they affect its growth and sustainability?'), ('Creative/Scenario-Based Question', 'Imagine a world where synthetic alternatives completely replace natural rubber. How would this impact the automotive industry and medical devices sector?'), ('Contextual/Relational Question', 'How do the rising industrial and infrastructure development in developing economies contribute to the growth of the rubber industry?'), ('User-Interactive Question', 'As a consumer, how would you prioritize environmental concerns versus the price and availability of natural rubber products?')]\n",
            "Storing data for file_id: 77c098f492e841019170e5eb977a63b9\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/77c098f492e841019170e5eb977a63b9.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/77c098f492e841019170e5eb977a63b9.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/7fe118e00411442ea91c951d5fa62338.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/7fe118e00411442ea91c951d5fa62338.json\n",
            "Error in extracting knowledge: Error code: 400 - {'error': {'message': \"This model's maximum context length is 4097 tokens. However, your messages resulted in 4746 tokens. Please reduce the length of the messages.\", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}\n",
            "Error in extracting dirty knowledge: Error in extracting knowledge\n",
            " Trying again\n",
            "Error in extracting knowledge: Error code: 400 - {'error': {'message': \"This model's maximum context length is 4097 tokens. However, your messages resulted in 4746 tokens. Please reduce the length of the messages.\", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}\n",
            "Error summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/7fe118e00411442ea91c951d5fa62338.md: Error in extracting knowledge\n",
            "Error summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/7fe118e00411442ea91c951d5fa62338.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/dd803c16e6e14ac6b279b29e3b40c229.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/dd803c16e6e14ac6b279b29e3b40c229.json\n",
            "Error in extracting knowledge: Error code: 400 - {'error': {'message': \"This model's maximum context length is 4097 tokens. However, your messages resulted in 4770 tokens. Please reduce the length of the messages.\", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}\n",
            "Error in extracting dirty knowledge: Error in extracting knowledge\n",
            " Trying again\n",
            "Error in extracting knowledge: Error code: 400 - {'error': {'message': \"This model's maximum context length is 4097 tokens. However, your messages resulted in 4770 tokens. Please reduce the length of the messages.\", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}\n",
            "Error summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/dd803c16e6e14ac6b279b29e3b40c229.md: Error in extracting knowledge\n",
            "Error summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/dd803c16e6e14ac6b279b29e3b40c229.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/848c4daf729c4d8f8f6ba2b1dc07d7dc.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/848c4daf729c4d8f8f6ba2b1dc07d7dc.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How have advancements in methods contributed to the isolation of mitochondria from plant cells?'), ('Analytical Question', 'What are the key steps involved in the process of isolating mitochondria from plant cells?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where a new method for isolating mitochondria from plant cells is discovered. How might this impact plant mitochondrial research?'), ('Contextual/Relational Question', 'How does the removal of chloroplasts and phenolic compounds pose challenges in the isolation of mitochondria from plant cells?'), ('User-Interactive Question', 'Have you ever encountered any difficulties in isolating mitochondria from plant cells? If so, how did you overcome them?')]\n",
            "Storing data for file_id: 848c4daf729c4d8f8f6ba2b1dc07d7dc\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/848c4daf729c4d8f8f6ba2b1dc07d7dc.json\n",
            "Questions: [('Content-Based Question', 'How are plant mitochondria isolated from cell culture using isopycnic centrifugation in density gradients of Percoll?'), ('Analytical Question', 'What are the key techniques used in the field of plant mitochondrial research, specifically in terms of sample preparation and fractionation methods?'), ('Creative/Scenario-Based Question', 'Imagine a new method for isolating plant mitochondria that does not involve density gradients. How might this alternative method impact the field of mitochondrial research?'), ('Contextual/Relational Question', 'How do the techniques described in the publications \"Isolation of Membranes and Organelles from Plant Cells\" and \"An improved method with a wider applicability to isolate plant mitochondria for mtDNA extraction\" contribute to our understanding of plant cell biology and molecular biology?'), ('User-Interactive Question', 'As a researcher in the field of mitochondrial research, how would you modify the existing methods for isolating plant mitochondria to improve efficiency and accuracy?')]\n",
            "Storing data for file_id: 848c4daf729c4d8f8f6ba2b1dc07d7dc\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/848c4daf729c4d8f8f6ba2b1dc07d7dc.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/848c4daf729c4d8f8f6ba2b1dc07d7dc.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6a6833e6bf9b4723b80b992cfd1c3c09.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6a6833e6bf9b4723b80b992cfd1c3c09.json\n",
            "Questions: [('Content-Based Question', 'How do recent developments in mitochondrial genome engineering, such as the use of nucleases and base editors, contribute to enhancing mitochondrial autonomy and treating hereditary mitochondrial diseases?'), ('Analytical Question', 'What are the potential challenges and opportunities presented by advances in mitochondrial genome engineering for treating mitochondrial diseases?'), ('Creative/Scenario-Based Question', 'Imagine a future where mitochondrial DNA replacement becomes a widely available treatment for mitochondrial diseases. How might this impact individuals with hereditary mitochondrial diseases and their quality of life?'), ('Contextual/Relational Question', 'How does the coordination between nuclear DNA and mitochondrial DNA contribute to maintaining mitochondrial autonomy, and what are the implications for understanding and treating mitochondrial diseases?'), ('User-Interactive Question', 'Based on the information provided, what are some potential applications of mitochondrial genome engineering in studying mitochondrial diseases and pathologies in different cellular contexts and stress conditions?')]\n",
            "Storing data for file_id: 6a6833e6bf9b4723b80b992cfd1c3c09\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6a6833e6bf9b4723b80b992cfd1c3c09.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How does the current state of mitochondrial genome engineering contribute to enhancing autonomy?'), ('Analytical Question', 'What are the main challenges faced in the field of mitochondrial genome engineering, and how can they be overcome to increase mitochondrial autonomy?'), ('Creative/Scenario-Based Question', 'Imagine a future where mitochondrial genome engineering has advanced significantly. How might increased mitochondrial autonomy impact the functioning of cells and organisms?'), ('Contextual/Relational Question', 'How does the interplay between nuclear and mitochondrial genomes influence the potential for enhancing autonomy through mitochondrial genome engineering?'), ('User-Interactive Question', 'How might the concept of increasing mitochondrial autonomy through genome engineering be received by different stakeholders, such as scientists, ethicists, and the general public?')]\n",
            "Storing data for file_id: 6a6833e6bf9b4723b80b992cfd1c3c09\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6a6833e6bf9b4723b80b992cfd1c3c09.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6a6833e6bf9b4723b80b992cfd1c3c09.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/15c0aba9ec7a4a2d99fd2a35dad345ed.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/15c0aba9ec7a4a2d99fd2a35dad345ed.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How does the Clark-type oxygen electrode measure oxygen levels in liquids?'), ('Analytical Question', 'What are the key components required to construct a DIY Clark-type oxygen electrode?'), ('Creative/Scenario-Based Question', 'Imagine you are conducting an experiment to measure oxygen levels in a lake. How would you use the DIY Clark-type oxygen electrode to obtain accurate readings?'), ('Contextual/Relational Question', 'How does the Clark-type oxygen electrode compare to other methods of measuring oxygen levels in liquids, in terms of accuracy and practicality?'), ('User-Interactive Question', 'Have you ever encountered a situation where measuring oxygen levels in a liquid would be beneficial? How do you think a DIY Clark-type oxygen electrode could be useful in that scenario?')]\n",
            "Storing data for file_id: 15c0aba9ec7a4a2d99fd2a35dad345ed\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/15c0aba9ec7a4a2d99fd2a35dad345ed.json\n",
            "Questions: [('Content-Based Question', 'How does the low temperature co-fired ceramic (LTCC) based microfluidic Clark-type oxygen sensor combine optical oxygen sensing and Clark electrode technology?'), ('Analytical Question', 'What are the advantages of using ceramic-based sensors and microfluidic technology for precise and efficient oxygen sensing in biosensing applications?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where the LTCC microfluidic Clark-type oxygen sensor is used in a medical setting. How could this sensor improve patient care and monitoring?'), ('Contextual/Relational Question', 'How does the real-time oxygen sensing capability of the LTCC microfluidic Clark-type oxygen sensor compare to other oxygen sensing technologies in terms of accuracy and continuous monitoring?'), ('User-Interactive Question', 'How would you envision the integration of the LTCC microfluidic Clark-type oxygen sensor into existing biosensing devices or systems? What potential challenges or benefits do you foresee?')]\n",
            "Storing data for file_id: 15c0aba9ec7a4a2d99fd2a35dad345ed\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/15c0aba9ec7a4a2d99fd2a35dad345ed.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/15c0aba9ec7a4a2d99fd2a35dad345ed.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/bcf05b9d20cc4a4db92660c566570f0c.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/bcf05b9d20cc4a4db92660c566570f0c.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How can the use of Wikidata APIs enhance data accessibility, integration, and quality enhancement in data applications?'), ('Analytical Question', 'What are the potential benefits of constructing knowledge graphs using Wikidata APIs for contextual understanding?'), ('Creative/Scenario-Based Question', 'Imagine you are developing a new data application. How could you leverage Wikidata APIs to foster creativity and innovation in your application?'), ('Contextual/Relational Question', 'How can connecting Wikidata APIs with other APIs enhance the user experience and enable real-time AI integration, leading to cost savings?'), ('User-Interactive Question', 'How might the challenges of data privacy, security, and maintaining high-quality data be addressed when utilizing Wikidata APIs in data-driven technology?')]\n",
            "Storing data for file_id: bcf05b9d20cc4a4db92660c566570f0c\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/bcf05b9d20cc4a4db92660c566570f0c.json\n",
            "Questions: [('Content-Based Question', 'How important is efficient data infrastructure and access for the success of AI-centric enterprises?'), ('Analytical Question', 'How can the development plan outlined in the Wikidata page on \"Development plan\" contribute to achieving AI-centric enterprise growth?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where an AI-centric enterprise lacks efficient data infrastructure and access. How would this impact its growth and success?'), ('Contextual/Relational Question', 'How do the predictions for data-driven growth in AI-centric enterprises in 2024, mentioned in the VentureBeat article, relate to the development plan outlined in the Wikidata page on \"Development plan\"?'), ('User-Interactive Question', 'In your opinion, what are some potential challenges that AI-centric enterprises may face in terms of data infrastructure and access, and how can they overcome them?')]\n",
            "Storing data for file_id: bcf05b9d20cc4a4db92660c566570f0c\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/bcf05b9d20cc4a4db92660c566570f0c.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/bcf05b9d20cc4a4db92660c566570f0c.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/e6a8e415bc984129a6b4a905fd3e2e26.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/e6a8e415bc984129a6b4a905fd3e2e26.json\n",
            "Questions: [('Content-Based Question', 'How does mitochondrial transfer contribute to protecting against ischemia-reperfusion injury in the myocardium?'), ('Analytical Question', 'What are the technical and biological challenges in isolating and evaluating the biochemical and functional properties of transferred mitochondria?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where autologous mitochondrial transplants become widely available. How could this technique be used to treat other mitochondrial diseases besides MERRF and MELAS?'), ('Contextual/Relational Question', 'How does the need for analyzing mitochondrial vitality at the single-cell level relate to the challenges posed by tissue and cellular heterogeneity?'), ('User-Interactive Question', 'In your opinion, what are some potential interdisciplinary collaborations that could enhance the study of mitochondrial mechanisms and their clinical applications?')]\n",
            "Storing data for file_id: e6a8e415bc984129a6b4a905fd3e2e26\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/e6a8e415bc984129a6b4a905fd3e2e26.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/e6a8e415bc984129a6b4a905fd3e2e26.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/97c3de33990e4fc6b7d4b70866e4c93c.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/97c3de33990e4fc6b7d4b70866e4c93c.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'What are some of the key ethical considerations that need to be addressed when it comes to mitochondrial manipulation in therapy?'), ('Analytical Question', 'How can international consensus and controlled clinical trials contribute to the safe and effective implementation of mitochondrial replacement techniques?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where gene therapy for mitochondrial diseases becomes widely accessible. What are some potential challenges that might arise in terms of ensuring the unique qualities of the mitochondrial organelle are considered?'), ('Contextual/Relational Question', 'How do the current advances in gene therapy for mitochondrial diseases relate to the safety concerns that need to be addressed in mitochondrial transfer and transplantation?'), ('User-Interactive Question', 'In your opinion, what would be the most effective approach to managing primary mitochondrial disease, taking into consideration the use of mitochondrial cocktails and other potential treatments?')]\n",
            "Storing data for file_id: 97c3de33990e4fc6b7d4b70866e4c93c\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/97c3de33990e4fc6b7d4b70866e4c93c.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'What are some of the ethical guidelines and safety measures that comprehensive sources establish for mitochondrial manipulation in therapy?'), ('Analytical Question', 'How do sources explore the clinical challenges associated with mitochondrial genetics and advancements in gene therapy?'), ('Creative/Scenario-Based Question', 'Imagine a future where mitochondrial therapies have been widely developed and applied in biotechnology. What are some potential benefits and risks that need to be considered?'), ('Contextual/Relational Question', 'How do the published sources contribute to the responsible development and application of mitochondrial therapies in biotechnology?'), ('User-Interactive Question', 'What are your thoughts on the importance of ethical guidelines and safety measures in the development and application of mitochondrial therapies in therapy? How do you think these measures can ensure responsible practices?')]\n",
            "Storing data for file_id: 97c3de33990e4fc6b7d4b70866e4c93c\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/97c3de33990e4fc6b7d4b70866e4c93c.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/97c3de33990e4fc6b7d4b70866e4c93c.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/62ead12d6fdc4e8faef06d6c4ef1e838.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/62ead12d6fdc4e8faef06d6c4ef1e838.json\n",
            "Questions: [('Content-Based Question', 'How does Gephi contribute to the exploration and visualization of complex relationships in different domains?'), ('Analytical Question', 'What are the basic steps involved in using Gephi for knowledge graph visualization?'), ('Creative/Scenario-Based Question', \"Imagine you are working on a research project that involves analyzing temporal networks. How can Gephi's dynamic graph feature be useful in visualizing and understanding the changes in relationships over time?\"), ('Contextual/Relational Question', \"How does Gephi's modular architecture and extensive customization through plugins contribute to its versatility as a graph analysis and visualization tool?\"), ('User-Interactive Question', 'Have you ever used Gephi or any other network analysis and visualization software? If so, what were your experiences and challenges in using such tools for exploring complex relationships?')]\n",
            "Storing data for file_id: 62ead12d6fdc4e8faef06d6c4ef1e838\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/62ead12d6fdc4e8faef06d6c4ef1e838.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/62ead12d6fdc4e8faef06d6c4ef1e838.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/9a4469c247bc48ac84dda8b05a6e350d.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/9a4469c247bc48ac84dda8b05a6e350d.json\n",
            "Error in extracting knowledge: Error code: 400 - {'error': {'message': \"This model's maximum context length is 4097 tokens. However, your messages resulted in 4342 tokens. Please reduce the length of the messages.\", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}\n",
            "Error in extracting dirty knowledge: Error in extracting knowledge\n",
            " Trying again\n",
            "Error in extracting knowledge: Error code: 400 - {'error': {'message': \"This model's maximum context length is 4097 tokens. However, your messages resulted in 4342 tokens. Please reduce the length of the messages.\", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}\n",
            "Error summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/9a4469c247bc48ac84dda8b05a6e350d.md: Error in extracting knowledge\n",
            "Error summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/9a4469c247bc48ac84dda8b05a6e350d.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/2c8c857b12bf4e6bb35ddb1733b20383.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/2c8c857b12bf4e6bb35ddb1733b20383.json\n",
            "Error in extracting knowledge: Error code: 400 - {'error': {'message': \"This model's maximum context length is 4097 tokens. However, your messages resulted in 5699 tokens. Please reduce the length of the messages.\", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}\n",
            "Error in extracting combined knowledge: Error in extracting knowledge\n",
            " Trying again\n",
            "Error in extracting knowledge: Error code: 400 - {'error': {'message': \"This model's maximum context length is 4097 tokens. However, your messages resulted in 5699 tokens. Please reduce the length of the messages.\", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}\n",
            "Error summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/2c8c857b12bf4e6bb35ddb1733b20383.md: Error in extracting knowledge\n",
            "Error summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/2c8c857b12bf4e6bb35ddb1733b20383.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/4b4b1e3e7adf40b59d5ef92031e3b2bf.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/4b4b1e3e7adf40b59d5ef92031e3b2bf.json\n",
            "Questions: [('Content-Based Question', 'How does the comparative study contribute to our understanding of mitochondrial research and its importance in understanding cellular function and disease?'), ('Analytical Question', 'What are the key factors that determine the efficiency and purity of mitochondrial extraction methods, and how do these factors impact the quality of mitochondrial fractions?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where a new mitochondrial extraction method is developed that promises higher efficiency and purity. How might this impact the field of mitochondrial research and our ability to study cellular function and disease?'), ('Contextual/Relational Question', 'How do the macroscale, microscale, and nanoscale techniques differ in terms of the level of analysis they provide and their suitability for investigating mitochondrial heterogeneity and dysfunction?'), ('User-Interactive Question', 'Have you ever encountered any challenges or limitations when working with mitochondrial extraction methods? How do you think ongoing developments in the field can address these challenges and enhance our understanding of mitochondrial biology?')]\n",
            "Storing data for file_id: 4b4b1e3e7adf40b59d5ef92031e3b2bf\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/4b4b1e3e7adf40b59d5ef92031e3b2bf.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/4b4b1e3e7adf40b59d5ef92031e3b2bf.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6d3c1580026546528356d583a9a268b9.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6d3c1580026546528356d583a9a268b9.json\n",
            "Questions: [('Content-Based Question', 'How does expanding the mitochondrial genome and enabling independent DNA replication contribute to enhancing the autonomy of mitochondria?'), ('Analytical Question', 'What are the technical challenges involved in engineering transcription, translation, and protein import systems within mitochondria to reduce their dependence on nuclear-encoded proteins?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where mitochondria achieve complete autonomy and no longer rely on nuclear-encoded proteins. How might this impact the overall functioning of cells and organisms?'), ('Contextual/Relational Question', 'How do considerations of off-target effects, delivery challenges, and technical and ethical aspects influence the exploration of genetic modifications to enhance the autonomy of mitochondria?'), ('User-Interactive Question', 'What are some potential evolutionary factors that might have influenced the current level of autonomy of mitochondria? How might these factors be relevant in the pursuit of enhancing mitochondrial autonomy?')]\n",
            "Storing data for file_id: 6d3c1580026546528356d583a9a268b9\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6d3c1580026546528356d583a9a268b9.json\n",
            "Questions: [('Content-Based Question', \"How can scientific innovation in manipulating cellular life's fundamental aspect be balanced with ethical concerns?\"), ('Analytical Question', 'What are the potential off-target effects of mitochondrial gene editing tools, and how should they be considered in research?'), ('Creative/Scenario-Based Question', 'Imagine a scenario where a breakthrough in mitochondrial gene editing allows for the correction of genetic diseases. How can we ensure that this technology is used responsibly and ethically?'), ('Contextual/Relational Question', \"How does the need to balance scientific innovation and ethical concerns in manipulating cellular life's fundamental aspect relate to other fields of research or technological advancements?\"), ('User-Interactive Question', 'What are your thoughts on the ethical considerations surrounding mitochondrial gene editing? How do you think society should navigate this delicate balance between scientific progress and ethical implications?')]\n",
            "Storing data for file_id: 6d3c1580026546528356d583a9a268b9\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6d3c1580026546528356d583a9a268b9.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/6d3c1580026546528356d583a9a268b9.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/f885211ff5164954a13d99ef9cfa039f.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/f885211ff5164954a13d99ef9cfa039f.json\n",
            "Error in extracting knowledge: Error code: 400 - {'error': {'message': \"This model's maximum context length is 4097 tokens. However, your messages resulted in 5287 tokens. Please reduce the length of the messages.\", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}\n",
            "Error in extracting dirty knowledge: Error in extracting knowledge\n",
            " Trying again\n",
            "Error in extracting knowledge: Error code: 400 - {'error': {'message': \"This model's maximum context length is 4097 tokens. However, your messages resulted in 5287 tokens. Please reduce the length of the messages.\", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}\n",
            "Error summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/f885211ff5164954a13d99ef9cfa039f.md: Error in extracting knowledge\n",
            "Error summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/f885211ff5164954a13d99ef9cfa039f.md\n",
            "Summarizing /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/95ebab163f43470bb1b3e59df5765140.md\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/95ebab163f43470bb1b3e59df5765140.json\n",
            "Error in summarizing article: Text cannot be empty\n",
            " Using last summary\n",
            "Questions: [('Content-Based Question', 'How does the SPR protocol for isolating plant mitochondria contribute to molecular biology research?'), ('Analytical Question', 'What are the key steps involved in the SPR protocol for isolating plant mitochondria, and how do they ensure the purity and integrity of the isolated mitochondria?'), ('Creative/Scenario-Based Question', 'Imagine you are a researcher studying the stress tolerance of different plant species. How could the SPR protocol for isolating plant mitochondria be used to investigate the role of mitochondria in plant stress responses?'), ('Contextual/Relational Question', 'How does the SPR protocol for isolating plant mitochondria compare to other methods used in molecular biology research to study mitochondrial genetics?'), ('User-Interactive Question', 'If you were to optimize the SPR protocol for isolating mitochondria from a specific plant species, what factors would you consider and how would you modify the protocol accordingly?')]\n",
            "Storing data for file_id: 95ebab163f43470bb1b3e59df5765140\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/95ebab163f43470bb1b3e59df5765140.json\n",
            "Questions: [('Content-Based Question', \"How does the researcher's improved method for extracting rice mitochondria contribute to advancements in mitochondrial research in plant species?\"), ('Analytical Question', \"What are the key steps in the researcher's method for extracting rice mitochondria, and how do they enhance efficiency and purity compared to previous methods?\"), ('Creative/Scenario-Based Question', \"Imagine a scenario where the researcher's method for extracting rice mitochondria is widely adopted. How might this impact future studies on mitochondrial research in other plant species?\"), ('Contextual/Relational Question', \"How does the article on mitochondrial research in plant species provide a broader context for understanding the significance of the researcher's improved method for extracting rice mitochondria?\"), ('User-Interactive Question', \"As a researcher in the field of plant biology, how might you incorporate the researcher's method for extracting rice mitochondria into your own studies on mitochondrial research in different plant species?\")]\n",
            "Storing data for file_id: 95ebab163f43470bb1b3e59df5765140\n",
            "Successfully saved data to /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/95ebab163f43470bb1b3e59df5765140.json\n",
            "Successfully summarized /Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/95ebab163f43470bb1b3e59df5765140.md\n"
          ]
        }
      ],
      "source": [
        "!pip install pydantic\n",
        "!pip install instructor\n",
        "!pip install openai\n",
        "from datetime import datetime\n",
        "from pydantic import BaseModel, Field\n",
        "\n",
        "import os\n",
        "import json\n",
        "import instructor\n",
        "from openai import OpenAI\n",
        "\n",
        "import re\n",
        "from typing import List\n",
        "\n",
        "\n",
        "#OUTPUT_FOLDER = \"/home/epas/Programming/ResearchAgentSwarm/Literature_Review/gpt_researcher_outputs/\" \n",
        "OUTPUT_FOLDER = \"/Users/tomriddle1/Documents/GitHub/gpt-researcher/outputs/\"\n",
        "#OUTPUT_FOLDER = \"gpt_researcher_outputs/\"\n",
        "#OUTPUT_FOLDER = \"Literature_Review/gpt_researcher_outputs/\"\n",
        "SUMMARY_JSON = \"summaries.json\"\n",
        "\n",
        "api_key = \"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\"\n",
        "\n",
        "# Optionally set the environment variable (if needed elsewhere)\n",
        "os.environ['OPENAI_API_KEY'] = api_key\n",
        "\n",
        "# Enum for prompt types\n",
        "    \n",
        "class SummaryStore:\n",
        "    def __init__(self, file_id): \n",
        "        self.file_id = file_id\n",
        "        self.file_path = f\"{OUTPUT_FOLDER}{file_id}.json\"\n",
        "        self._create_file_if_not_exists()\n",
        "\n",
        "    def _create_file_if_not_exists(self):\n",
        "        if not os.path.exists(self.file_path):\n",
        "            # Initialize with empty data\n",
        "            empty_data = [] \n",
        "            self._save(empty_data)\n",
        "    \n",
        "    def store(self, summary, clean_entities,dirty_entities, file_id, article, references, topic, hypothetical_questions, knowledge):\n",
        "        data = { \n",
        "            \"file_id\": file_id,\n",
        "            \"article\": article,\n",
        "            \"summary\": summary,\n",
        "            \"clean_entities\": clean_entities,\n",
        "            \"dirty_entities\": dirty_entities,\n",
        "            \"references\": references,\n",
        "            \"topics\": topic,\n",
        "            \"hypothetical_questions\": hypothetical_questions,\n",
        "            \"knowledge_triplets\": knowledge,\n",
        "            \"timestamp\": datetime.now().isoformat()\n",
        "        }\n",
        "\n",
        "        existing_data = self.load()\n",
        "        existing_data.append(data)\n",
        "        print(f\"Storing data for file_id: {file_id}\")  # Log storing action\n",
        "        self._save(existing_data)\n",
        "\n",
        "    def load(self):\n",
        "        if os.path.exists(self.file_path):\n",
        "            with open(self.file_path, \"r\") as f:\n",
        "                return json.load(f)\n",
        "        else:\n",
        "            return []\n",
        "\n",
        "    def _save(self, content):\n",
        "        try:\n",
        "            with open(self.file_path, \"w\") as f:\n",
        "                json.dump(content, f)\n",
        "            print(f\"Successfully saved data to {self.file_path}\")  # Log success message\n",
        "        except Exception as e:\n",
        "            print(f\"Error saving data to {self.file_path}: {e}\")  # Log error message\n",
        "\n",
        "    \n",
        "\n",
        "def count_words(text):\n",
        "    return len(text.split())\n",
        "\n",
        "def build_system_prompt(prompt_type: str):\n",
        "    # read from file \"entity_dense_prompt.md\"\n",
        "    if prompt_type == \"Enitity Dense\":\n",
        "        with open(\"entity_dense_prompt.md\", \"r\") as f:\n",
        "            system_prompt = f.read()\n",
        "    if prompt_type == \"SPR\":\n",
        "        with open(\"sparse_prime_representation.md\", \"r\") as f:\n",
        "            system_prompt = f.read()\n",
        "    if prompt_type == \"Get Entities\":\n",
        "        with open(\"get_entities.md\", \"r\") as f:\n",
        "            system_prompt = f.read()\n",
        "    if prompt_type == \"Get Topic\":\n",
        "        with open(\"get_topic.md\", \"r\") as f:\n",
        "            system_prompt = f.read()\n",
        "    if prompt_type == \"Get Hypothetical Questions\":\n",
        "        with open(\"get_hypothetical_questions.md\", \"r\") as f:\n",
        "            system_prompt = f.read()\n",
        "    if prompt_type == \"Get Knowledge\":\n",
        "        with open(\"get_knowlege_graph_triples.md\", \"r\") as f:\n",
        "            system_prompt = f.read()\n",
        "    return f\"{system_prompt}\"\n",
        "\n",
        "def parse_response(response):\n",
        "\n",
        "    # Get the text content from the single completion \n",
        "    completion = response.choices[0]\n",
        "    text = completion.message.content\n",
        "\n",
        "    # Remove unnecessary newlines and whitespace    \n",
        "    text = text.strip()  \n",
        "\n",
        "    # Could add additional parsing logic here \n",
        "\n",
        "    return text\n",
        "\n",
        "\n",
        "def generate_summary(text: str, summary_type: str, model: str = \"gpt-3.5-turbo-0613\", temp: float = 0.45, max_tokens: int = 800 ):\n",
        "    client = instructor.patch(OpenAI(api_key=api_key))\n",
        "    if not text:\n",
        "        raise ValueError(\"Text cannot be empty\")\n",
        "\n",
        "    if temp < 0 or temp > 1:\n",
        "       raise ValueError(\"Temperature should be between 0 and 1\")\n",
        "    \n",
        "    try: \n",
        "        # summarization code\n",
        "        if summary_type == \"Entity Dense\":\n",
        "            #print(f\"System Prompt: {build_system_prompt(prompt_type='Enitity Dense')}\")\n",
        "            response = client.chat.completions.create(\n",
        "                model=model,\n",
        "                temperature=temp,\n",
        "                max_tokens=max_tokens,\n",
        "                max_retries=3,\n",
        "                messages=[\n",
        "                    {\"role\": \"system\", \"content\": build_system_prompt(prompt_type=\"Enitity Dense\")},\n",
        "                    {\"role\": \"user\", \"content\": text}\n",
        "                ],\n",
        "            )\n",
        "        if summary_type == \"SPR\":\n",
        "            #print(f\"System Prompt: {build_system_prompt(prompt_type='SPR')}\")\n",
        "            response = client.chat.completions.create(\n",
        "                model=model,\n",
        "                temperature=temp,\n",
        "                max_tokens=max_tokens,\n",
        "                max_retries=3,\n",
        "                messages=[\n",
        "                    {\"role\": \"system\", \"content\": build_system_prompt(prompt_type=\"SPR\")},\n",
        "                    {\"role\": \"user\", \"content\": text}\n",
        "                ],\n",
        "            )\n",
        "        summary = parse_response(response)\n",
        "    except Exception as e:\n",
        "        print(f\"Error in summarizing article: {e}\\n Occured in generate_summary function\")\n",
        "        # Break out of the loop if there is an error\n",
        "        return None\n",
        "    \n",
        "    if not summary:\n",
        "        raise RuntimeError(\"Summary generation failed\")\n",
        "\n",
        "    return summary\n",
        "\n",
        "\n",
        "def get_entity_dense_sumary(article, initial_summary, num_iterations=3):\n",
        "    summary_chain = [initial_summary]\n",
        "    \n",
        "    all_entities_dict = {}\n",
        "    clean_entities,  dirty_entities = get_entities(article)\n",
        "    all_entities_dict[\"clean_entities\"] = clean_entities\n",
        "    all_entities_dict[\"dirty_entities\"] = dirty_entities\n",
        "\n",
        "    try:\n",
        "        for _ in range(num_iterations):\n",
        "            missing_entities = [entity for entity in clean_entities if entity not in summary_chain[-1]]\n",
        "            condensed_entities = generate_summary(text=\",\".join(missing_entities), summary_type=\"SPR\")\n",
        "            request = build_sumary_request(article, summary_chain[-1], condensed_entities)\n",
        "            new_summary = generate_summary(text=request, summary_type=\"Entity Dense\")  \n",
        "            summary_chain.append(new_summary)        \n",
        "        return summary_chain[-1], all_entities_dict\n",
        "    except Exception as e:\n",
        "        print(f\"Error in summarizing article: {e}\\n Using last summary\")\n",
        "        # Break out of the loop if there is an error\n",
        "        return summary_chain[-1], all_entities_dict\n",
        "    \n",
        "\n",
        "def get_entities(article: str, model=\"gpt-3.5-turbo-0613\"):\n",
        "    client = instructor.patch(OpenAI(api_key=api_key))\n",
        "\n",
        "    if not article:\n",
        "        raise ValueError(\"Article text cannot be empty\")\n",
        "\n",
        "    entities = []\n",
        "\n",
        "    sentences = split_to_sentences(article)\n",
        "        \n",
        "    chunk_size = 5\n",
        "    overlap = 1\n",
        "    \n",
        "    for i in range(0, len(sentences), chunk_size-overlap): \n",
        "        start = i\n",
        "        end = i + chunk_size\n",
        "        if end > len(sentences):\n",
        "            end = len(sentences)\n",
        "            \n",
        "        chunk = sentences[start:end]\n",
        "        chunk_text = \". \".join(chunk)\n",
        "        try:\n",
        "            response = client.chat.completions.create(\n",
        "                model=model,\n",
        "                messages=[\n",
        "                        {\"role\": \"system\", \"content\": build_system_prompt(prompt_type=\"Get Entities\")},\n",
        "                        {\"role\": \"user\", \"content\": chunk_text}\n",
        "                    ],\n",
        "                temperature=0.7\n",
        "            )\n",
        "\n",
        "            entities.extend(_parse_entities(response))\n",
        "        except Exception as e:\n",
        "            print(f\"Error in extracting entities: {e}\")\n",
        "            # Break out of the loop if there is an error\n",
        "            return None\n",
        "        \n",
        "    return clean_and_separate_entities(entities)\n",
        "    \n",
        "\n",
        "def _parse_entities(response):\n",
        "    # Parses the generated response to extract a list of entity strings\n",
        "    entities = [] \n",
        "    entity_text =  parse_response(response)\n",
        "    #print(f'Entity text: {entity_text}')\n",
        "\n",
        "    # Naive splitting on commas for example output \n",
        "    entities = [e.strip() for e in entity_text.split(\",\")] \n",
        "    entities = [e for e in entities if e]\n",
        "    \n",
        "    return entities\n",
        "\n",
        "\n",
        "def build_knowledge_graph_request(article, clean_entities=None, dirty_entities=None, prev_knowledge=None):\n",
        "        request = f\"Article: {article}\\n\\n\"\n",
        "        if clean_entities:\n",
        "            request += f\"Clean Entities: {clean_entities}\\n\\n\"\n",
        "        if dirty_entities:\n",
        "            request += f\"Dirty Entities: {dirty_entities}\\n\\n\"\n",
        "        if prev_knowledge:\n",
        "            request += f\"Do Not Repeat Previous Knowledge: {prev_knowledge}\\n\\n\"\n",
        "        \n",
        "        client = instructor.patch(OpenAI(api_key=api_key))\n",
        "        try:\n",
        "            response = client.chat.completions.create(\n",
        "                model=\"gpt-3.5-turbo-0613\",\n",
        "                temperature=0.6,\n",
        "                max_retries=3,\n",
        "                messages=[\n",
        "                    {\"role\": \"system\", \"content\": build_system_prompt(prompt_type=\"Get Knowledge\")},\n",
        "                    {\"role\": \"user\", \"content\": request}\n",
        "                ],\n",
        "            )\n",
        "            knowledge = parse_response(response)\n",
        "            return knowledge\n",
        "        except Exception as e:\n",
        "            print(f\"Error in extracting knowledge: {e}\")\n",
        "            # Break out of the loop if there is an error\n",
        "            raise ValueError(\"Error in extracting knowledge\")\n",
        "\n",
        "\n",
        "def build_sumary_request(article, prev_summary, missing_entities):\n",
        "\n",
        "    request = f\"Article: {article}\\n\\n\"\n",
        "    request += f\"Previous Summary: {prev_summary}\\n\\n\" \n",
        "    request += f\"Missing Entities: {missing_entities}\\n\\n\"\n",
        "    return request\n",
        "\n",
        "def split_to_sentences(text):\n",
        "    # logic to split text into sentences \n",
        "    return re.split(r\"[.!?]\\s\", text)\n",
        "\n",
        "   \n",
        "def get_article_chunks(article, chunk_size=800 ):\n",
        "    total_words = count_words(article) \n",
        "    if total_words <= chunk_size:\n",
        "        return [article]\n",
        "    \n",
        "    sentences = split_to_sentences(article)\n",
        "    \n",
        "    chunks = []\n",
        "    current_chunk = []\n",
        "    curr_len = 0\n",
        "    \n",
        "    for sentence in sentences:\n",
        "        sentence_words = count_words(sentence)  \n",
        "        if curr_len + sentence_words < chunk_size:\n",
        "            # add sentence if under chunk size\n",
        "            current_chunk.append(sentence)\n",
        "            curr_len += sentence_words \n",
        "        else:\n",
        "            # otherwise save chunk and reset\n",
        "            chunks.append(\" \".join(current_chunk)) \n",
        "            current_chunk = [sentence]\n",
        "            curr_len = sentence_words\n",
        "            \n",
        "    if current_chunk:\n",
        "        chunks.append(\" \".join(current_chunk))\n",
        "        \n",
        "    return chunks\n",
        "import re\n",
        "\n",
        "def extract_references(file_path):\n",
        "\n",
        "    with open(file_path) as f:\n",
        "        text = f.read() \n",
        "\n",
        "    start_idx = text.find(\"## References\")\n",
        "\n",
        "    if start_idx >= 0:\n",
        "        refs = text[start_idx:]\n",
        "        refs = refs.replace(\"## References\", \"\")\n",
        "        return refs\n",
        "\n",
        "    return \"\"\n",
        "\n",
        "def request_topics(summary):\n",
        "    client = instructor.patch(OpenAI(api_key=api_key))\n",
        "    try:\n",
        "        response = client.chat.completions.create(\n",
        "        model=\"gpt-3.5-turbo-0613\",\n",
        "        temperature=0.4,\n",
        "        max_retries=3,\n",
        "        messages=[\n",
        "            {\"role\": \"system\", \"content\": build_system_prompt(prompt_type=\"Get Topic\")},\n",
        "            {\"role\": \"user\", \"content\": summary}])\n",
        "        topic = extract_topics_with_justification(parse_response(response))\n",
        "    except Exception as e:\n",
        "        print(f\"Error in extracting topics: {e}\")\n",
        "        # Break out of the loop if there is an error\n",
        "        return None\n",
        "    return topic\n",
        "\n",
        "def request_hypothetical_questions(summary):\n",
        "    client = instructor.patch(OpenAI(api_key=api_key))\n",
        "    try:\n",
        "        response = client.chat.completions.create(\n",
        "        model=\"gpt-3.5-turbo-0613\",\n",
        "        temperature=0.4,\n",
        "        max_retries=3,\n",
        "        messages=[\n",
        "            {\"role\": \"system\", \"content\": build_system_prompt(prompt_type=\"Get Hypothetical Questions\")},\n",
        "            {\"role\": \"user\", \"content\": summary}])\n",
        "        questions = extract_hypothetical_questions(parse_response(response))\n",
        "        #questions = parse_response(response)\n",
        "    except Exception as e:\n",
        "        print(f\"Error in extracting hypothetical questions: {e}\")\n",
        "        # Break out of the loop if there is an error\n",
        "        return None\n",
        "    return questions\n",
        "\n",
        "def extract_info(summary):\n",
        "    # NLP logic to extract topic and hypothetical questions \n",
        "    while True:\n",
        "        topic = request_topics(summary)\n",
        "        if topic:\n",
        "            break\n",
        "    while True:\n",
        "        questions = request_hypothetical_questions(summary)\n",
        "        if questions:\n",
        "            break\n",
        "    return topic, questions\n",
        "\n",
        "def extract_knowledge(article, clean_entities, dirty_entities):\n",
        "    # NLP logic to extract knowledge from the article\n",
        "    knowledge = \"\"\n",
        "    if not article:\n",
        "        raise ValueError(\"Article text cannot be empty\")\n",
        "\n",
        "    try:\n",
        "        clean_knowledge = build_knowledge_graph_request(article=article, clean_entities=clean_entities)\n",
        "        knowledge += clean_knowledge\n",
        "    except Exception as e:\n",
        "        print(f\"Error in extracting clean knowledge: {e}\\n Trying again\")\n",
        "        clean_knowledge = build_knowledge_graph_request(article=article, clean_entities=clean_entities)\n",
        "        knowledge += clean_knowledge\n",
        "    try:\n",
        "        dirty_knowledge = build_knowledge_graph_request(article=article, dirty_entities=dirty_entities)\n",
        "        knowledge += dirty_knowledge\n",
        "    except Exception as e:\n",
        "        print(f\"Error in extracting dirty knowledge: {e}\\n Trying again\")\n",
        "        dirty_knowledge = build_knowledge_graph_request(article=article, dirty_entities=dirty_entities)\n",
        "        knowledge += dirty_knowledge\n",
        "\n",
        "    try:\n",
        "        combined_knowledge = build_knowledge_graph_request(article=article, prev_knowledge=knowledge)\n",
        "        knowledge += combined_knowledge\n",
        "    except Exception as e:\n",
        "        print(f\"Error in extracting combined knowledge: {e}\\n Trying again\")\n",
        "        combined_knowledge = build_knowledge_graph_request(article=article, prev_knowledge=knowledge)\n",
        "        knowledge += combined_knowledge\n",
        "    return clean_entity_relationships(knowledge)\n",
        "\n",
        "def Incrementally_Refine_Article_Summary(article_info):\n",
        "    file_id = article_info[\"file_id\"]\n",
        "    file_path = article_info[\"file_path\"]\n",
        "    \n",
        "    store = SummaryStore(file_id)\n",
        "\n",
        "    references = extract_references(file_path)\n",
        "    urls = extract_urls(references)\n",
        "    if urls:\n",
        "        # create dictionary of urls and references\n",
        "        references = {\"urls\": urls, \"references\": references}\n",
        "    #print(f\"References: {references}\")\n",
        "\n",
        "    with open(file_path) as f:\n",
        "        article_text = f.read()\n",
        "    \n",
        "    article_chunks = get_article_chunks(article_text)\n",
        "\n",
        "    try:\n",
        "        chunk_num = 0\n",
        "        for chunk in article_chunks:\n",
        "            # Generate an initial summary for each chunk\n",
        "            initial_summary = generate_summary(text=chunk, summary_type=\"SPR\")\n",
        "            \n",
        "            # Generate a refined summary for each chunk\n",
        "            refined_sumary, entities = get_entity_dense_sumary(chunk, initial_summary)\n",
        "\n",
        "            # Extract Knowledge from the article and entities\n",
        "            knowledge_triplets = extract_knowledge(chunk, clean_entities=entities[\"clean_entities\"], dirty_entities=entities[\"dirty_entities\"])\n",
        "\n",
        "            # Extract the topic and hypothetical questions from the refined summary\n",
        "            topic, questions = extract_info(refined_sumary)\n",
        "\n",
        "            # Store the summary, entities, and citation\n",
        "            chunk_name = f\"Chunk # {chunk_num}.\\n{chunk}\"\n",
        "            \n",
        "            store.store(summary=refined_sumary, \n",
        "                        file_id=file_id, \n",
        "                        clean_entities=entities[\"clean_entities\"],\n",
        "                        dirty_entities=entities[\"dirty_entities\"],\n",
        "                        article=chunk_name, \n",
        "                        references=references, \n",
        "                        topic=topic, \n",
        "                        hypothetical_questions=questions,\n",
        "                        knowledge=knowledge_triplets\n",
        "                        )\n",
        "            chunk_num += 1\n",
        "        # return success\n",
        "        return True\n",
        "\n",
        "    except Exception as e: \n",
        "        print(f\"Error summarizing {file_path}: {e}\")\n",
        "        return None\n",
        "\n",
        "\n",
        "import codecs\n",
        "\n",
        "def is_bibliography(file_path):\n",
        "\n",
        "    with codecs.open(file_path, 'rb') as f:\n",
        "        first_line = f.readline()\n",
        "        if b'# Bibliography Recommendation Report:' in first_line:\n",
        "            return True\n",
        "    return False\n",
        "\n",
        "def get_article_list():\n",
        "    articles = []\n",
        "    \n",
        "    for file_name in os.listdir(OUTPUT_FOLDER):\n",
        "        file_path = os.path.join(OUTPUT_FOLDER, file_name)\n",
        "        \n",
        "        # Skip bibliography files\n",
        "        if is_bibliography(file_path):\n",
        "            continue\n",
        "            \n",
        "        if file_name.endswith(\".md\"):\n",
        "            file_id = get_file_id(file_name)\n",
        "            summary_file_path = os.path.join(OUTPUT_FOLDER, f\"{file_id}.json\")\n",
        "\n",
        "            if not os.path.exists(summary_file_path):\n",
        "                info = {\n",
        "                    \"file_id\": file_id, \n",
        "                    \"file_path\": file_path\n",
        "                }\n",
        "                articles.append(info)\n",
        "\n",
        "    return articles\n",
        "\n",
        "def get_file_id(file_name):\n",
        "    # Extract base name without extension\n",
        "    return os.path.splitext(file_name)[0]\n",
        "\n",
        "article_list = get_article_list()\n",
        "if article_list:\n",
        "    for article_info in article_list:\n",
        "        print(f\"Summarizing {article_info['file_path']}\")\n",
        "        success = Incrementally_Refine_Article_Summary(article_info)\n",
        "        if success:\n",
        "            print(f\"Successfully summarized {article_info['file_path']}\")\n",
        "        else:\n",
        "            print(f\"Error summarizing {article_info['file_path']}\")\n",
        "else:\n",
        "    print(\"No articles to summarize\")\n",
        "# open summary.json to see the results \n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Neo4j Graph Database"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "from neo4j import GraphDatabase\n",
        "\n",
        "uri = \"bolt://localhost:7687\"  # Replace with your Neo4j instance URI\n",
        "username = \"neo4j\"  # Replace with your username\n",
        "password = \"password\"  # Replace with your password\n",
        "\n",
        "driver = GraphDatabase.driver(uri, auth=(username, password))\n",
        "\n",
        "def load_summaries_to_neo4j(summaries, driver):\n",
        "    with driver.session() as session:\n",
        "        for summary in summaries:\n",
        "            # Call the add_summary function within a Neo4j transaction\n",
        "            session.write_transaction(add_summary, summary)\n",
        "\n",
        "def add_summary(tx, summary_data):\n",
        "    # Create Article node\n",
        "    tx.run(\"MERGE (a:Article {id: $file_id, content: $article})\", summary_data)\n",
        "\n",
        "    # Create Summary node and link to Article\n",
        "    tx.run(\"MERGE (s:Summary {content: $summary}) \"\n",
        "           \"MERGE (a)-[:HAS_SUMMARY]->(s)\", summary_data)\n",
        "\n",
        "    # Handling arrays in summary_data\n",
        "    handle_arrays(tx, summary_data, 'clean_entities', 'Entity', 'MENTIONS', 'e', 'a')\n",
        "    handle_arrays(tx, summary_data, 'dirty_entities', 'DirtyEntity', 'MENTIONS_DIRTY', 'de', 'a')\n",
        "    handle_arrays(tx, summary_data, 'topics', 'Topic', 'HAS_TOPIC', 't', 'a')\n",
        "    handle_arrays(tx, summary_data, 'hypothetical_questions', 'HypotheticalQuestion', 'HAS_QUESTION', 'q', 'a')\n",
        "    \n",
        "    # Handle knowledge triplets\n",
        "    if 'knowledge_triplets' in summary_data:\n",
        "        for triplet in summary_data[\"knowledge_triplets\"]:\n",
        "            tx.run(\"MERGE (s:Subject {name: $subject}) \"\n",
        "                   \"MERGE (t:Target {name: $target}) \"\n",
        "                   \"MERGE (s)-[:RELATIONSHIP {type: $relationship}]->(t)\",\n",
        "                   {'subject': triplet['subject'], 'relationship': triplet['relationship'], 'target': triplet['target']})\n",
        "\n",
        "    # Add references (assuming references are a list of URLs or texts)\n",
        "    if 'references' in summary_data and 'urls' in summary_data['references']:\n",
        "        for url in summary_data[\"references\"][\"urls\"]:\n",
        "            tx.run(\"MERGE (r:Reference {url: $url}) \"\n",
        "                   \"MERGE (a)-[:HAS_REFERENCE]->(r)\", {'url': url, 'article': summary_data['article']})\n",
        "\n",
        "def handle_arrays(tx, summary_data, key, node_type, relationship, node_alias, article_alias):\n",
        "    if key in summary_data:\n",
        "        for item in summary_data[key]:\n",
        "            # For topics with justifications\n",
        "            if key == 'topics' and isinstance(item, dict):\n",
        "                tx.run(\"MERGE (t:Topic {name: $topic}) \"\n",
        "                       \"MERGE (s)-[:HAS_TOPIC {justification: $justification}]->(t)\",\n",
        "                       {'topic': item['topic'], 'justification': item['justification'], 's': summary_data['summary']})\n",
        "\n",
        "            # For hypothetical questions with question types\n",
        "            elif key == 'hypothetical_questions' and isinstance(item, dict):\n",
        "                tx.run(\"MERGE (hq:HypotheticalQuestion {content: $question}) \"\n",
        "                       \"MERGE (s)-[:HAS_QUESTION {type: $question_type}]->(hq)\",\n",
        "                       {'question': item['question'], 'question_type': item['question_type'], 's': summary_data['summary']})\n",
        "\n",
        "            # For simple string lists like clean and dirty entities\n",
        "            elif isinstance(item, str):\n",
        "                tx.run(f\"MERGE ({node_alias}:{node_type} {{name: $name}}) \"\n",
        "                       f\"MERGE (a)-[:{relationship}]->({node_alias})\",\n",
        "                       {'name': item, 'a': summary_data['article']})\n",
        "\n",
        "\n",
        "# Load summaries from JSON folder\n",
        "summaries = []\n",
        "for file_name in os.listdir(OUTPUT_FOLDER):\n",
        "    file_path = os.path.join(OUTPUT_FOLDER, file_name)\n",
        "    if file_name.endswith(\".json\"):\n",
        "        with open(file_path) as f:\n",
        "            summaries.extend(json.load(f))\n",
        "\n",
        "# Load summaries to Neo4j\n",
        "load_summaries_to_neo4j(summaries, driver)\n",
        "\n",
        "# Close the driver connection\n",
        "\n",
        "driver.close()\n",
        "\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "collapsed_sections": [
        "5ecdPw0gF6Q0",
        "D30JwOfQWT_u",
        "zv0KV0sUFH4Z",
        "0yeeQ2K-GD10",
        "snYDZh2OG6ns",
        "1h43Eh0PSDHb",
        "F9zcFI2FHiIJ",
        "2g8Xb9JmIq3f",
        "KNp6_8aRJa-m",
        "D9qdcO7XKY35",
        "Kdis0QlPJ3-k",
        "KUbjOcq8LR0A",
        "x0zLIv1i75gJ",
        "tWpIbwqlLkKH",
        "mdwmgmy2MCxt",
        "DBbpIf8EOmCH"
      ],
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.13"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
